<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Deep Learning Tutorial &#8212; CoderData</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=d75fae25" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinxdoc.css?v=34905f61" />
    <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css?v=2aa19091" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=08cfd9eb" />
    <script src="../_static/documentation_options.js?v=8d563738"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="canonical" href="https://pnnl-compbio.github.io/coderdata/_notebooks/docsite_tutorial.html" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="How to Contribute" href="../contribution.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="Related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../contribution.html" title="How to Contribute"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">CoderData</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Deep Learning Tutorial</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="Deep-Learning-Tutorial">
<h1>Deep Learning Tutorial<a class="headerlink" href="#Deep-Learning-Tutorial" title="Link to this heading">¶</a></h1>
<p>In this tutorial we will demonstrate how to create a basic deep learning model using train, test, and validation sets. For the model a subset of the bladderpdo dataset is used where transcriptomics and drug structures are the inputs and drug response values are the output. Throughout the tutorial we will be using functions included in coderdata. For more infomration on functions used from the package look to the <a class="reference external" href="../APIreference.html">coderdata/API reference</a>.</p>
<p>Import coderdata, pandas, and numpy.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">coderdata</span> <span class="k">as</span> <span class="nn">cd</span>
</pre></div>
</div>
</div>
<section id="Download-and-View-Datasets">
<h2>Download and View Datasets<a class="headerlink" href="#Download-and-View-Datasets" title="Link to this heading">¶</a></h2>
<p>To view the availabe datasets within the coderdata API use the following functions.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span><span class="o">.</span><span class="n">list_datasets</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
beataml: Beat acute myeloid leukemia (BeatAML) focuses on acute myeloid leukemia tumor data. Data includes drug response, proteomics, and transcriptomics datasets.
bladderpdo: Tumor Evolution and Drug Response in Patient-Derived Organoid Models of Bladder Cancer Data includes transcriptomics, mutations, copy number, and drug response data.
ccle: Cancer Cell Line Encyclopedia (CCLE).
cptac: The Clinical Proteomic Tumor Analysis Consortium (CPTAC) project is a collaborative network funded by the National Cancer Institute (NCI) focused on improving our understanding of cancer biology through the integration of transcriptomic, proteomic, and genomic data.
ctrpv2: Cancer Therapeutics Response Portal version 2 (CTRPv2)
fimm: Institute for Molecular Medicine Finland (FIMM) dataset.
gcsi: The Genentech Cell Line Screening Initiative (gCSI)
gdscv1: Genomics of Drug Sensitivity in Cancer (GDSC) v1
gdscv2: Genomics of Drug Sensitivity in Cancer (GDSC) v2
hcmi: Human Cancer Models Initiative (HCMI) encompasses numerous cancer types and includes cell line, organoid, and tumor data. Data includes the transcriptomics, somatic mutation, and copy number datasets.
mpnst: Malignant Peripheral Nerve Sheath Tumor is a rare, agressive sarcoma that affects peripheral nerves throughout the body.
mpnstpdx: Patient derived xenograft data for MPNST
nci60: National Cancer Institute 60
pancpdo: Organoid Profiling Identifies Common Responders to Chemotherapy in Pancreatic Cancer Data includes transcriptomics, mutations, copy number, and drug response data.
prism: Profiling Relative Inhibition Simultaneously in Mixtures
sarcpdo: The landscape of drug sensitivity and resistance in sarcoma Data includes transcriptomics, mutations, and drug response data.
</pre></div></div>
</div>
<p>To download and load the needed dataset(s) use the function below. For this example we will be using bladderpdo.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="s2">&quot;bladderpdo&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line">Example Output:</div>
<div class="line">Downloaded ‘<a class="reference external" href="https://ndownloader.figshare.com/files/53779532">https://ndownloader.figshare.com/files/53779532</a>’ to ‘c:<span class="math">\Path</span><span class="math">\genes</span>.csv.gz’</div>
<div class="line">Downloaded ‘<a class="reference external" href="https://ndownloader.figshare.com/files/53956211">https://ndownloader.figshare.com/files/53956211</a>’ to ‘c:<span class="math">\Path</span><span class="math">\bladderpdo</span>_copy_number.csv.gz’</div>
<div class="line">Downloaded ‘<a class="reference external" href="https://ndownloader.figshare.com/files/53956214">https://ndownloader.figshare.com/files/53956214</a>’ to ‘c:<span class="math">\Path</span><span class="math">\bladderpdo</span>_drugs.tsv.gz’</div>
<div class="line">Downloaded ‘<a class="reference external" href="https://ndownloader.figshare.com/files/53956217">https://ndownloader.figshare.com/files/53956217</a>’ to ‘c:<span class="math">\Path</span><span class="math">\bladderpdo</span>_drug_descriptors.tsv.gz</div>
<div class="line">Downloaded ‘<a class="reference external" href="https://ndownloader.figshare.com/files/53956220">https://ndownloader.figshare.com/files/53956220</a>’ to ‘c:<span class="math">\Path</span><span class="math">\bladderpdo</span>_experiments.tsv.gz’</div>
<div class="line">Downloaded ‘<a class="reference external" href="https://ndownloader.figshare.com/files/53956223">https://ndownloader.figshare.com/files/53956223</a>’ to ‘c:<span class="math">\Path</span><span class="math">\bladderpdo</span>_mutations.csv.gz’</div>
<div class="line">Downloaded ‘<a class="reference external" href="https://ndownloader.figshare.com/files/53956226">https://ndownloader.figshare.com/files/53956226</a>’ to ‘c:<span class="math">\Path</span><span class="math">\bladderpdo</span>_samples.csv’</div>
<div class="line">Downloaded ‘<a class="reference external" href="https://ndownloader.figshare.com/files/53956232">https://ndownloader.figshare.com/files/53956232</a>’ to ‘c:<span class="math">\Path</span><span class="math">\bladderpdo</span>_transcriptomics.csv.gz’</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bladderpdo</span><span class="o">=</span><span class="n">cd</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;bladderpdo&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line">Example Output:</div>
<div class="line">Importing raw data …</div>
<div class="line">Importing ‘copy_number’ from c:<span class="math">\Path</span><span class="math">\bladderpdo</span>_copy_number.csv.gz … DONE</div>
<div class="line">Importing ‘drugs’ from c:<span class="math">\Path</span><span class="math">\bladderpdo</span>_drugs.tsv.gz … DONE</div>
<div class="line">Importing ‘drug_descriptors’ from c:<span class="math">\Path</span><span class="math">\bladderpdo</span>_drug_descriptors.tsv.gz … DONE</div>
<div class="line">Importing ‘experiments’ from c:<span class="math">\Path</span><span class="math">\bladderpdo</span>_experiments.tsv.gz … DONE</div>
<div class="line">Importing ‘mutations’ from c:<span class="math">\Path</span><span class="math">\bladderpdo</span>_mutations.csv.gz … DONE</div>
<div class="line">Importing ‘samples’ from c:<span class="math">\Path</span><span class="math">\bladderpdo</span>_samples.csv … DONE</div>
<div class="line">Importing ‘transcriptomics’ from c:<span class="math">\Path</span><span class="math">\bladderpdo</span>_transcriptomics.csv.gz … DONE</div>
<div class="line">Importing ‘genes’ from c:<span class="math">\Path</span><span class="math">\genes</span>.csv.gz … DONE</div>
<div class="line">Importing raw data … DONE</div>
</div>
<p>The function <code class="docutils literal notranslate"><span class="pre">type()</span></code> can be used to view what the class of any object is as seen below, bladderpdo is a dataset.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">type</span><span class="p">(</span><span class="n">bladderpdo</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
coderdata.dataset.dataset.Dataset
</pre></div></div>
</div>
<p>To view all the datatypes included with a dataset object use the function below. The <code class="docutils literal notranslate"><span class="pre">info()</span></code> command allows us to see which datatypes are included in the bladderpdo dataset object. We will be using the Transcriptomics, Drugs, and Experiments Data in our example. Note that each datatype has different format options and arguments. To view all options for each type navigate to the <a class="reference external" href="../usage.html">usage</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;datatypes&quot;</span><span class="p">:</span><span class="n">bladderpdo</span><span class="o">.</span><span class="n">types</span><span class="p">()})</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>datatypes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>transcriptomics</td>
    </tr>
    <tr>
      <th>1</th>
      <td>mutations</td>
    </tr>
    <tr>
      <th>2</th>
      <td>copy_number</td>
    </tr>
    <tr>
      <th>3</th>
      <td>samples</td>
    </tr>
    <tr>
      <th>4</th>
      <td>drugs</td>
    </tr>
    <tr>
      <th>5</th>
      <td>experiments</td>
    </tr>
    <tr>
      <th>6</th>
      <td>genes</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Examples are shown below of how to view a regular dataset and the specificed datatype.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bladderpdo</span><span class="o">.</span><span class="n">drugs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>improve_drug_id</th>
      <th>chem_name</th>
      <th>pubchem_id</th>
      <th>canSMILES</th>
      <th>InChIKey</th>
      <th>formula</th>
      <th>weight</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>SMI_8294</td>
      <td>bi-2536</td>
      <td>11364421</td>
      <td>CCC1C(=O)N(C2=CN=C(N=C2N1C3CCCC3)NC4=C(C=C(C=C...</td>
      <td>XQVVPGYIWAGRNI-JOCHJYFZSA-N</td>
      <td>C28H39N7O3</td>
      <td>521.7</td>
    </tr>
    <tr>
      <th>1</th>
      <td>SMI_8294</td>
      <td>((r)-4-[(8-cyclopentyl-7-et-5,6,7,8-tetrahydro...</td>
      <td>11364421</td>
      <td>CCC1C(=O)N(C2=CN=C(N=C2N1C3CCCC3)NC4=C(C=C(C=C...</td>
      <td>XQVVPGYIWAGRNI-JOCHJYFZSA-N</td>
      <td>C28H39N7O3</td>
      <td>521.7</td>
    </tr>
    <tr>
      <th>2</th>
      <td>SMI_8294</td>
      <td>en300-18167228</td>
      <td>11364421</td>
      <td>CCC1C(=O)N(C2=CN=C(N=C2N1C3CCCC3)NC4=C(C=C(C=C...</td>
      <td>XQVVPGYIWAGRNI-JOCHJYFZSA-N</td>
      <td>C28H39N7O3</td>
      <td>521.7</td>
    </tr>
    <tr>
      <th>3</th>
      <td>SMI_8294</td>
      <td>4-((r)-8-cyclopentyl-7-ethyl-5-methyl-6-oxo-5,...</td>
      <td>11364421</td>
      <td>CCC1C(=O)N(C2=CN=C(N=C2N1C3CCCC3)NC4=C(C=C(C=C...</td>
      <td>XQVVPGYIWAGRNI-JOCHJYFZSA-N</td>
      <td>C28H39N7O3</td>
      <td>521.7</td>
    </tr>
    <tr>
      <th>4</th>
      <td>SMI_8294</td>
      <td>bcpp000342</td>
      <td>11364421</td>
      <td>CCC1C(=O)N(C2=CN=C(N=C2N1C3CCCC3)NC4=C(C=C(C=C...</td>
      <td>XQVVPGYIWAGRNI-JOCHJYFZSA-N</td>
      <td>C28H39N7O3</td>
      <td>521.7</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>5015</th>
      <td>SMI_438</td>
      <td>metatrexan</td>
      <td>126941</td>
      <td>CN(CC1=CN=C2C(=N1)C(=NC(=N2)N)N)C3=CC=C(C=C3)C...</td>
      <td>FBOZXECLQNJBKD-ZDUSSCGKSA-N</td>
      <td>C20H22N8O5</td>
      <td>454.4</td>
    </tr>
    <tr>
      <th>5016</th>
      <td>SMI_438</td>
      <td>hms2233o18</td>
      <td>126941</td>
      <td>CN(CC1=CN=C2C(=N1)C(=NC(=N2)N)N)C3=CC=C(C=C3)C...</td>
      <td>FBOZXECLQNJBKD-ZDUSSCGKSA-N</td>
      <td>C20H22N8O5</td>
      <td>454.4</td>
    </tr>
    <tr>
      <th>5017</th>
      <td>SMI_438</td>
      <td>amethopterin l-</td>
      <td>126941</td>
      <td>CN(CC1=CN=C2C(=N1)C(=NC(=N2)N)N)C3=CC=C(C=C3)C...</td>
      <td>FBOZXECLQNJBKD-ZDUSSCGKSA-N</td>
      <td>C20H22N8O5</td>
      <td>454.4</td>
    </tr>
    <tr>
      <th>5018</th>
      <td>SMI_438</td>
      <td>tocris-1230</td>
      <td>126941</td>
      <td>CN(CC1=CN=C2C(=N1)C(=NC(=N2)N)N)C3=CC=C(C=C3)C...</td>
      <td>FBOZXECLQNJBKD-ZDUSSCGKSA-N</td>
      <td>C20H22N8O5</td>
      <td>454.4</td>
    </tr>
    <tr>
      <th>5019</th>
      <td>SMI_438</td>
      <td>amethopterin (hydrate); cl14377 (hydrate); wr1...</td>
      <td>126941</td>
      <td>CN(CC1=CN=C2C(=N1)C(=NC(=N2)N)N)C3=CC=C(C=C3)C...</td>
      <td>FBOZXECLQNJBKD-ZDUSSCGKSA-N</td>
      <td>C20H22N8O5</td>
      <td>454.4</td>
    </tr>
  </tbody>
</table>
<p>5020 rows × 7 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bladderpdo</span><span class="o">.</span><span class="n">transcriptomics</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>entrez_id</th>
      <th>transcriptomics</th>
      <th>improve_sample_id</th>
      <th>source</th>
      <th>study</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>7105</td>
      <td>69.277129</td>
      <td>5553</td>
      <td>Gene Expression Omnibus</td>
      <td>Lee et al. 2018 Bladder PDOs</td>
    </tr>
    <tr>
      <th>1</th>
      <td>64102</td>
      <td>0.027787</td>
      <td>5553</td>
      <td>Gene Expression Omnibus</td>
      <td>Lee et al. 2018 Bladder PDOs</td>
    </tr>
    <tr>
      <th>2</th>
      <td>8813</td>
      <td>81.403736</td>
      <td>5553</td>
      <td>Gene Expression Omnibus</td>
      <td>Lee et al. 2018 Bladder PDOs</td>
    </tr>
    <tr>
      <th>3</th>
      <td>57147</td>
      <td>20.024448</td>
      <td>5553</td>
      <td>Gene Expression Omnibus</td>
      <td>Lee et al. 2018 Bladder PDOs</td>
    </tr>
    <tr>
      <th>4</th>
      <td>55732</td>
      <td>29.819016</td>
      <td>5553</td>
      <td>Gene Expression Omnibus</td>
      <td>Lee et al. 2018 Bladder PDOs</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>816096</th>
      <td>140032</td>
      <td>0.024479</td>
      <td>5502</td>
      <td>Gene Expression Omnibus</td>
      <td>Lee et al. 2018 Bladder PDOs</td>
    </tr>
    <tr>
      <th>816097</th>
      <td>9782</td>
      <td>44.949383</td>
      <td>5502</td>
      <td>Gene Expression Omnibus</td>
      <td>Lee et al. 2018 Bladder PDOs</td>
    </tr>
    <tr>
      <th>816098</th>
      <td>102724737</td>
      <td>0.024479</td>
      <td>5502</td>
      <td>Gene Expression Omnibus</td>
      <td>Lee et al. 2018 Bladder PDOs</td>
    </tr>
    <tr>
      <th>816099</th>
      <td>101669762</td>
      <td>2.203401</td>
      <td>5502</td>
      <td>Gene Expression Omnibus</td>
      <td>Lee et al. 2018 Bladder PDOs</td>
    </tr>
    <tr>
      <th>816100</th>
      <td>100527949</td>
      <td>0.024479</td>
      <td>5502</td>
      <td>Gene Expression Omnibus</td>
      <td>Lee et al. 2018 Bladder PDOs</td>
    </tr>
  </tbody>
</table>
<p>816101 rows × 5 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bladderpdo</span><span class="o">.</span><span class="n">experiments</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>source</th>
      <th>improve_sample_id</th>
      <th>improve_drug_id</th>
      <th>study</th>
      <th>time</th>
      <th>time_unit</th>
      <th>dose_response_metric</th>
      <th>dose_response_value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Synapse</td>
      <td>5473</td>
      <td>SMI_32023</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>fit_auc</td>
      <td>0.7725</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Synapse</td>
      <td>5473</td>
      <td>SMI_55473</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>fit_auc</td>
      <td>0.9277</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Synapse</td>
      <td>5473</td>
      <td>SMI_27249</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>fit_auc</td>
      <td>0.8538</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Synapse</td>
      <td>5473</td>
      <td>SMI_55606</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>fit_auc</td>
      <td>0.8478</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Synapse</td>
      <td>5473</td>
      <td>SMI_20004</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>fit_auc</td>
      <td>0.8825</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>32995</th>
      <td>Synapse</td>
      <td>5606</td>
      <td>SMI_438</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>dss</td>
      <td>0.1221</td>
    </tr>
    <tr>
      <th>32996</th>
      <td>Synapse</td>
      <td>5606</td>
      <td>SMI_52219</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>dss</td>
      <td>0.1104</td>
    </tr>
    <tr>
      <th>32997</th>
      <td>Synapse</td>
      <td>5606</td>
      <td>SMI_34999</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>dss</td>
      <td>0.1202</td>
    </tr>
    <tr>
      <th>32998</th>
      <td>Synapse</td>
      <td>5606</td>
      <td>SMI_55689</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>dss</td>
      <td>0.0000</td>
    </tr>
    <tr>
      <th>32999</th>
      <td>Synapse</td>
      <td>5606</td>
      <td>SMI_10376</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>dss</td>
      <td>0.0000</td>
    </tr>
  </tbody>
</table>
<p>33000 rows × 8 columns</p>
</div></div>
</div>
<p>Coderdata has a function <code class="docutils literal notranslate"><span class="pre">format()</span></code> that will reformat any dataset.datatype with the specificed arguments. Each datatype has different arguments specific to its contents and to view each datatypes argument options visit <a class="reference external" href="../usage.html">coderdata/usage</a>. Here we will use the bladderpdo.experiment to format. We want to only include <em>auc</em> metric values in our dataframe for this tutorial, however, to view other metric options for a dataset use the pandas unique function.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;dose_response_metric&quot;</span><span class="p">:</span><span class="n">bladderpdo</span><span class="o">.</span><span class="n">experiments</span><span class="o">.</span><span class="n">dose_response_metric</span><span class="o">.</span><span class="n">unique</span><span class="p">()})</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dose_response_metric</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>fit_auc</td>
    </tr>
    <tr>
      <th>1</th>
      <td>fit_ic50</td>
    </tr>
    <tr>
      <th>2</th>
      <td>fit_ec50</td>
    </tr>
    <tr>
      <th>3</th>
      <td>fit_r2</td>
    </tr>
    <tr>
      <th>4</th>
      <td>fit_ec50se</td>
    </tr>
    <tr>
      <th>5</th>
      <td>fit_einf</td>
    </tr>
    <tr>
      <th>6</th>
      <td>fit_hs</td>
    </tr>
    <tr>
      <th>7</th>
      <td>aac</td>
    </tr>
    <tr>
      <th>8</th>
      <td>auc</td>
    </tr>
    <tr>
      <th>9</th>
      <td>dss</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Set <code class="docutils literal notranslate"><span class="pre">bladderpdo.experiments</span></code> to only include samples where the <code class="docutils literal notranslate"><span class="pre">dose_response_metric</span></code> is <em>auc</em>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bladderpdo</span><span class="o">.</span><span class="n">experiments</span> <span class="o">=</span> <span class="n">bladderpdo</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">data_type</span><span class="o">=</span><span class="s1">&#39;experiments&#39;</span><span class="p">,</span> <span class="n">metrics</span><span class="o">=</span><span class="s1">&#39;auc&#39;</span><span class="p">)</span>
<span class="n">bladderpdo</span><span class="o">.</span><span class="n">experiments</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>source</th>
      <th>improve_sample_id</th>
      <th>improve_drug_id</th>
      <th>study</th>
      <th>time</th>
      <th>time_unit</th>
      <th>dose_response_metric</th>
      <th>dose_response_value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>26400</th>
      <td>Synapse</td>
      <td>5473</td>
      <td>SMI_32023</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.7530</td>
    </tr>
    <tr>
      <th>26401</th>
      <td>Synapse</td>
      <td>5473</td>
      <td>SMI_55473</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.9067</td>
    </tr>
    <tr>
      <th>26402</th>
      <td>Synapse</td>
      <td>5473</td>
      <td>SMI_27249</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.8972</td>
    </tr>
    <tr>
      <th>26403</th>
      <td>Synapse</td>
      <td>5473</td>
      <td>SMI_55606</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.8320</td>
    </tr>
    <tr>
      <th>26404</th>
      <td>Synapse</td>
      <td>5473</td>
      <td>SMI_20004</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.8937</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>29695</th>
      <td>Synapse</td>
      <td>5606</td>
      <td>SMI_438</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.4095</td>
    </tr>
    <tr>
      <th>29696</th>
      <td>Synapse</td>
      <td>5606</td>
      <td>SMI_52219</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.6260</td>
    </tr>
    <tr>
      <th>29697</th>
      <td>Synapse</td>
      <td>5606</td>
      <td>SMI_34999</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.8730</td>
    </tr>
    <tr>
      <th>29698</th>
      <td>Synapse</td>
      <td>5606</td>
      <td>SMI_55689</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.9406</td>
    </tr>
    <tr>
      <th>29699</th>
      <td>Synapse</td>
      <td>5606</td>
      <td>SMI_10376</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.9813</td>
    </tr>
  </tbody>
</table>
<p>3300 rows × 8 columns</p>
</div></div>
</div>
</section>
</section>
<section id="Develop-Deep-Learning-Model">
<h1>Develop Deep Learning Model<a class="headerlink" href="#Develop-Deep-Learning-Model" title="Link to this heading">¶</a></h1>
<div class="line-block">
<div class="line">We will be using the keras functional API, for information on the keras API visit <a class="reference external" href="https://keras.io/api">keras.io</a>. This allows us to use multiple inputs in our model. You may need to install a couple of these dependencies to proceed.</div>
<div class="line">Dependencies that are not included in coderdata:</div>
</div>
<ul class="simple">
<li><p>tensorflow (Note: tensorflow is only supported up to the python version 3.12)</p></li>
<li><p>matplotlib</p></li>
<li><p>rdkit</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">keras</span>
<span class="c1">#from numpy import loadtxt</span>
<span class="kn">from</span> <span class="nn">tensorflow.keras.models</span> <span class="kn">import</span> <span class="n">Sequential</span>
<span class="kn">from</span> <span class="nn">tensorflow.keras.layers</span> <span class="kn">import</span> <span class="n">Dense</span>
<span class="kn">from</span> <span class="nn">sklearn.preprocessing</span> <span class="kn">import</span> <span class="n">MinMaxScaler</span>
<span class="c1">#from sklearn.model_selection import train_test_split</span>
<span class="kn">from</span> <span class="nn">rdkit</span> <span class="kn">import</span> <span class="n">Chem</span>
<span class="kn">from</span> <span class="nn">rdkit.Chem</span> <span class="kn">import</span> <span class="n">AllChem</span>
<span class="kn">from</span> <span class="nn">rdkit.Chem.rdFingerprintGenerator</span> <span class="kn">import</span> <span class="n">GetMorganGenerator</span>
<span class="kn">from</span> <span class="nn">keras.models</span> <span class="kn">import</span> <span class="n">Model</span>
<span class="kn">from</span> <span class="nn">keras.layers</span> <span class="kn">import</span> <span class="n">Input</span><span class="p">,</span> <span class="n">Dense</span><span class="p">,</span> <span class="n">concatenate</span>
<span class="kn">from</span> <span class="nn">keras.optimizers</span> <span class="kn">import</span> <span class="n">Adam</span>
<span class="kn">from</span> <span class="nn">keras.losses</span> <span class="kn">import</span> <span class="n">MeanSquaredError</span>
<span class="kn">from</span> <span class="nn">keras.metrics</span> <span class="kn">import</span> <span class="n">MeanAbsoluteError</span>
<span class="kn">from</span> <span class="nn">keras</span> <span class="kn">import</span> <span class="n">layers</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="kn">import</span> <span class="n">mean_absolute_error</span><span class="p">,</span> <span class="n">mean_squared_error</span><span class="p">,</span> <span class="n">r2_score</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">sklearn.linear_model</span> <span class="kn">import</span> <span class="n">LinearRegression</span>
</pre></div>
</div>
</div>
<p>Coderdata has 2 split functions a two-way and three-way split that can be used with any dataset. Below are example of how to use both functions but for this tutorial we will be using the three-way splits for our model.</p>
<p><strong>Quick tip:</strong> To view all available arguments for any function use the <code class="docutils literal notranslate"><span class="pre">help()</span></code> function below.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">bladderpdo</span><span class="o">.</span><span class="n">split_train_other</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Help on method split_train_other in module coderdata.dataset.dataset:

split_train_other(split_type: &#34;Literal[&#39;mixed-set&#39;, &#39;drug-blind&#39;, &#39;cancer-blind&#39;]&#34; = &#39;mixed-set&#39;, ratio: &#39;tuple[int, int]&#39; = (8, 2), stratify_by: &#39;Optional[str]&#39; = None, balance: &#39;bool&#39; = False, random_state: &#39;Optional[Union[int, RandomState]]&#39; = None, **kwargs: &#39;dict&#39;) -&gt; &#39;TwoWaySplit&#39; method of coderdata.dataset.dataset.Dataset instance

</pre></div></div>
</div>
<p>Here is an example of a two-way split that is trained on bladderpdo dataset dataframe. To view the definitions of different <code class="docutils literal notranslate"><span class="pre">split_type</span></code> and the associated arguments view the <a class="reference external" href="../APIreference.html">API reference</a>. Note that the other split functions will use similar arguments.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">two_way_split</span><span class="o">=</span> <span class="n">bladderpdo</span><span class="o">.</span><span class="n">split_train_other</span><span class="p">(</span><span class="n">split_type</span> <span class="o">=</span> <span class="s1">&#39;drug-blind&#39;</span><span class="p">,</span>
                                            <span class="n">ratio</span> <span class="o">=</span> <span class="p">[</span><span class="mi">8</span><span class="p">,</span><span class="mi">2</span><span class="p">],</span>
                                            <span class="n">random_state</span> <span class="o">=</span> <span class="mi">42</span><span class="p">,</span>
                                            <span class="p">)</span>
</pre></div>
</div>
</div>
<p>The split function dataframe can be used with any datatype however here we will use the experiments datatype. Now to view both the train and other splits use the call below, you will see that that the <code class="docutils literal notranslate"><span class="pre">train</span></code> set is ~80% of the original bladderpdo.experiments and the <code class="docutils literal notranslate"><span class="pre">other</span></code> set is ~20% aligning with the values input for <code class="docutils literal notranslate"><span class="pre">ratio</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">two_way_split</span><span class="o">.</span><span class="n">train</span><span class="o">.</span><span class="n">experiments</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>source</th>
      <th>improve_sample_id</th>
      <th>improve_drug_id</th>
      <th>study</th>
      <th>time</th>
      <th>time_unit</th>
      <th>dose_response_metric</th>
      <th>dose_response_value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Synapse</td>
      <td>5473</td>
      <td>SMI_20004</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.8937</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Synapse</td>
      <td>5473</td>
      <td>SMI_27249</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.8972</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Synapse</td>
      <td>5473</td>
      <td>SMI_28189</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.7877</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Synapse</td>
      <td>5473</td>
      <td>SMI_32023</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.7530</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Synapse</td>
      <td>5473</td>
      <td>SMI_34999</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.7171</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">two_way_split</span><span class="o">.</span><span class="n">other</span><span class="o">.</span><span class="n">experiments</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>source</th>
      <th>improve_sample_id</th>
      <th>improve_drug_id</th>
      <th>study</th>
      <th>time</th>
      <th>time_unit</th>
      <th>dose_response_metric</th>
      <th>dose_response_value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Synapse</td>
      <td>5473</td>
      <td>SMI_25616</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.6962</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Synapse</td>
      <td>5474</td>
      <td>SMI_25616</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.6962</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Synapse</td>
      <td>5475</td>
      <td>SMI_25616</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.6962</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Synapse</td>
      <td>5476</td>
      <td>SMI_25616</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.6962</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Synapse</td>
      <td>5477</td>
      <td>SMI_25616</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.6962</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>The three-way split uses similar input arguments as the two-way split. However, the three-way split provides a train, test, and validation set.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">split</span> <span class="o">=</span> <span class="n">bladderpdo</span><span class="o">.</span><span class="n">split_train_test_validate</span><span class="p">(</span><span class="n">split_type</span><span class="o">=</span><span class="s1">&#39;mixed-set&#39;</span><span class="p">,</span>
                                             <span class="n">ratio</span> <span class="o">=</span> <span class="p">[</span><span class="mi">8</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span>
                                             <span class="n">random_state</span> <span class="o">=</span> <span class="mi">42</span><span class="p">,</span>
                                             <span class="n">thresh</span> <span class="o">=</span> <span class="mf">0.8</span>
                                             <span class="p">)</span>
</pre></div>
</div>
</div>
<p>The CoderData package includes its own split function that is intended for only coderdata specific datasets. Here is an overview of what each split is and how it will be used.</p>
<p>Train (80%) : This gathers majority of the data and it allows the model to learn from the patterns and relationships of the data. This step involves iteratively adjusting its parameters to bettter fit the model.</p>
<p>Validate (10%): Gathering generally the smallest portion of data is used to refine the model’s hyperparameters. It evaluates the training performance by comparing it to validation performance based on overfitting or underfitting. Here it will follow each training iteration (epoch) and evaluating the performance based on the validation set.</p>
<p>Test (10%) : Finally to evaluate the final model performance the test data will be used to provide an estimate to how the model will perform on unseen data</p>
<p><strong>Note:</strong> The percentage of data pulled for each step may vary but should typically follows this size comparison (Train &gt; Test &gt;= Validate). For a more in depth explanation of how train, test, and validate work visit <a class="reference external" href="https://mlu-explain.github.io/train-test-validation/">mlu-explain.github.io</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">split</span><span class="o">.</span><span class="n">train</span><span class="o">.</span><span class="n">experiments</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>source</th>
      <th>improve_sample_id</th>
      <th>improve_drug_id</th>
      <th>study</th>
      <th>time</th>
      <th>time_unit</th>
      <th>dose_response_metric</th>
      <th>dose_response_value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Synapse</td>
      <td>5553</td>
      <td>SMI_55531</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.9275</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Synapse</td>
      <td>5556</td>
      <td>SMI_55473</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.2615</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Synapse</td>
      <td>5479</td>
      <td>SMI_55606</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.8320</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Synapse</td>
      <td>5566</td>
      <td>SMI_25708</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.7642</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Synapse</td>
      <td>5564</td>
      <td>SMI_16144</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.7298</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>2635</th>
      <td>Synapse</td>
      <td>5528</td>
      <td>SMI_27186</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.9594</td>
    </tr>
    <tr>
      <th>2636</th>
      <td>Synapse</td>
      <td>5529</td>
      <td>SMI_10282</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.8449</td>
    </tr>
    <tr>
      <th>2637</th>
      <td>Synapse</td>
      <td>5532</td>
      <td>SMI_25708</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.8323</td>
    </tr>
    <tr>
      <th>2638</th>
      <td>Synapse</td>
      <td>5523</td>
      <td>SMI_42845</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.6559</td>
    </tr>
    <tr>
      <th>2639</th>
      <td>Synapse</td>
      <td>5604</td>
      <td>SMI_35935</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.4328</td>
    </tr>
  </tbody>
</table>
<p>2640 rows × 8 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">split</span><span class="o">.</span><span class="n">test</span><span class="o">.</span><span class="n">experiments</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>source</th>
      <th>improve_sample_id</th>
      <th>improve_drug_id</th>
      <th>study</th>
      <th>time</th>
      <th>time_unit</th>
      <th>dose_response_metric</th>
      <th>dose_response_value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Synapse</td>
      <td>5522</td>
      <td>SMI_51528</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.5353</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Synapse</td>
      <td>5544</td>
      <td>SMI_20644</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>1.0000</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Synapse</td>
      <td>5563</td>
      <td>SMI_55531</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.9126</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Synapse</td>
      <td>5539</td>
      <td>SMI_20644</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.9315</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Synapse</td>
      <td>5490</td>
      <td>SMI_16144</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.7088</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>325</th>
      <td>Synapse</td>
      <td>5484</td>
      <td>SMI_8294</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.7684</td>
    </tr>
    <tr>
      <th>326</th>
      <td>Synapse</td>
      <td>5491</td>
      <td>SMI_18136</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.7549</td>
    </tr>
    <tr>
      <th>327</th>
      <td>Synapse</td>
      <td>5532</td>
      <td>SMI_20004</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.9606</td>
    </tr>
    <tr>
      <th>328</th>
      <td>Synapse</td>
      <td>5548</td>
      <td>SMI_659</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.7099</td>
    </tr>
    <tr>
      <th>329</th>
      <td>Synapse</td>
      <td>5490</td>
      <td>SMI_55531</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.9967</td>
    </tr>
  </tbody>
</table>
<p>330 rows × 8 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">split</span><span class="o">.</span><span class="n">validate</span><span class="o">.</span><span class="n">experiments</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>source</th>
      <th>improve_sample_id</th>
      <th>improve_drug_id</th>
      <th>study</th>
      <th>time</th>
      <th>time_unit</th>
      <th>dose_response_metric</th>
      <th>dose_response_value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Synapse</td>
      <td>5603</td>
      <td>SMI_42965</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.7433</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Synapse</td>
      <td>5555</td>
      <td>SMI_55473</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.2615</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Synapse</td>
      <td>5493</td>
      <td>SMI_40409</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.5723</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Synapse</td>
      <td>5553</td>
      <td>SMI_46771</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.2187</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Synapse</td>
      <td>5489</td>
      <td>SMI_27249</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.7267</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>325</th>
      <td>Synapse</td>
      <td>5541</td>
      <td>SMI_50331</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.5340</td>
    </tr>
    <tr>
      <th>326</th>
      <td>Synapse</td>
      <td>5537</td>
      <td>SMI_25616</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.8480</td>
    </tr>
    <tr>
      <th>327</th>
      <td>Synapse</td>
      <td>5530</td>
      <td>SMI_31885</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.8442</td>
    </tr>
    <tr>
      <th>328</th>
      <td>Synapse</td>
      <td>5542</td>
      <td>SMI_28189</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.5715</td>
    </tr>
    <tr>
      <th>329</th>
      <td>Synapse</td>
      <td>5603</td>
      <td>SMI_5060</td>
      <td>Lee etal 2018 Bladder PDOs</td>
      <td>6</td>
      <td>days</td>
      <td>auc</td>
      <td>0.7683</td>
    </tr>
  </tbody>
</table>
<p>330 rows × 8 columns</p>
</div></div>
</div>
<div class="line-block">
<div class="line">The coderdata function <code class="docutils literal notranslate"><span class="pre">save()</span></code> can be used to keep the split for later use.</div>
<div class="line"><strong>Note:</strong> The save function can be used with any coderdata Dataset object. However only individual splits can be saved not the full split.</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">split</span><span class="o">.</span><span class="n">train</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s1">&#39;/Path/coderdata/beataml_train.pickle&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<section id="Preparing-the-Data">
<h2>Preparing the Data<a class="headerlink" href="#Preparing-the-Data" title="Link to this heading">¶</a></h2>
<section id="Impute-Missing-Data">
<h3>Impute Missing Data<a class="headerlink" href="#Impute-Missing-Data" title="Link to this heading">¶</a></h3>
<div class="line-block">
<div class="line">Now we will prepare the transcriptomic and data structure columns to ensure it is formated correclty to be merged with our drug response values and then used within our model.</div>
<div class="line">First impute missing values in the bladderpdo transcriptomics data. Here we take the global mean, however, this may not be the best method for you or needed for your datasets.</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Impute and transcriptomics if there are any based on global mean.</span>
<span class="n">columns_to_fill</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">]</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">columns_to_fill</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">bladderpdo</span><span class="o">.</span><span class="n">transcriptomics</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="n">bladderpdo</span><span class="o">.</span><span class="n">transcriptomics</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)]:</span>
        <span class="n">bladderpdo</span><span class="o">.</span><span class="n">transcriptomics</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="n">bladderpdo</span><span class="o">.</span><span class="n">transcriptomics</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Scale-Input-Data">
<h3>Scale Input Data<a class="headerlink" href="#Scale-Input-Data" title="Link to this heading">¶</a></h3>
<p>(Optional) Here use a sklearn function to scale the transcriptomics data in a small range of non-negative numbers. This step can be helpful when there is a wide range of values in the data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scaler</span><span class="o">=</span><span class="n">MinMaxScaler</span><span class="p">()</span>
<span class="n">bladderpdo</span><span class="o">.</span><span class="n">transcriptomics</span><span class="p">[[</span><span class="s2">&quot;transcriptomics&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">scaler</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">bladderpdo</span><span class="o">.</span><span class="n">transcriptomics</span><span class="p">[[</span><span class="s2">&quot;transcriptomics&quot;</span><span class="p">]])</span>
</pre></div>
</div>
</div>
<p>Drop any duplicated transcriptomics data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bladderpdo</span><span class="o">.</span><span class="n">transcriptomics</span><span class="p">[[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>transcriptomics</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1.572852e-03</td>
    </tr>
    <tr>
      <th>1</th>
      <td>3.745602e-07</td>
    </tr>
    <tr>
      <th>2</th>
      <td>1.848217e-03</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4.544482e-04</td>
    </tr>
    <tr>
      <th>4</th>
      <td>6.768580e-04</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>From the <code class="docutils literal notranslate"><span class="pre">bladderpdo.drugs</span></code> dataframe we are pulling the canSMILES column and using rdkit to convert them into Morgan fingerprints. The resulting <code class="docutils literal notranslate"><span class="pre">fingerprint</span></code> column will then be used as another input along with transcriptomics for the model.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Convert SMILES to Morgan fingerprints</span>
<span class="k">def</span> <span class="nf">smiles_to_fingerprint</span><span class="p">(</span><span class="n">smiles</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">smiles</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">mol</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span> <span class="c1"># Handle cases where MolFromSmiles fails</span>
                <span class="k">return</span> <span class="kc">None</span>
        <span class="n">morgan_gen</span><span class="o">=</span> <span class="n">GetMorganGenerator</span><span class="p">(</span><span class="n">radius</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">fpSize</span><span class="o">=</span><span class="mi">1024</span><span class="p">)</span>
        <span class="n">fingerprint</span> <span class="o">=</span> <span class="n">morgan_gen</span><span class="o">.</span><span class="n">GetFingerprint</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>  <span class="c1"># Adjust radius and nBits as needed</span>
        <span class="n">fingerprint_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">fingerprint</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">fingerprint_array</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error processing SMILES &#39;</span><span class="si">{</span><span class="n">smiles</span><span class="si">}</span><span class="s2">&#39;: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>


<span class="c1"># smiles_to_fingerprint(merged_df.canSMILES[0])</span>
<span class="c1"># # Apply the function to your SMILES column</span>
<span class="n">bladderpdo</span><span class="o">.</span><span class="n">drugs</span><span class="p">[</span><span class="s1">&#39;fingerprint&#39;</span><span class="p">]</span><span class="o">=</span> <span class="n">bladderpdo</span><span class="o">.</span><span class="n">drugs</span><span class="p">[</span><span class="s1">&#39;canSMILES&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">smiles_to_fingerprint</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
</section>
<section id="Merge-Data">
<h3>Merge Data<a class="headerlink" href="#Merge-Data" title="Link to this heading">¶</a></h3>
<p>Group transcriptomics data into an array based on the <code class="docutils literal notranslate"><span class="pre">improve_sample_id</span></code>. This is in prepration for the model because the input values must be in an array. Then drop any duplicated transriptomics values with the same array.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">group_df</span> <span class="o">=</span> <span class="n">bladderpdo</span><span class="o">.</span><span class="n">transcriptomics</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;improve_sample_id&#39;</span><span class="p">)[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="n">group_df</span><span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">group_df</span><span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">group_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>improve_sample_id</th>
      <th>transcriptomics</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5473</td>
      <td>[0.0004120739513098224, 0.005498299890739134, ...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>5477</td>
      <td>[0.03584645254124297, 0.0006806983197478193, 0...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>5482</td>
      <td>[0.0010883074665564165, 0.004632827848666922, ...</td>
    </tr>
    <tr>
      <th>3</th>
      <td>5485</td>
      <td>[0.0014773197417206076, 0.0038554997961206556,...</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5488</td>
      <td>[0.000516024688106854, 0.0020666180961583826, ...</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>We will merge the split dataframes with the <code class="docutils literal notranslate"><span class="pre">bladderpdo.drugs</span></code> dataframe based on the <code class="docutils literal notranslate"><span class="pre">improve_drug_id</span></code> first and then in the next step with will merge the dataframes with the <code class="docutils literal notranslate"><span class="pre">transcriptomics</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">split</span><span class="o">.</span><span class="n">train</span><span class="o">.</span><span class="n">experiments</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">split</span><span class="o">.</span><span class="n">train</span><span class="o">.</span><span class="n">experiments</span><span class="p">[[</span><span class="s1">&#39;improve_sample_id&#39;</span><span class="p">,</span><span class="s1">&#39;improve_drug_id&#39;</span><span class="p">,</span> <span class="s1">&#39;dose_response_value&#39;</span><span class="p">]],</span>
                                   <span class="n">bladderpdo</span><span class="o">.</span><span class="n">drugs</span><span class="p">[[</span><span class="s1">&#39;improve_drug_id&#39;</span><span class="p">,</span><span class="s1">&#39;fingerprint&#39;</span><span class="p">]],</span>
                                   <span class="n">on</span><span class="o">=</span><span class="s1">&#39;improve_drug_id&#39;</span><span class="p">,</span>
                                   <span class="n">how</span><span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>

<span class="n">split</span><span class="o">.</span><span class="n">test</span><span class="o">.</span><span class="n">experiments</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">split</span><span class="o">.</span><span class="n">test</span><span class="o">.</span><span class="n">experiments</span><span class="p">[[</span><span class="s1">&#39;improve_sample_id&#39;</span><span class="p">,</span><span class="s1">&#39;improve_drug_id&#39;</span><span class="p">,</span> <span class="s1">&#39;dose_response_value&#39;</span><span class="p">]],</span>
                                   <span class="n">bladderpdo</span><span class="o">.</span><span class="n">drugs</span><span class="p">[[</span><span class="s1">&#39;improve_drug_id&#39;</span><span class="p">,</span><span class="s1">&#39;fingerprint&#39;</span><span class="p">]],</span>
                                   <span class="n">on</span><span class="o">=</span><span class="s1">&#39;improve_drug_id&#39;</span><span class="p">,</span>
                                   <span class="n">how</span><span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>

<span class="n">split</span><span class="o">.</span><span class="n">validate</span><span class="o">.</span><span class="n">experiments</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">split</span><span class="o">.</span><span class="n">validate</span><span class="o">.</span><span class="n">experiments</span><span class="p">[[</span><span class="s1">&#39;improve_sample_id&#39;</span><span class="p">,</span><span class="s1">&#39;improve_drug_id&#39;</span><span class="p">,</span> <span class="s1">&#39;dose_response_value&#39;</span><span class="p">]],</span>
                                   <span class="n">bladderpdo</span><span class="o">.</span><span class="n">drugs</span><span class="p">[[</span><span class="s1">&#39;improve_drug_id&#39;</span><span class="p">,</span><span class="s1">&#39;fingerprint&#39;</span><span class="p">]],</span>
                                   <span class="n">on</span><span class="o">=</span><span class="s1">&#39;improve_drug_id&#39;</span><span class="p">,</span>
                                   <span class="n">how</span><span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>To the split <code class="docutils literal notranslate"><span class="pre">bladderpdo.experiments</span></code> dataframes add <code class="docutils literal notranslate"><span class="pre">transcriptomics</span></code> arrays based on <code class="docutils literal notranslate"><span class="pre">improve_sample_id</span></code>. Then drop all columns besides our input values, <code class="docutils literal notranslate"><span class="pre">transcriptomics</span></code> and <code class="docutils literal notranslate"><span class="pre">fingerprint</span></code>, with the corresponding outputs <code class="docutils literal notranslate"><span class="pre">dose_response_value</span></code>. This is repeated for the train, test, and validate dataframes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">train_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">split</span><span class="o">.</span><span class="n">train</span><span class="o">.</span><span class="n">experiments</span><span class="p">,</span> <span class="n">group_df</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;improve_sample_id&#39;</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>
<span class="n">train_df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">train_df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;dose_response_value&#39;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">train_df</span> <span class="o">=</span> <span class="n">train_df</span><span class="p">[[</span><span class="s1">&#39;dose_response_value&#39;</span><span class="p">,</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">,</span> <span class="s1">&#39;fingerprint&#39;</span> <span class="p">]]</span>

<span class="n">test_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">split</span><span class="o">.</span><span class="n">test</span><span class="o">.</span><span class="n">experiments</span><span class="p">,</span> <span class="n">group_df</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;improve_sample_id&#39;</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>
<span class="n">test_df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">test_df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;dose_response_value&#39;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">test_df</span> <span class="o">=</span> <span class="n">test_df</span><span class="p">[[</span><span class="s1">&#39;dose_response_value&#39;</span><span class="p">,</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">,</span><span class="s1">&#39;fingerprint&#39;</span><span class="p">]]</span>

<span class="n">validate_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">split</span><span class="o">.</span><span class="n">validate</span><span class="o">.</span><span class="n">experiments</span><span class="p">,</span> <span class="n">group_df</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;improve_sample_id&#39;</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>
<span class="n">validate_df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">validate_df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;dose_response_value&#39;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">validate_df</span> <span class="o">=</span> <span class="n">validate_df</span><span class="p">[[</span><span class="s1">&#39;dose_response_value&#39;</span><span class="p">,</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">,</span><span class="s1">&#39;fingerprint&#39;</span><span class="p">]]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">train_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dose_response_value</th>
      <th>transcriptomics</th>
      <th>fingerprint</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.9275</td>
      <td>[0.0031862060625194033, 2.728004558250126e-05,...</td>
      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, ...</td>
    </tr>
    <tr>
      <th>107</th>
      <td>0.2615</td>
      <td>[0.026989577082694893, 0.002483812226380167, 0...</td>
      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>
    </tr>
    <tr>
      <th>870</th>
      <td>0.5044</td>
      <td>[0.0031862060625194033, 2.728004558250126e-05,...</td>
      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>
    </tr>
    <tr>
      <th>1055</th>
      <td>0.7376</td>
      <td>[0.044913496837260425, 0.0005729141409851485, ...</td>
      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>
    </tr>
    <tr>
      <th>1168</th>
      <td>0.9244</td>
      <td>[0.0, 0.0013422786305636884, 0.006175404786558...</td>
      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">test_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dose_response_value</th>
      <th>transcriptomics</th>
      <th>fingerprint</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>308</th>
      <td>0.9126</td>
      <td>[0.0007689920202373915, 0.002660806480477618, ...</td>
      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, ...</td>
    </tr>
    <tr>
      <th>836</th>
      <td>0.8929</td>
      <td>[0.008620298094601964, 0.003253032313722315, 0...</td>
      <td>[0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>
    </tr>
    <tr>
      <th>1334</th>
      <td>0.7684</td>
      <td>[0.0010883074665564165, 0.004632827848666922, ...</td>
      <td>[0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>
    </tr>
    <tr>
      <th>1500</th>
      <td>0.4067</td>
      <td>[0.0031862060625194033, 2.728004558250126e-05,...</td>
      <td>[0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>
    </tr>
    <tr>
      <th>2137</th>
      <td>0.2959</td>
      <td>[0.026989577082694893, 0.002483812226380167, 0...</td>
      <td>[0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">validate_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dose_response_value</th>
      <th>transcriptomics</th>
      <th>fingerprint</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>171</th>
      <td>0.5723</td>
      <td>[0.0007678460342894769, 0.0014434679191595028,...</td>
      <td>[0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, ...</td>
    </tr>
    <tr>
      <th>254</th>
      <td>0.2187</td>
      <td>[0.0031862060625194033, 2.728004558250126e-05,...</td>
      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>
    </tr>
    <tr>
      <th>436</th>
      <td>0.7674</td>
      <td>[0.008620298094601964, 0.003253032313722315, 0...</td>
      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>
    </tr>
    <tr>
      <th>506</th>
      <td>0.6207</td>
      <td>[0.002719332067378844, 0.0028423386644789166, ...</td>
      <td>[0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, ...</td>
    </tr>
    <tr>
      <th>589</th>
      <td>0.7549</td>
      <td>[0.000516024688106854, 0.0020666180961583826, ...</td>
      <td>[0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p><strong>Quick Tip:</strong> View the length of the transcriptomics array to then later decide the shape that is input into the model. Here we have multiple size arrays so we will need to pad or truncate to ensure a consist shape. An example of that will be shown later in the tutorial.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;Array Length&quot;</span><span class="p">:</span><span class="n">validate_df</span><span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">len</span><span class="p">)</span><span class="o">.</span><span class="n">unique</span><span class="p">()})</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Array Length</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>3996</td>
    </tr>
    <tr>
      <th>1</th>
      <td>4741</td>
    </tr>
    <tr>
      <th>2</th>
      <td>4040</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4766</td>
    </tr>
    <tr>
      <th>4</th>
      <td>3964</td>
    </tr>
    <tr>
      <th>5</th>
      <td>195</td>
    </tr>
    <tr>
      <th>6</th>
      <td>4207</td>
    </tr>
    <tr>
      <th>7</th>
      <td>4067</td>
    </tr>
    <tr>
      <th>8</th>
      <td>3723</td>
    </tr>
    <tr>
      <th>9</th>
      <td>4184</td>
    </tr>
    <tr>
      <th>10</th>
      <td>3698</td>
    </tr>
    <tr>
      <th>11</th>
      <td>4083</td>
    </tr>
    <tr>
      <th>12</th>
      <td>4236</td>
    </tr>
    <tr>
      <th>13</th>
      <td>2985</td>
    </tr>
    <tr>
      <th>14</th>
      <td>4193</td>
    </tr>
    <tr>
      <th>15</th>
      <td>4845</td>
    </tr>
    <tr>
      <th>16</th>
      <td>4491</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
</section>
<section id="Set-the-Model-Parameters">
<h2>Set the Model Parameters<a class="headerlink" href="#Set-the-Model-Parameters" title="Link to this heading">¶</a></h2>
<p>Here we set the model parameters. You may modify the inputs, layers, activation functions, and outputs of the model. This is a general setup when using a keras model, but this may need to be adjusted to better suit your data. Here we use <code class="docutils literal notranslate"><span class="pre">activation</span> <span class="pre">=</span> <span class="pre">'relu'</span></code> for the output layer, note that this activation is only intended when dealing with non-negative values (how the transcriptomic inputs were scaled previously).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Define Inputs</span>
<span class="n">transcriptomics_input</span> <span class="o">=</span> <span class="n">keras</span><span class="o">.</span><span class="n">Input</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="p">(</span><span class="mi">3000</span><span class="p">,),</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;transcriptomics&quot;</span><span class="p">)</span>
<span class="n">fingerprint_input</span> <span class="o">=</span> <span class="n">keras</span><span class="o">.</span><span class="n">Input</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="p">(</span><span class="mi">1024</span><span class="p">,),</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;fingerprint&quot;</span><span class="p">)</span>

<span class="c1"># You may adjust these layers and activation functions to get better (or worse) results.</span>
<span class="c1"># Merge all available features into a single large vector via concatenation</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">layers</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">transcriptomics_input</span><span class="p">,</span> <span class="n">fingerprint_input</span><span class="p">])</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">layers</span><span class="o">.</span><span class="n">Dense</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="n">activation</span><span class="o">=</span><span class="s1">&#39;relu&#39;</span><span class="p">)(</span><span class="n">x</span><span class="p">)</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">layers</span><span class="o">.</span><span class="n">Dense</span><span class="p">(</span><span class="mi">64</span><span class="p">,</span> <span class="n">activation</span><span class="o">=</span><span class="s2">&quot;relu&quot;</span><span class="p">)(</span><span class="n">x</span><span class="p">)</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">layers</span><span class="o">.</span><span class="n">Dropout</span><span class="p">(</span><span class="mf">0.5</span><span class="p">)(</span><span class="n">x</span><span class="p">)</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">layers</span><span class="o">.</span><span class="n">Dense</span><span class="p">(</span><span class="mi">32</span><span class="p">,</span> <span class="n">activation</span><span class="o">=</span><span class="s1">&#39;relu&#39;</span><span class="p">)(</span><span class="n">x</span><span class="p">)</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">layers</span><span class="o">.</span><span class="n">Dense</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="n">activation</span><span class="o">=</span><span class="s1">&#39;relu&#39;</span><span class="p">)(</span><span class="n">x</span><span class="p">)</span>


<span class="c1"># Priority prediction is here. You may add layers (and an output) after this too.</span>
<span class="c1"># This just allows for an early prediction in case you have a large datset and want preliminary results.</span>
<span class="n">priority_pred</span> <span class="o">=</span> <span class="n">layers</span><span class="o">.</span><span class="n">Dense</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;priority&quot;</span><span class="p">,</span> <span class="n">activation</span> <span class="o">=</span><span class="s1">&#39;relu&#39;</span><span class="p">)(</span><span class="n">x</span><span class="p">)</span>  <span class="c1"># Using the activation relu.</span>

<span class="c1"># Instantiate an end-to-end model predicting both priority and department</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">keras</span><span class="o">.</span><span class="n">Model</span><span class="p">(</span>
    <span class="n">inputs</span><span class="o">=</span> <span class="p">[</span><span class="n">transcriptomics_input</span><span class="p">,</span> <span class="n">fingerprint_input</span><span class="p">],</span>
    <span class="n">outputs</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;priority&quot;</span><span class="p">:</span> <span class="n">priority_pred</span><span class="p">},</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Compile-the-Model">
<h2>Compile the Model<a class="headerlink" href="#Compile-the-Model" title="Link to this heading">¶</a></h2>
<p>Here we used a basic keras <code class="docutils literal notranslate"><span class="pre">compile()</span></code> input with the Adam optimizer. For other functions options visit <a class="reference external" href="https://keras.io/api/models/model_training_apis/#compile-method">keras.io/compile-method</a> or optimizer options <a class="reference external" href="https://keras.io/api/optimizers/">keras.io/optimizers</a>. Then view the model summary, this can be helpful when adjusting parameters and seeing the model layers. For more information on the model summary visit
<a class="reference external" href="https://keras.io/api/models/model_training_apis/#compile-method">keras.io/summary-method</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
    <span class="n">optimizer</span><span class="o">=</span><span class="n">keras</span><span class="o">.</span><span class="n">optimizers</span><span class="o">.</span><span class="n">Adam</span><span class="p">(</span><span class="n">learning_rate</span><span class="o">=</span><span class="mf">0.0001</span><span class="p">),</span>
    <span class="n">loss</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;priority&quot;</span><span class="p">:</span> <span class="n">keras</span><span class="o">.</span><span class="n">losses</span><span class="o">.</span><span class="n">MeanSquaredError</span><span class="p">(),</span>
    <span class="p">},</span>
    <span class="n">metrics</span><span class="o">=</span><span class="p">[</span><span class="n">MeanAbsoluteError</span><span class="p">()],</span>
    <span class="n">loss_weights</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;priority&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
<span class="p">)</span>

<span class="n">model</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-weight: bold">Model: "functional"</span>
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold"> Layer (type)        </span>┃<span style="font-weight: bold"> Output Shape      </span>┃<span style="font-weight: bold">    Param # </span>┃<span style="font-weight: bold"> Connected to      </span>┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ transcriptomics     │ (<span style="color: #00d7ff; text-decoration-color: #00d7ff">None</span>, <span style="color: #00af00; text-decoration-color: #00af00">3000</span>)      │          <span style="color: #00af00; text-decoration-color: #00af00">0</span> │ -                 │
│ (<span style="color: #0087ff; text-decoration-color: #0087ff">InputLayer</span>)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ fingerprint         │ (<span style="color: #00d7ff; text-decoration-color: #00d7ff">None</span>, <span style="color: #00af00; text-decoration-color: #00af00">1024</span>)      │          <span style="color: #00af00; text-decoration-color: #00af00">0</span> │ -                 │
│ (<span style="color: #0087ff; text-decoration-color: #0087ff">InputLayer</span>)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ concatenate         │ (<span style="color: #00d7ff; text-decoration-color: #00d7ff">None</span>, <span style="color: #00af00; text-decoration-color: #00af00">4024</span>)      │          <span style="color: #00af00; text-decoration-color: #00af00">0</span> │ transcriptomics[<span style="color: #00af00; text-decoration-color: #00af00">…</span> │
│ (<span style="color: #0087ff; text-decoration-color: #0087ff">Concatenate</span>)       │                   │            │ fingerprint[<span style="color: #00af00; text-decoration-color: #00af00">0</span>][<span style="color: #00af00; text-decoration-color: #00af00">0</span>] │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense (<span style="color: #0087ff; text-decoration-color: #0087ff">Dense</span>)       │ (<span style="color: #00d7ff; text-decoration-color: #00d7ff">None</span>, <span style="color: #00af00; text-decoration-color: #00af00">16</span>)        │     <span style="color: #00af00; text-decoration-color: #00af00">64,400</span> │ concatenate[<span style="color: #00af00; text-decoration-color: #00af00">0</span>][<span style="color: #00af00; text-decoration-color: #00af00">0</span>] │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_1 (<span style="color: #0087ff; text-decoration-color: #0087ff">Dense</span>)     │ (<span style="color: #00d7ff; text-decoration-color: #00d7ff">None</span>, <span style="color: #00af00; text-decoration-color: #00af00">64</span>)        │      <span style="color: #00af00; text-decoration-color: #00af00">1,088</span> │ dense[<span style="color: #00af00; text-decoration-color: #00af00">0</span>][<span style="color: #00af00; text-decoration-color: #00af00">0</span>]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout (<span style="color: #0087ff; text-decoration-color: #0087ff">Dropout</span>)   │ (<span style="color: #00d7ff; text-decoration-color: #00d7ff">None</span>, <span style="color: #00af00; text-decoration-color: #00af00">64</span>)        │          <span style="color: #00af00; text-decoration-color: #00af00">0</span> │ dense_1[<span style="color: #00af00; text-decoration-color: #00af00">0</span>][<span style="color: #00af00; text-decoration-color: #00af00">0</span>]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_2 (<span style="color: #0087ff; text-decoration-color: #0087ff">Dense</span>)     │ (<span style="color: #00d7ff; text-decoration-color: #00d7ff">None</span>, <span style="color: #00af00; text-decoration-color: #00af00">32</span>)        │      <span style="color: #00af00; text-decoration-color: #00af00">2,080</span> │ dropout[<span style="color: #00af00; text-decoration-color: #00af00">0</span>][<span style="color: #00af00; text-decoration-color: #00af00">0</span>]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_3 (<span style="color: #0087ff; text-decoration-color: #0087ff">Dense</span>)     │ (<span style="color: #00d7ff; text-decoration-color: #00d7ff">None</span>, <span style="color: #00af00; text-decoration-color: #00af00">12</span>)        │        <span style="color: #00af00; text-decoration-color: #00af00">396</span> │ dense_2[<span style="color: #00af00; text-decoration-color: #00af00">0</span>][<span style="color: #00af00; text-decoration-color: #00af00">0</span>]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ priority (<span style="color: #0087ff; text-decoration-color: #0087ff">Dense</span>)    │ (<span style="color: #00d7ff; text-decoration-color: #00d7ff">None</span>, <span style="color: #00af00; text-decoration-color: #00af00">1</span>)         │         <span style="color: #00af00; text-decoration-color: #00af00">13</span> │ dense_3[<span style="color: #00af00; text-decoration-color: #00af00">0</span>][<span style="color: #00af00; text-decoration-color: #00af00">0</span>]     │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-weight: bold"> Total params: </span><span style="color: #00af00; text-decoration-color: #00af00">67,977</span> (265.54 KB)
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-weight: bold"> Trainable params: </span><span style="color: #00af00; text-decoration-color: #00af00">67,977</span> (265.54 KB)
</pre></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-weight: bold"> Non-trainable params: </span><span style="color: #00af00; text-decoration-color: #00af00">0</span> (0.00 B)
</pre></div>
</div>
</section>
<section id="Run-your-Model">
<h2>Run your Model<a class="headerlink" href="#Run-your-Model" title="Link to this heading">¶</a></h2>
<p>Before you can run the mdoel you may need to pad/truncate the input arrays. As mentioned earlier this is because the arrays must be the same length in order for the model to run correctly. Here we adjust both the <code class="docutils literal notranslate"><span class="pre">transcriptomics</span></code> and <code class="docutils literal notranslate"><span class="pre">fingerprint</span></code> inputs for all split dataframes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Set target_length as the maximum length</span>
<span class="k">def</span> <span class="nf">pad_sequence</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">target_length</span><span class="p">,</span> <span class="n">pad_value</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Pad or truncate a sequence to a target length.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">target_length</span><span class="p">:</span>  <span class="c1"># Truncate if longer than target_length</span>
        <span class="k">return</span> <span class="n">seq</span><span class="p">[:</span><span class="n">target_length</span><span class="p">]</span>
    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">target_length</span><span class="p">:</span>  <span class="c1"># Pad with zeros if shorter than target_length</span>
        <span class="k">return</span> <span class="n">seq</span> <span class="o">+</span> <span class="p">[</span><span class="n">pad_value</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">target_length</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">seq</span>

<span class="c1"># Example: Create fixed-length arrays</span>
<span class="n">train_df</span><span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">train_df</span><span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pad_sequence</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">target_length</span><span class="o">=</span><span class="mi">3000</span><span class="p">))</span>
<span class="n">test_df</span><span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">test_df</span><span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pad_sequence</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">target_length</span><span class="o">=</span><span class="mi">3000</span><span class="p">))</span>
<span class="n">validate_df</span><span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">validate_df</span><span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pad_sequence</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">target_length</span><span class="o">=</span><span class="mi">3000</span><span class="p">))</span>

<span class="n">train_df</span><span class="p">[</span><span class="s1">&#39;fingerprint&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">train_df</span><span class="p">[</span><span class="s1">&#39;fingerprint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pad_sequence</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">target_length</span><span class="o">=</span><span class="mi">1024</span><span class="p">))</span>
<span class="n">test_df</span><span class="p">[</span><span class="s1">&#39;fingerprint&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">test_df</span><span class="p">[</span><span class="s1">&#39;fingerprint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pad_sequence</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">target_length</span><span class="o">=</span><span class="mi">1024</span><span class="p">))</span>
<span class="n">validate_df</span><span class="p">[</span><span class="s1">&#39;fingerprint&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">validate_df</span><span class="p">[</span><span class="s1">&#39;fingerprint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pad_sequence</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">target_length</span><span class="o">=</span><span class="mi">1024</span><span class="p">))</span>
<br/></pre></div>
</div>
</div>
<p>Define the output data, <code class="docutils literal notranslate"><span class="pre">dose_response_value</span></code>, in prepartion of running the model.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">y_train</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">train_df</span><span class="p">[</span><span class="s1">&#39;dose_response_value&#39;</span><span class="p">])</span>
<span class="n">y_test</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">test_df</span><span class="p">[</span><span class="s1">&#39;dose_response_value&#39;</span><span class="p">])</span>
<span class="n">y_validate</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">validate_df</span><span class="p">[</span><span class="s1">&#39;dose_response_value&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<p>Now use the keras <code class="docutils literal notranslate"><span class="pre">fit()</span></code> function with the train and validation data. For more information on this function visit <a class="reference external" href="https://keras.io/api/models/model_training_apis/#fit-method">kera.io/fit-method</a>. Note you may adjust the <code class="docutils literal notranslate"><span class="pre">epoch</span></code> and <code class="docutils literal notranslate"><span class="pre">bath_size</span></code> to better fit your model.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Train the model</span>
<span class="n">history</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span>
    <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">train_df</span><span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()),</span>
     <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">train_df</span><span class="p">[</span><span class="s1">&#39;fingerprint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
     <span class="p">],</span>
    <span class="n">y_train</span><span class="p">,</span>
    <span class="n">validation_data</span><span class="o">=</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">validate_df</span><span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()),</span>
                      <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">validate_df</span><span class="p">[</span><span class="s1">&#39;fingerprint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
                      <span class="p">],</span>
                     <span class="n">y_validate</span><span class="p">),</span>
    <span class="n">epochs</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>  <span class="c1"># Number of epochs</span>
    <span class="n">batch_size</span><span class="o">=</span><span class="mi">32</span><span class="p">,</span>  <span class="c1"># Batch size</span>
    <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span> <span class="c1"># Reduces the output (set equal to 1 if you want to see the output steps)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Evaluate-your-Model">
<h2>Evaluate your Model<a class="headerlink" href="#Evaluate-your-Model" title="Link to this heading">¶</a></h2>
<p>Now we will use the test data to evaluate our model. Information on the <code class="docutils literal notranslate"><span class="pre">evaluate()</span></code> function, <a class="reference external" href="https://keras.io/api/models/model_training_apis/#evaluate-method">kera.io/evaluate-method</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">losses</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">evaluate</span><span class="p">(</span>
    <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">test_df</span><span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()),</span>
     <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">test_df</span><span class="p">[</span><span class="s1">&#39;fingerprint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">())],</span>
    <span class="n">y_test</span><span class="p">,</span>
    <span class="n">return_dict</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">losses</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-bold">3/3</span> <span class="ansi-green-fg">━━━━━━━━━━━━━━━━━━━━</span> <span class="ansi-bold">0s</span> 17ms/step - loss: 0.0197 - mean_absolute_error: 0.1211
{&#39;loss&#39;: 0.01972586289048195, &#39;mean_absolute_error&#39;: 0.12107016891241074}
</pre></div></div>
</div>
<p>Here is a site giving a guide on how to evaluate your model, visit <a class="reference external" href="https://machinelearningmastery.com/learning-curves-for-diagnosing-machine-learning-model-performance/">machinelearning.com</a>. The site gives examples of similary graphs shown below demonstrating what overfitting, underfitting, and correct models may output.</p>
<section id="Plot-Loss-Function-and-Error-of-the-Model">
<h3>Plot Loss Function and Error of the Model<a class="headerlink" href="#Plot-Loss-Function-and-Error-of-the-Model" title="Link to this heading">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># summarize history for accuracy</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">history</span><span class="o">.</span><span class="n">history</span><span class="p">[</span><span class="s1">&#39;val_mean_absolute_error&#39;</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">history</span><span class="o">.</span><span class="n">history</span><span class="p">[</span><span class="s1">&#39;mean_absolute_error&#39;</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Mean Absolute Error of Deep Learning Model&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Mean Absolute Error&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;epoch&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">([</span><span class="s1">&#39;Validation Set&#39;</span><span class="p">,</span><span class="s1">&#39;Training Set&#39;</span><span class="p">,],</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper right&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="c1"># summarize history for loss</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">history</span><span class="o">.</span><span class="n">history</span><span class="p">[</span><span class="s1">&#39;val_loss&#39;</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">history</span><span class="o">.</span><span class="n">history</span><span class="p">[</span><span class="s1">&#39;loss&#39;</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Loss Function of Deep Learning Model&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Loss&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;epoch&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">([</span><span class="s1">&#39;Validation Set&#39;</span><span class="p">,</span><span class="s1">&#39;Training Set&#39;</span><span class="p">],</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper right&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/_notebooks_docsite_tutorial_83_0.png" src="../_images/_notebooks_docsite_tutorial_83_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/_notebooks_docsite_tutorial_83_1.png" src="../_images/_notebooks_docsite_tutorial_83_1.png" />
</div>
</div>
</section>
<section id="Prediction-vs-True-Values">
<h3>Prediction vs True Values<a class="headerlink" href="#Prediction-vs-True-Values" title="Link to this heading">¶</a></h3>
<p>We will now use the test data to see how our model did in making predictions based off of the train and validation data it was given. Using graph and summary statistics we can see how well the model was able to reflect the given data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># make a prediction</span>
<span class="n">predict</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">test_df</span><span class="p">[</span><span class="s1">&#39;transcriptomics&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()),</span>
     <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">test_df</span><span class="p">[</span><span class="s1">&#39;fingerprint&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">())])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-bold">3/3</span> <span class="ansi-green-fg">━━━━━━━━━━━━━━━━━━━━</span> <span class="ansi-bold">0s</span> 31ms/step
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">new_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
    <span class="s1">&#39;Predicted Values&#39;</span><span class="p">:</span> <span class="n">predict</span><span class="p">[</span><span class="s2">&quot;priority&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
    <span class="s1">&#39;True Values&#39;</span><span class="p">:</span> <span class="n">y_test</span>
<span class="p">})</span>
<span class="n">new_df</span><span class="p">[</span><span class="s1">&#39;Predicted Values&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_df</span><span class="p">[</span><span class="s1">&#39;Predicted Values&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="k">else</span> <span class="n">x</span><span class="p">)</span>
<span class="n">sorted_df</span> <span class="o">=</span> <span class="n">new_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;True Values&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  <span class="c1"># Change &#39;ascending&#39; to False for descending order</span>
<span class="n">sorted_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Predicted Values</th>
      <th>True Values</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>4</th>
      <td>0.432947</td>
      <td>0.2959</td>
    </tr>
    <tr>
      <th>13</th>
      <td>0.340147</td>
      <td>0.3021</td>
    </tr>
    <tr>
      <th>15</th>
      <td>0.548971</td>
      <td>0.3230</td>
    </tr>
    <tr>
      <th>74</th>
      <td>0.497476</td>
      <td>0.3567</td>
    </tr>
    <tr>
      <th>34</th>
      <td>0.371287</td>
      <td>0.3596</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">true_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sorted_df</span><span class="p">[</span><span class="s2">&quot;True Values&quot;</span><span class="p">])</span>
<span class="n">predicted_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sorted_df</span><span class="p">[</span><span class="s2">&quot;Predicted Values&quot;</span><span class="p">])</span>

<span class="c1"># Plot the true values and predicted values</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">true_values</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">predicted_values</span><span class="p">)</span>

<span class="c1"># Fit trendlines</span>
<span class="n">true_values_trendline</span> <span class="o">=</span> <span class="n">LinearRegression</span><span class="p">()</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">true_values</span><span class="p">))</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">true_values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">true_values</span><span class="p">))</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
<span class="n">predicted_values_trendline</span> <span class="o">=</span> <span class="n">LinearRegression</span><span class="p">()</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">predicted_values</span><span class="p">))</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">predicted_values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">predicted_values</span><span class="p">))</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>

<span class="c1"># Plot trendlines</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">true_values_trendline</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">predicted_values_trendline</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;orange&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>

<span class="c1"># Customize plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Predicted AUC vs True AUC&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Drug Response AUC&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Observation Values&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">([</span><span class="s1">&#39;True Values&#39;</span><span class="p">,</span> <span class="s1">&#39;Predicted Values&#39;</span><span class="p">,</span> <span class="s1">&#39;True Values Trendline&#39;</span><span class="p">,</span> <span class="s1">&#39;Predicted Values Trendline&#39;</span><span class="p">],</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;lower right&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/_notebooks_docsite_tutorial_88_0.png" src="../_images/_notebooks_docsite_tutorial_88_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">true_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">new_df</span><span class="p">[</span><span class="s2">&quot;True Values&quot;</span><span class="p">])</span>
<span class="n">predicted_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">new_df</span><span class="p">[</span><span class="s2">&quot;Predicted Values&quot;</span><span class="p">])</span>

<span class="c1"># Plot the true values and predicted values</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">true_values</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">predicted_values</span><span class="p">)</span>

<span class="c1"># Fit trendlines</span>
<span class="n">true_values_trendline</span> <span class="o">=</span> <span class="n">LinearRegression</span><span class="p">()</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">true_values</span><span class="p">))</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">true_values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">true_values</span><span class="p">))</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
<span class="n">predicted_values_trendline</span> <span class="o">=</span> <span class="n">LinearRegression</span><span class="p">()</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">predicted_values</span><span class="p">))</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">predicted_values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">predicted_values</span><span class="p">))</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>


<span class="c1"># Customize plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Predicted AUC vs True AUC&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Drug Response AUC&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Observation Values&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">([</span><span class="s1">&#39;True Values&#39;</span><span class="p">,</span> <span class="s1">&#39;Predicted Values&#39;</span><span class="p">,</span> <span class="s1">&#39;True Values Trendline&#39;</span><span class="p">,</span> <span class="s1">&#39;Predicted Values Trendline&#39;</span><span class="p">],</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;lower right&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/_notebooks_docsite_tutorial_89_0.png" src="../_images/_notebooks_docsite_tutorial_89_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<span class="c1"># Visualizing the regression results</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">y_test</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">predicted_values</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="p">[</span><span class="n">y_test</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">y_test</span><span class="o">.</span><span class="n">max</span><span class="p">()],</span> <span class="n">y</span><span class="o">=</span><span class="p">[</span><span class="n">y_test</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">y_test</span><span class="o">.</span><span class="n">max</span><span class="p">()],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;orange&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;True vs Predicted AUC Values&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;True AUC Values&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Predicted AUC Values&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/_notebooks_docsite_tutorial_90_0.png" src="../_images/_notebooks_docsite_tutorial_90_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate Summary Statistics</span>


<span class="c1"># Calculate Mean Absolute Error (MAE)</span>
<span class="n">mae</span> <span class="o">=</span> <span class="n">mean_absolute_error</span><span class="p">(</span><span class="n">new_df</span><span class="p">[</span><span class="s1">&#39;True Values&#39;</span><span class="p">],</span> <span class="n">new_df</span><span class="p">[</span><span class="s1">&#39;Predicted Values&#39;</span><span class="p">])</span>

<span class="c1"># Calculate Root Mean Squared Error (RMSE)</span>
<span class="n">rmse</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">mean_squared_error</span><span class="p">(</span><span class="n">new_df</span><span class="p">[</span><span class="s1">&#39;True Values&#39;</span><span class="p">],</span> <span class="n">new_df</span><span class="p">[</span><span class="s1">&#39;Predicted Values&#39;</span><span class="p">]))</span>

<span class="c1"># Calculate R-squared (R2) score</span>
<span class="n">r2</span> <span class="o">=</span> <span class="n">r2_score</span><span class="p">(</span><span class="n">new_df</span><span class="p">[</span><span class="s1">&#39;True Values&#39;</span><span class="p">],</span> <span class="n">new_df</span><span class="p">[</span><span class="s1">&#39;Predicted Values&#39;</span><span class="p">])</span>

<span class="n">summary_statistics</span> <span class="o">=</span> <span class="n">new_df</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>

<span class="c1"># Print the statistics</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Mean Absolute Error (MAE):&quot;</span><span class="p">,</span> <span class="n">mae</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Root Mean Squared Error (RMSE):&quot;</span><span class="p">,</span> <span class="n">rmse</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;R-squared (R2) Score:&quot;</span><span class="p">,</span> <span class="n">r2</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="c1"># Print summary statistics</span>
<span class="nb">print</span><span class="p">(</span><span class="n">summary_statistics</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Mean Absolute Error (MAE): 0.12107017955400727
Root Mean Squared Error (RMSE): 0.14044879395085597
R-squared (R2) Score: 0.40249192535799005


       Predicted Values  True Values
count         88.000000    88.000000
mean           0.665929     0.749632
std            0.118047     0.182738
min            0.340147     0.295900
25%            0.586784     0.668400
50%            0.673757     0.776050
75%            0.753646     0.890800
max            0.885833     0.997100
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Side by side comparison for first 50 values.</span>
<span class="n">new_df</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">50</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Predicted Values</th>
      <th>True Values</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.849208</td>
      <td>0.9126</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.837691</td>
      <td>0.8929</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.632486</td>
      <td>0.7684</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.548004</td>
      <td>0.4067</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.432947</td>
      <td>0.2959</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/coderdata_logo3.jpg" alt="Logo of CoderData"/>
            </a></p><div  class="sphinxdoc-sidebar">
    <h3>Resource Links</h3>
    <ul>
        <li><a href="https://github.com/PNNL-CompBio/coderdata/" target="_blank">Github</a></li>
        <li><a href="https://pypi.org/project/coderdata/" target="_blank">PYPI</a></li>
        <li><a href="https://figshare.com/articles/dataset/CODERData2_1_0/28823159" target="_blank">FigShare</a></li>
    </ul>
</div>

<div class="sphinxdoc-sidebar">
    <h3>Quick Access</h3>
    <ul>
        <li><a href="../index.html">Home</a></li>
        <li><a href="../usage.html">Usage</a></li>
        <li><a href="../APIreference.html">API Reference</a></li>
        <li><a href="../datasets_included.html">Datasets</a></li>
        <li><a href="#">Tutorial</a></li>
        <li><a href="../contribution_guide.html">Contributing</a></li>
    </ul>
</div>
  <div>
    <h3><a href="../index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">Deep Learning Tutorial</a><ul>
<li><a class="reference internal" href="#Download-and-View-Datasets">Download and View Datasets</a></li>
</ul>
</li>
<li><a class="reference internal" href="#Develop-Deep-Learning-Model">Develop Deep Learning Model</a><ul>
<li><a class="reference internal" href="#Preparing-the-Data">Preparing the Data</a><ul>
<li><a class="reference internal" href="#Impute-Missing-Data">Impute Missing Data</a></li>
<li><a class="reference internal" href="#Scale-Input-Data">Scale Input Data</a></li>
<li><a class="reference internal" href="#Merge-Data">Merge Data</a></li>
</ul>
</li>
<li><a class="reference internal" href="#Set-the-Model-Parameters">Set the Model Parameters</a></li>
<li><a class="reference internal" href="#Compile-the-Model">Compile the Model</a></li>
<li><a class="reference internal" href="#Run-your-Model">Run your Model</a></li>
<li><a class="reference internal" href="#Evaluate-your-Model">Evaluate your Model</a><ul>
<li><a class="reference internal" href="#Plot-Loss-Function-and-Error-of-the-Model">Plot Loss Function and Error of the Model</a></li>
<li><a class="reference internal" href="#Prediction-vs-True-Values">Prediction vs True Values</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  </div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="Related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../contribution.html" title="How to Contribute"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">CoderData</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Deep Learning Tutorial</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2025, Sara Gosline, Yannick Mahlich, Jeremy Jacobson, Marissa Andrews.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.1.3.
    </div>
  </body>
</html>