{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1c301587",
   "metadata": {},
   "outputs": [],
   "source": [
    "import synapseclient\n",
    "import pandas as pd\n",
    "import os\n",
    "import wget\n",
    "from synapseclient import Project, Folder, File, Link\n",
    "import requests\n",
    "import numpy as np\n",
    "# from synapseclient import Activity\n",
    "# from synapseclient import Entity, Project, Folder, File, Link\n",
    "# from synapseclient import Evaluation, Submission, SubmissionStatus\n",
    "# from synapseclient import Wiki\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "58dde524",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "UPGRADE AVAILABLE\n",
      "\n",
      "A more recent version of the Synapse Client (3.0.0) is available. Your version (2.7.2) can be upgraded by typing:\n",
      "    pip install --upgrade synapseclient\n",
      "\n",
      "Python Synapse Client version 3.0.0 release notes\n",
      "\n",
      "https://python-docs.synapse.org/build/html/news.html\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome, JJacobson95!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "syn = synapseclient.Synapse()\n",
    "PAT = \"ENTER YOUR PERSONAL ACCESS TOKEN HERE\"\n",
    "\n",
    "\n",
    "syn.login(authToken=PAT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4df1ab58",
   "metadata": {},
   "outputs": [],
   "source": [
    "entity = syn.get('syn51674470', downloadLocation='.')\n",
    "entity = syn.get('syn25714248', downloadLocation='.')\n",
    "entity = syn.get('syn32533104', downloadLocation='.')\n",
    "entity = syn.get('syn32529921', downloadLocation='.')\n",
    "entity = syn.get('syn26642974', downloadLocation='.')\n",
    "entity = syn.get('syn26427390', downloadLocation='.')\n",
    "\n",
    "#patient data\n",
    "entity = syn.get('syn26642974', downloadLocation='.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "43faf1f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def download_from_github(raw_url, save_path):\n",
    "    response = requests.get(raw_url)\n",
    "\n",
    "    with open(save_path, 'wb') as f:\n",
    "        f.write(response.content)\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "93105a33",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def retrieve_figshare_data(url):\n",
    "    \"\"\"\n",
    "    Retrieve data from a given figshare URL.\n",
    "    \n",
    "    Return: File name (genes.csv)\n",
    "    \"\"\"\n",
    "    files_0 = os.listdir()\n",
    "    wget.download(url)\n",
    "    files_1 = os.listdir()\n",
    "    new_file = str(next(iter(set(files_1) - set(files_0))))\n",
    "    return new_file\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "703bf998",
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_and_combine(df, data_type, entrez_map_file, improve_map_file, sample_map=None):\n",
    "    \"\"\"\n",
    "    Maps and combines dataframes based on their data_type. It then merges \n",
    "    the final dataframe with provided metadata.\n",
    "    \"\"\"\n",
    "    # Initialize the list to hold mapped dataframes\n",
    "\n",
    "    # Load mapping files\n",
    "    samples = sample_map\n",
    "    genes = pd.read_csv(entrez_map_file)          # Map gene_name to entrez_id\n",
    "    \n",
    "    improve = pd.read_csv(improve_map_file)        # Map sample_id to improve_sample_id\n",
    "\n",
    "    # Process each dataframe based on its data_type\n",
    "#     for df in dataframe_list:\n",
    "    if data_type == \"transcriptomics\":\n",
    "        mapped_df = df.merge(genes, left_on='Gene', right_on='gene_symbol', how='left').reindex(\n",
    "                        columns=['transcriptomics', 'entrez_id', \"sample_id\"])\n",
    "#         mapped_df.drop(columns=['gene_id', 'gene_name', 'gene_type'], inplace=True)\n",
    "\n",
    "\n",
    "    elif data_type == \"mutations\":\n",
    "#         mapped_df = df.reindex(columns=['Entrez_Gene_Id', 'symbol', 'hgvsc'])\n",
    "        mapped_df = df.merge(genes, left_on='symbol', right_on='gene_symbol', how='left').reindex(\n",
    "                        columns=['hgvsc', 'entrez_id', \"alt_ID\"])\n",
    "        mapped_df.rename(columns={\"hgvsc\": \"mutations\"}, inplace=True)\n",
    "        mapped_df.rename(columns={\"alt_ID\": \"sample_id\"}, inplace=True)\n",
    "        mapped_df.rename(columns={\"Entrez_Gene_Id\": \"entrez_id\"}, inplace=True)\n",
    "\n",
    "    elif data_type == \"proteomics\":\n",
    "\n",
    "        # Update 'sampleID' in mapped_ids dataframe\n",
    "        sample_map['sampleID'] = sample_map['sampleID'].str.split('_').apply(lambda x: x[2])\n",
    "\n",
    "        # Rename 'sampleID' to 'id'\n",
    "        sample_map.rename(columns={\"sampleID\": \"id\"}, inplace=True)\n",
    "#         print(\"sample_map\", sample_map)\n",
    "        # Reindex the columns of mapped_ids dataframe\n",
    "        sample_map = mapped_ids.reindex(columns=['labId', \"id\"])        \n",
    "        \n",
    "        # Merge p_df with mapped_ids\n",
    "        df = df.merge(mapped_ids, left_on='id', right_on='id', how='left')\n",
    "\n",
    "        # Reindex the columns of p_df\n",
    "        df = df.reindex(columns=[\"Protein\", \"proteomics\", \"labId\"])\n",
    "\n",
    "        # Merge df with genes and reindex + rename columns\n",
    "        mapped_df = df.merge(genes, left_on='Protein', right_on='gene_symbol', how='left').reindex(\n",
    "            columns=['proteomics', 'entrez_id', 'labId'])\n",
    "\n",
    "        mapped_df.rename(\n",
    "            columns={\n",
    "                \"labId\": \"sample_id\",\n",
    "            },\n",
    "            inplace=True\n",
    "        )\n",
    "\n",
    "\n",
    "\n",
    "    # Add improve ID then Drop the sample_id and other_id columns\n",
    "    \n",
    "    mapped_df = pd.merge(mapped_df, improve[['other_id', 'improve_sample_id']], left_on='sample_id', right_on='other_id', how='left')\n",
    "    mapped_df.drop(columns=['sample_id', 'other_id'], inplace=True)\n",
    "    mapped_df.insert(0, 'improve_sample_id', mapped_df.pop('improve_sample_id'))\n",
    "    mapped_df['source'] = 'synapse'\n",
    "    mapped_df['study'] = 'BeatAML'\n",
    "\n",
    "    # Concatenate the list of dataframes into one final dataframe\n",
    "    final_dataframe = mapped_df\n",
    "\n",
    "\n",
    "    return final_dataframe\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6f628db6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100% [..........................................................................] 8156240 / 8156240"
     ]
    }
   ],
   "source": [
    "gene_url = \"https://figshare.com/ndownloader/files/40576109?private_link=525f7777039f4610ef47\"\n",
    "\n",
    "entrez_map_file = retrieve_figshare_data(gene_url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "32a36604",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r",
      "  0% [                                                                              ]     0 / 23019\r",
      " 35% [...........................                                                   ]  8192 / 23019\r",
      " 71% [.......................................................                       ] 16384 / 23019\r",
      "100% [..............................................................................] 23019 / 23019"
     ]
    }
   ],
   "source": [
    "samples_url = \"https://figshare.com/ndownloader/files/42289053?private_link=f05259d19b8c34a9a53d\"\n",
    "\n",
    "improve_map_file = retrieve_figshare_data(samples_url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e28c0b5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Transcriptomics\n",
    "#entrez_id\n",
    "#transcriptomics\n",
    "#improve_sample_id\n",
    "#source\n",
    "#study\n",
    "\n",
    "t_df = pd.read_csv(\"BeatAML_Waves1to4_RNA_data_normalized.txt\", sep = '\\t')\n",
    "\n",
    "t_df = t_df.reset_index().rename(columns={'index': 'Gene'})\n",
    "\n",
    "t_df = pd.melt(t_df, id_vars=['Gene'], var_name='sample_id', value_name='transcriptomics')\n",
    "\n",
    "# t_df\n",
    "\n",
    "t_df = map_and_combine(t_df, \"transcriptomics\", entrez_map_file, improve_map_file)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "48d7deb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "t_df\n",
    "t_df.to_csv(\"transcriptomics.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "165eb748",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "92a4d7ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Proteomics\n",
    "#entrez_id\n",
    "#proteomics\n",
    "#improve_sample_id\n",
    "#source\n",
    "#study\n",
    "\n",
    "\n",
    "\n",
    "#How to read in samples file\n",
    "# df = pd.read_csv(\"PNNL_clinical_summary_12_08_2021_updated_OS_4patients_02_28_2022.txt\", \n",
    "#                  sep='\\s+', \n",
    "#                  encoding='mac-roman',  \n",
    "#                  quotechar='\"',\n",
    "#                  error_bad_lines=False,\n",
    "#                  warn_bad_lines=True,\n",
    "#                  index_col=None)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "p_df = pd.read_csv(\"ptrc_ex10_crosstab_global_gene_corrected.txt\", sep = '\\t')\n",
    "\n",
    "p_df = p_df.reset_index().rename(columns={'index': 'Protein'})\n",
    "\n",
    "p_df = pd.melt(p_df, id_vars=['Protein'], var_name='id', value_name='proteomics')\n",
    "\n",
    "p_df\n",
    "\n",
    "# mapped_ids = pd.read_csv(\"Data Available for Proteomic Samples.xlsx\")\n",
    "mapped_ids = pd.read_excel(\"Data Available for Proteomic Samples.xlsx\")\n",
    "\n",
    "p_df = map_and_combine(p_df, \"proteomics\", entrez_map_file, improve_map_file, mapped_ids)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "16839bfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_df\n",
    "p_df.to_csv(\"proteomics.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e3b6fb4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Mutations\n",
    "#entrez_id\n",
    "#mutation\n",
    "#improve_sample_id\n",
    "#source\n",
    "#study\n",
    "\n",
    "\n",
    "m_df = pd.read_csv(\"beataml_wes_wv1to4_mutations_177_samples.txt\", sep = '\\t')\n",
    "m_df = map_and_combine(m_df, \"mutations\", entrez_map_file, improve_map_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "85c652ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "m_df\n",
    "m_df.to_csv(\"mutations.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d3bd25d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#Note - This file was converted from txt to csv due to byte reading issues.\n",
    "#Note - assign_improve_ids.py was run on this prior to here.\n",
    "# summary = pd.read_csv(\"PNNL_clinical_summary_12_08_2021_updated_OS_4patients_02_28_2022.csv\")\n",
    "\n",
    "# samples = summary[[\"labId\",\"improve_sample_id\",\"specificDxAtInclusion\",\"specimenType\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f580544e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# samples = summary[[\"labId\",\"improve_sample_id\",\"specificDxAtInclusion\",\"specimenType\"]]\n",
    "# samples.rename(columns={\"labId\": \"other_id\"}, inplace=True)\n",
    "# samples.rename(columns={\"specificDxAtInclusion\": \"other_names\"}, inplace=True)\n",
    "# samples.rename(columns={\"specimenType\": \"common_name\"}, inplace=True)\n",
    "\n",
    "\n",
    "# samples[\"cancer_type\"] = \"ACUTE MYELOID LEUKAEMIA\"\n",
    "\n",
    "# samples[\"model_type\"] = \"ex vivo\"\n",
    "# samples[\"other_id_source\"] = \"beatAML\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b2b3de82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "18119843",
   "metadata": {},
   "outputs": [],
   "source": [
    "# samples.to_csv(\"samples.csv\", index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "28d17da7",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# def retreive_drug_info(compound_name):\n",
    "#     url = f\"https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/name/{compound_name}/property/CanonicalSMILES,IsomericSMILES,InChIKey,MolecularFormula,MolecularWeight/JSON\"\n",
    "\n",
    "#     response = requests.get(url)\n",
    "#     data = response.json()\n",
    "\n",
    "#     properties = data[\"PropertyTable\"][\"Properties\"][0]\n",
    "#     canSMILES = properties[\"CanonicalSMILES\"]\n",
    "#     isoSMILES = properties[\"IsomericSMILES\"]\n",
    "#     inchikey = properties[\"InChIKey\"]\n",
    "#     formula = properties[\"MolecularFormula\"]\n",
    "#     weight = properties[\"MolecularWeight\"]\n",
    "\n",
    "#     return(canSMILES, isoSMILES, inchikey, formula, weight)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "632d8945",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ee39eaa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a2867680",
   "metadata": {},
   "outputs": [],
   "source": [
    "def retrieve_drug_info(compound_name):\n",
    "    if pd.isna(compound_name):\n",
    "        return np.nan, np.nan, np.nan, np.nan, np.nan\n",
    "    \n",
    "    url = f\"https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/name/{compound_name}/property/CanonicalSMILES,IsomericSMILES,InChIKey,MolecularFormula,MolecularWeight/JSON\"\n",
    "    response = requests.get(url)\n",
    "\n",
    "    if response.status_code != 200:\n",
    "        return np.nan, np.nan, np.nan, np.nan, np.nan\n",
    "    \n",
    "    data = response.json()\n",
    "    if \"PropertyTable\" in data:\n",
    "        properties = data[\"PropertyTable\"][\"Properties\"][0]\n",
    "        canSMILES = properties.get(\"CanonicalSMILES\", np.nan)\n",
    "        isoSMILES = properties.get(\"IsomericSMILES\", np.nan)\n",
    "        inchikey = properties.get(\"InChIKey\", np.nan)\n",
    "        formula = properties.get(\"MolecularFormula\", np.nan)\n",
    "        weight = properties.get(\"MolecularWeight\", np.nan)\n",
    "\n",
    "        return canSMILES, isoSMILES, inchikey, formula, weight\n",
    "    else:\n",
    "        return np.nan, np.nan, np.nan, np.nan, np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "117220fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def update_dataframe_with_pubchem(d_df):\n",
    "    # Get unique chem_name values and retrieve their data\n",
    "    chem_names = d_df['chem_name'].dropna().unique()\n",
    "    chem_data_dict = {}\n",
    "    for name in chem_names:\n",
    "        print(\"Attempting to call pubchem API for chem_name: \", name)\n",
    "        chem_data_dict[name] = retrieve_drug_info(name)\n",
    "\n",
    "    # Identify rows whose chem_name failed and retrieve the other_name for those rows\n",
    "    failed_chem_names = {k for k, v in chem_data_dict.items() if all(pd.isna(val) for val in v)}\n",
    "    other_names = d_df[d_df['chem_name'].isin(failed_chem_names)]['other_name'].dropna().unique()\n",
    "    other_data_dict = {}\n",
    "    for name in other_names:\n",
    "        print(\"Attempting to call pubchem API for other_name: \", name)\n",
    "        other_data_dict[name] = retrieve_drug_info(name)\n",
    "\n",
    "    # Combine both dictionaries for easy lookup\n",
    "    data_dict = {**chem_data_dict, **other_data_dict}\n",
    "\n",
    "    # Update the DataFrame using the data dictionary\n",
    "    for idx, row in d_df.iterrows():\n",
    "        if row['chem_name'] in data_dict and not all(pd.isna(val) for val in data_dict[row['chem_name']]):\n",
    "            values = data_dict[row['chem_name']]\n",
    "        else:\n",
    "            values = data_dict.get(row['other_name'], (np.nan, np.nan, np.nan, np.nan, np.nan))\n",
    "        \n",
    "        d_df.at[idx, \"canSMILES\"] = values[0]\n",
    "        d_df.at[idx, \"isoSMILES\"] = values[1]\n",
    "        d_df.at[idx, \"inchikey\"] = values[2]\n",
    "        d_df.at[idx, \"formula\"] = values[3]\n",
    "        d_df.at[idx, \"weight\"] = values[4]\n",
    "    \n",
    "    return d_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "63091737",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_drug_info(d_df,drug_map):\n",
    "    result_df = d_df.merge(drug_map[['isoSMILES', 'improve_drug_id']], on='isoSMILES', how='left')\n",
    "    result_df = result_df[[\"sample_id\",\"improve_drug_id\",\"chem_name\", \"auc\",\"formula\",\"weight\", \"canSMILES\",\"inchikey\",\"isoSMILES\"]]\n",
    "    return result_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ad5fb84f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drug map file\n",
    "# drug_url = retrieve_figshare_data(\"https://figshare.com/ndownloader/files/41259273?private_link=525f7777039f4610ef47\")\n",
    "def format_drug_map(drug_map_path):\n",
    "    drug_map = pd.read_csv(drug_map_path, sep = \"\\t\")\n",
    "    drug_map = drug_map.drop_duplicates(subset='isoSMILES', keep='first')\n",
    "    return drug_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "359d79b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Drug Response\n",
    "def format_drug_df(drug_path):\n",
    "    d_df = pd.read_csv(drug_path, index_col=None)\n",
    "    d_df.drop(\"Unnamed: 0\",axis='columns',inplace=True)\n",
    "    d_df[['chem_name', 'other_name']] = d_df['inhibitor'].str.extract(r'^(.*?)\\s*(?:\\((.+)\\))?$')\n",
    "    d_df[\"chem_name\"] = d_df[\"chem_name\"].str.replace('\\s-\\s', ':')\n",
    "    return d_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "81364664",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_improve_id(previous_df, new_df):\n",
    "    \n",
    "    # Extract the maximum value of improve_drug_id from the previous dataframe\n",
    "    # Assuming the format is SMI_number. We get rid of 'SMI_' and then convert to int.\n",
    "    max_id = max([int(val.replace('SMI_', '')) for val in previous_df['improve_drug_id'].tolist() if pd.notnull(val) and val.startswith('SMI_')])\n",
    "\n",
    "    # Identify isoSMILES in the new dataframe that don't exist in the old dataframe\n",
    "    unique_new_smiles = set(new_df['isoSMILES']) - set(previous_df['isoSMILES'])\n",
    "    \n",
    "    # Identify rows in the new dataframe with isoSMILES that are unique and where improve_drug_id is NaN\n",
    "    mask = (new_df['isoSMILES'].isin(unique_new_smiles)) & (new_df['improve_drug_id'].isna())\n",
    "    \n",
    "    # Create a mapping of unique new isoSMILES to improve_drug_id\n",
    "    id_map = {}\n",
    "    for smiles in unique_new_smiles:\n",
    "        max_id += 1\n",
    "        id_map[smiles] = f\"SMI_{max_id}\"\n",
    "    \n",
    "    # Apply the mapping to the new dataframe for rows with unique isoSMILES and NaN improve_drug_id\n",
    "    new_df.loc[mask, 'improve_drug_id'] = new_df['isoSMILES'].map(id_map)\n",
    "    \n",
    "    return new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "19600335",
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_exp_to_improve(df,improve_map_file):\n",
    "    improve = pd.read_csv(improve_map_file)        # Map sample_id to improve_sample_id\n",
    "    mapped_df = pd.merge(df, improve[['other_id', 'improve_sample_id']], left_on='sample_id', right_on='other_id', how='left')\n",
    "    mapped_df.drop(columns=['sample_id', 'other_id'], inplace=True)\n",
    "    mapped_df.insert(0, 'improve_sample_id', mapped_df.pop('improve_sample_id'))\n",
    "    mapped_df['source'] = 'synapse'\n",
    "    mapped_df['study'] = 'BeatAML'\n",
    "    mapped_df['ic50'] = np.nan\n",
    "    mapped_df['ec50'] = np.nan\n",
    "    mapped_df['ec50se'] = np.nan\n",
    "    mapped_df['einf'] = np.nan\n",
    "    mapped_df['hs'] = np.nan\n",
    "    mapped_df['aac1'] = np.nan\n",
    "    mapped_df['auc1'] = np.nan\n",
    "    mapped_df['dss1'] = np.nan\n",
    "    return mapped_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "5863e0ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_sixth_from_last_quote(s):\n",
    "    reversed_s = s[::-1]  # Reverse the string\n",
    "    count = 0\n",
    "    index_to_remove = None\n",
    "    for i, char in enumerate(reversed_s):\n",
    "        if char == '\"':\n",
    "            count += 1\n",
    "        if count == 6:\n",
    "            index_to_remove = i\n",
    "            break\n",
    "    # Remove the 6th occurrence of the quote character\n",
    "    if index_to_remove is not None:\n",
    "        reversed_s = reversed_s[:index_to_remove] + reversed_s[index_to_remove+1:]\n",
    "    return reversed_s[::-1]  # Reverse again to obtain the original order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d46d5914",
   "metadata": {},
   "outputs": [],
   "source": [
    "def modify_patient_file(file_path):\n",
    "    with open(file_path, 'r', encoding='ISO-8859-1') as f:\n",
    "        lines = f.readlines()\n",
    "    # Modify the 110th line (0-based index is 109)\n",
    "    lines[109] = remove_sixth_from_last_quote(lines[109])\n",
    "    with open(file_path, 'w', encoding='utf-8') as f:\n",
    "        f.writelines(lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cac8406a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_samples_file()\n",
    "    file_path = 'PNNL_clinical_summary_12_08_2021_updated_OS_4patients_02_28_2022.txt'\n",
    "    modify_patient_file(file_path)\n",
    "    samples = pd.read_csv('PNNL_clinical_summary_12_08_2021_updated_OS_4patients_02_28_2022.txt', delimiter=' ', skipinitialspace=True, encoding='ISO-8859-1', quotechar='\"')\n",
    "    maxval = max(pd.read_csv('https://raw.githubusercontent.com/PNNL-CompBio/candleDataProcessing/main/hcmi/samples.csv').improve_sample_id)\n",
    "    # Create a dictionary that maps unique lab-id values to sequential integers\n",
    "    mapping = {labId: i for i, labId in enumerate(samples['labId'].unique(), start=(int(maxval)+1))}\n",
    "    # Use the map method to create the new column based on the lab-id column\n",
    "    samples['improve_sample_id'] = samples['labId'].map(mapping)\n",
    "    samples = samples[[\"labId\",\"improve_sample_id\",\"specificDxAtInclusion\",\"specimenType\"]]\n",
    "    samples.rename(columns={\"labId\": \"other_id\"}, inplace=True)\n",
    "    samples.rename(columns={\"specificDxAtInclusion\": \"other_names\"}, inplace=True)\n",
    "    samples.rename(columns={\"specimenType\": \"common_name\"}, inplace=True)\n",
    "    samples[\"cancer_type\"] = \"ACUTE MYELOID LEUKAEMIA\"\n",
    "    samples[\"model_type\"] = \"ex vivo\"\n",
    "    samples[\"other_id_source\"] = \"beatAML\"\n",
    "    samples.to_csv(\"samples.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cf60b0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "generate_samples_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c64b10a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[WARNING] /var/folders/wt/_yr7rg_13t76sq_q5cw57hxw0000gn/T/ipykernel_84914/2428304630.py:6: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  d_df[\"chem_name\"] = d_df[\"chem_name\"].str.replace('\\s-\\s', ':')\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attempting to call pubchem API for chem_name:  AKT Inhibitor IV\n",
      "Attempting to call pubchem API for chem_name:  AKT Inhibitor X\n",
      "Attempting to call pubchem API for chem_name:  AMPK Inhibitor\n",
      "Attempting to call pubchem API for chem_name:  AST-487\n",
      "Attempting to call pubchem API for chem_name:  AZD1152-HQPA\n",
      "Attempting to call pubchem API for chem_name:  Alisertib\n",
      "Attempting to call pubchem API for chem_name:  Bortezomib\n",
      "Attempting to call pubchem API for chem_name:  CYT387\n",
      "Attempting to call pubchem API for chem_name:  Canertinib\n",
      "Attempting to call pubchem API for chem_name:  Crenolanib\n",
      "Attempting to call pubchem API for chem_name:  Doramapimod\n",
      "Attempting to call pubchem API for chem_name:  Dovitinib\n",
      "Attempting to call pubchem API for chem_name:  Erlotinib\n",
      "Attempting to call pubchem API for chem_name:  Flavopiridol\n",
      "Attempting to call pubchem API for chem_name:  GDC-0941\n",
      "Attempting to call pubchem API for chem_name:  GW-2580\n",
      "Attempting to call pubchem API for chem_name:  Gefitinib\n",
      "Attempting to call pubchem API for chem_name:  Go6976\n",
      "Attempting to call pubchem API for chem_name:  H-89\n",
      "Attempting to call pubchem API for chem_name:  INK-128\n",
      "Attempting to call pubchem API for chem_name:  Idelalisib\n",
      "Attempting to call pubchem API for chem_name:  Imatinib\n",
      "Attempting to call pubchem API for chem_name:  JAK Inhibitor I\n",
      "Attempting to call pubchem API for chem_name:  JNJ-7706621\n",
      "Attempting to call pubchem API for chem_name:  JNK Inhibitor II\n",
      "Attempting to call pubchem API for chem_name:  KN92\n",
      "Attempting to call pubchem API for chem_name:  KN93\n",
      "Attempting to call pubchem API for chem_name:  LY-333531\n",
      "Attempting to call pubchem API for chem_name:  LY294002\n",
      "Attempting to call pubchem API for chem_name:  Lapatinib\n",
      "Attempting to call pubchem API for chem_name:  Linifanib\n",
      "Attempting to call pubchem API for chem_name:  MEK1/2 Inhibitor\n",
      "Attempting to call pubchem API for chem_name:  Midostaurin\n",
      "Attempting to call pubchem API for chem_name:  Motesanib\n",
      "Attempting to call pubchem API for chem_name:  NF-kB Activation Inhibitor\n",
      "Attempting to call pubchem API for chem_name:  PD153035\n",
      "Attempting to call pubchem API for chem_name:  PD98059\n",
      "Attempting to call pubchem API for chem_name:  PI-103\n",
      "Attempting to call pubchem API for chem_name:  PP2\n",
      "Attempting to call pubchem API for chem_name:  PRT062070\n",
      "Attempting to call pubchem API for chem_name:  Pazopanib\n",
      "Attempting to call pubchem API for chem_name:  Pelitinib\n",
      "Attempting to call pubchem API for chem_name:  Quizartinib\n",
      "Attempting to call pubchem API for chem_name:  R406\n",
      "Attempting to call pubchem API for chem_name:  RAF265\n",
      "Attempting to call pubchem API for chem_name:  Rapamycin\n",
      "Attempting to call pubchem API for chem_name:  Roscovitine\n",
      "Attempting to call pubchem API for chem_name:  SB-202190\n",
      "Attempting to call pubchem API for chem_name:  SB-203580\n",
      "Attempting to call pubchem API for chem_name:  SB-431542\n",
      "Attempting to call pubchem API for chem_name:  SNS-032\n",
      "Attempting to call pubchem API for chem_name:  STO609\n",
      "Attempting to call pubchem API for chem_name:  SU14813\n",
      "Attempting to call pubchem API for chem_name:  SYK Inhibitor\n",
      "Attempting to call pubchem API for chem_name:  Sorafenib\n",
      "Attempting to call pubchem API for chem_name:  Tandutinib\n",
      "Attempting to call pubchem API for chem_name:  Tofacitinib\n",
      "Attempting to call pubchem API for chem_name:  Tozasertib\n",
      "Attempting to call pubchem API for chem_name:  VX-745\n",
      "Attempting to call pubchem API for chem_name:  Vandetanib\n",
      "Attempting to call pubchem API for chem_name:  Vatalanib\n",
      "Attempting to call pubchem API for chem_name:  YM-155\n",
      "Attempting to call pubchem API for chem_name:  p38 MAP Kinase Inhibitor\n",
      "Attempting to call pubchem API for chem_name:  Dasatinib\n",
      "Attempting to call pubchem API for chem_name:  Nilotinib\n",
      "Attempting to call pubchem API for chem_name:  Ponatinib\n",
      "Attempting to call pubchem API for chem_name:  Sunitinib\n",
      "Attempting to call pubchem API for chem_name:  A-674563\n",
      "Attempting to call pubchem API for chem_name:  AT7519\n",
      "Attempting to call pubchem API for chem_name:  AZD1480\n",
      "Attempting to call pubchem API for chem_name:  Afatinib\n",
      "Attempting to call pubchem API for chem_name:  BEZ235\n",
      "Attempting to call pubchem API for chem_name:  BI-2536\n",
      "Attempting to call pubchem API for chem_name:  BMS-345541\n",
      "Attempting to call pubchem API for chem_name:  Bosutinib\n",
      "Attempting to call pubchem API for chem_name:  CI-1040\n",
      "Attempting to call pubchem API for chem_name:  Cediranib\n",
      "Attempting to call pubchem API for chem_name:  Crizotinib\n",
      "Attempting to call pubchem API for chem_name:  Foretinib\n",
      "Attempting to call pubchem API for chem_name:  GDC-0879\n",
      "Attempting to call pubchem API for chem_name:  GSK-1838705A\n",
      "Attempting to call pubchem API for chem_name:  GSK690693\n",
      "Attempting to call pubchem API for chem_name:  Ibrutinib\n",
      "Attempting to call pubchem API for chem_name:  JNJ-28312141\n",
      "Attempting to call pubchem API for chem_name:  KI20227\n",
      "Attempting to call pubchem API for chem_name:  KW-2449\n",
      "Attempting to call pubchem API for chem_name:  Lestaurtinib\n",
      "Attempting to call pubchem API for chem_name:  MLN120B\n",
      "Attempting to call pubchem API for chem_name:  Masitinib\n",
      "Attempting to call pubchem API for chem_name:  NVP-TAE684\n",
      "Attempting to call pubchem API for chem_name:  Neratinib\n",
      "Attempting to call pubchem API for chem_name:  PD173955\n",
      "Attempting to call pubchem API for chem_name:  PHA-665752\n",
      "Attempting to call pubchem API for chem_name:  PLX-4720\n",
      "Attempting to call pubchem API for chem_name:  PP242\n",
      "Attempting to call pubchem API for chem_name:  R547\n",
      "Attempting to call pubchem API for chem_name:  SGX-523\n",
      "Attempting to call pubchem API for chem_name:  Selumetinib\n",
      "Attempting to call pubchem API for chem_name:  Staurosporine\n",
      "Attempting to call pubchem API for chem_name:  TG100-115\n",
      "Attempting to call pubchem API for chem_name:  TG101348\n",
      "Attempting to call pubchem API for chem_name:  Trametinib\n",
      "Attempting to call pubchem API for chem_name:  Axitinib\n",
      "Attempting to call pubchem API for chem_name:  Vargetef\n",
      "Attempting to call pubchem API for chem_name:  Ruxolitinib\n",
      "Attempting to call pubchem API for chem_name:  CHIR-99021\n",
      "Attempting to call pubchem API for chem_name:  GSK-1904529A\n",
      "Attempting to call pubchem API for chem_name:  MK-2206\n",
      "Attempting to call pubchem API for chem_name:  MGCD-265\n",
      "Attempting to call pubchem API for chem_name:  17-AAG\n",
      "Attempting to call pubchem API for chem_name:  ABT-737\n",
      "Attempting to call pubchem API for chem_name:  Cabozantinib\n",
      "Attempting to call pubchem API for chem_name:  Elesclomol\n",
      "Attempting to call pubchem API for chem_name:  JNJ-38877605\n",
      "Attempting to call pubchem API for chem_name:  KU-55933\n",
      "Attempting to call pubchem API for chem_name:  MLN8054\n",
      "Attempting to call pubchem API for chem_name:  NVP-ADW742\n",
      "Attempting to call pubchem API for chem_name:  PHT-427\n",
      "Attempting to call pubchem API for chem_name:  PRT062607\n",
      "Attempting to call pubchem API for chem_name:  Regorafenib\n",
      "Attempting to call pubchem API for chem_name:  S31-201\n",
      "Attempting to call pubchem API for chem_name:  SU11274\n",
      "Attempting to call pubchem API for chem_name:  Saracatinib\n",
      "Attempting to call pubchem API for chem_name:  Tivozanib\n",
      "Attempting to call pubchem API for chem_name:  Vismodegib\n",
      "Attempting to call pubchem API for chem_name:  XAV-939\n",
      "Attempting to call pubchem API for chem_name:  Vemurafenib\n",
      "Attempting to call pubchem API for chem_name:  Arsenic Trioxide\n",
      "Attempting to call pubchem API for chem_name:  Axitinib:Doramapimod\n",
      "Attempting to call pubchem API for chem_name:  Axitinib:GW-2580\n",
      "Attempting to call pubchem API for chem_name:  Axitinib:KI20227\n",
      "Attempting to call pubchem API for chem_name:  Axitinib:Motesanib\n",
      "Attempting to call pubchem API for chem_name:  Azacytidine\n",
      "Attempting to call pubchem API for chem_name:  Azacytidine :Vemurafenib\n",
      "Attempting to call pubchem API for chem_name:  Azacytidine:Quizartinib\n",
      "Attempting to call pubchem API for chem_name:  DBZ\n",
      "Attempting to call pubchem API for chem_name:  Dasatinib:GW-2580\n",
      "Attempting to call pubchem API for chem_name:  Dasatinib:Midostaurin\n",
      "Attempting to call pubchem API for chem_name:  Dasatinib:NF-kb ActInh\n",
      "Attempting to call pubchem API for chem_name:  Dasatinib:Pazopanib\n",
      "Attempting to call pubchem API for chem_name:  Dasatinib:Sorafenib\n",
      "Attempting to call pubchem API for chem_name:  Doramapimod:Dasatinib\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attempting to call pubchem API for chem_name:  Doramapimod:GW-2580\n",
      "Attempting to call pubchem API for chem_name:  Doramapimod:Idelalisib\n",
      "Attempting to call pubchem API for chem_name:  Doramapimod:Midostaurin\n",
      "Attempting to call pubchem API for chem_name:  Doramapimod:NF-kb ActInh\n",
      "Attempting to call pubchem API for chem_name:  Doramapimod:Ruxolitinib\n",
      "Attempting to call pubchem API for chem_name:  Doramapimod:Sorafenib\n",
      "Attempting to call pubchem API for chem_name:  GW-2580:Doramapimod\n",
      "Attempting to call pubchem API for chem_name:  GW-2580:KI20227\n",
      "Attempting to call pubchem API for chem_name:  GW-2580:Midostaurin\n",
      "Attempting to call pubchem API for chem_name:  GW-2580:Pazopanib\n",
      "Attempting to call pubchem API for chem_name:  GW-2580:Sunitinib\n",
      "Attempting to call pubchem API for chem_name:  Idelalisib:Dasatinib\n",
      "Attempting to call pubchem API for chem_name:  Idelalisib:GW-2580\n",
      "Attempting to call pubchem API for chem_name:  Idelalisib:Midostaurin\n",
      "Attempting to call pubchem API for chem_name:  Idelalisib:Ruxolitinib\n",
      "Attempting to call pubchem API for chem_name:  Linifanib:Ruxolitinib\n",
      "Attempting to call pubchem API for chem_name:  Midostaurin:Alistertib\n",
      "Attempting to call pubchem API for chem_name:  Midostaurin:Ruxolitinib\n",
      "Attempting to call pubchem API for chem_name:  Midostaurin:Sorafenib\n",
      "Attempting to call pubchem API for chem_name:  Motesanib:Doramapimod\n",
      "Attempting to call pubchem API for chem_name:  Motesanib:GW-2580\n",
      "Attempting to call pubchem API for chem_name:  Motesanib:KI20227\n",
      "Attempting to call pubchem API for chem_name:  Motesanib:Sunitinib\n",
      "Attempting to call pubchem API for chem_name:  NFkB Actinhib:Midostaurin\n",
      "Attempting to call pubchem API for chem_name:  NFkB Actinhib:Vemurafenib\n",
      "Attempting to call pubchem API for chem_name:  NVPTAE684: Midostaurin\n",
      "Attempting to call pubchem API for chem_name:  NVPTAE684:Foretinib\n",
      "Attempting to call pubchem API for chem_name:  Pazopanib:Midostaurin\n",
      "Attempting to call pubchem API for chem_name:  Quizartinib:Arsenic\n",
      "Attempting to call pubchem API for chem_name:  Quizartinib:NVPTAE684\n",
      "Attempting to call pubchem API for chem_name:  Quizartinib:NVPTAE684:Arsenic\n",
      "Attempting to call pubchem API for chem_name:  Ruxolitinib :NF-kb ActInh\n",
      "Attempting to call pubchem API for chem_name:  Ruxolitinib:Alistertib\n",
      "Attempting to call pubchem API for chem_name:  Ruxolitinib:Dasatinib\n",
      "Attempting to call pubchem API for chem_name:  AGI-5198\n",
      "Attempting to call pubchem API for chem_name:  AGI-6780\n",
      "Attempting to call pubchem API for chem_name:  Bay 11-7085\n",
      "Attempting to call pubchem API for chem_name:  JQ1\n",
      "Attempting to call pubchem API for chem_name:  Lenvatinib\n",
      "Attempting to call pubchem API for chem_name:  Venetoclax\n",
      "Attempting to call pubchem API for chem_name:  Volasertib\n",
      "Attempting to call pubchem API for chem_name:  AGI combo\n",
      "Attempting to call pubchem API for chem_name:  AGI combo:Crizotinib\n",
      "Attempting to call pubchem API for chem_name:  AGI combo:Quizartinib\n",
      "Attempting to call pubchem API for chem_name:  Alisertib:Crizotinib\n",
      "Attempting to call pubchem API for chem_name:  Axitinib:Bosutinib\n",
      "Attempting to call pubchem API for chem_name:  Cabozantinib:Ruxolitinib\n",
      "Attempting to call pubchem API for chem_name:  Crizotinib:Quizartinib\n",
      "Attempting to call pubchem API for chem_name:  Cytarabine\n",
      "Attempting to call pubchem API for chem_name:  Cytarabine:Nutlin 3a\n",
      "Attempting to call pubchem API for chem_name:  Cytarabine:Vismodegib\n",
      "Attempting to call pubchem API for chem_name:  Dasatinib:AGI combo\n",
      "Attempting to call pubchem API for chem_name:  Dasatinib:JQ1\n",
      "Attempting to call pubchem API for chem_name:  Dasatinib:Quizartinib\n",
      "Attempting to call pubchem API for chem_name:  Dasatinib:Rapamycin\n",
      "Attempting to call pubchem API for chem_name:  Dasatinib:Trametinib\n",
      "Attempting to call pubchem API for chem_name:  Dasatinib:Vatalanib\n",
      "Attempting to call pubchem API for chem_name:  Dasatinib:Velcade\n",
      "Attempting to call pubchem API for chem_name:  Dasatinib:Venetoclax\n",
      "Attempting to call pubchem API for chem_name:  Doramapimod:AGI combo\n",
      "Attempting to call pubchem API for chem_name:  Doramapimod:Venetoclax\n",
      "Attempting to call pubchem API for chem_name:  Entospletinib\n",
      "Attempting to call pubchem API for chem_name:  Idelalisib:AGI combo\n",
      "Attempting to call pubchem API for chem_name:  Idelalisib:JQ1\n",
      "Attempting to call pubchem API for chem_name:  Idelalisib:Quizartinib\n",
      "Attempting to call pubchem API for chem_name:  Idelalisib:Trametinib\n",
      "Attempting to call pubchem API for chem_name:  Idelalisib:Venetoclax\n",
      "Attempting to call pubchem API for chem_name:  Nutlin 3a\n",
      "Attempting to call pubchem API for chem_name:  Quizartinib:Ruxolitinib\n",
      "Attempting to call pubchem API for chem_name:  Quizartinib:Trametinib\n",
      "Attempting to call pubchem API for chem_name:  Rapamycin:Trametinib\n",
      "Attempting to call pubchem API for chem_name:  Ruxolitinib:AGI combo\n",
      "Attempting to call pubchem API for chem_name:  Ruxolitinib:JQ1\n",
      "Attempting to call pubchem API for chem_name:  Ruxolitinib:Sunitinib\n",
      "Attempting to call pubchem API for chem_name:  Ruxolitinib:Venetoclax\n",
      "Attempting to call pubchem API for chem_name:  Sorafenib:Entospletinib\n",
      "Attempting to call pubchem API for chem_name:  Sorafenib:JQ1\n",
      "Attempting to call pubchem API for chem_name:  Sorafenib:Venetoclax\n",
      "Attempting to call pubchem API for chem_name:  Trametinib:Bortezomib\n",
      "Attempting to call pubchem API for chem_name:  Trametinib:Vargetef\n",
      "Attempting to call pubchem API for chem_name:  Vandetanib:Vemurafenib\n",
      "Attempting to call pubchem API for chem_name:  AGI combo:Sorafenib\n",
      "Attempting to call pubchem API for chem_name:  Doramapimod:JQ1\n",
      "Attempting to call pubchem API for chem_name:  Doramapimod:Palbociclib\n",
      "Attempting to call pubchem API for chem_name:  Doramapimod:Panobinostat\n",
      "Attempting to call pubchem API for chem_name:  Ibrutinib:Quizartinib\n",
      "Attempting to call pubchem API for chem_name:  Idelalisib:Palbociclib\n",
      "Attempting to call pubchem API for chem_name:  Idelalisib:Panobinostat\n",
      "Attempting to call pubchem API for chem_name:  JQ1:Palbociclib\n",
      "Attempting to call pubchem API for chem_name:  JQ1:Panobinostat\n",
      "Attempting to call pubchem API for chem_name:  OTX-015\n",
      "Attempting to call pubchem API for chem_name:  Palbociclib\n",
      "Attempting to call pubchem API for chem_name:  Panobinostat\n",
      "Attempting to call pubchem API for chem_name:  Panobinostat:Palbociclib\n",
      "Attempting to call pubchem API for chem_name:  Quizartinib:JQ1\n",
      "Attempting to call pubchem API for chem_name:  Quizartinib:Palbociclib\n",
      "Attempting to call pubchem API for chem_name:  Quizartinib:Panobinostat\n",
      "Attempting to call pubchem API for chem_name:  Quizartinib:Venetoclax\n",
      "Attempting to call pubchem API for chem_name:  Ruxolitinib:Palbociclib\n",
      "Attempting to call pubchem API for chem_name:  Ruxolitinib:Panobinostat\n",
      "Attempting to call pubchem API for chem_name:  Sorafenib:Palbociclib\n",
      "Attempting to call pubchem API for chem_name:  Sorafenib:Panobinostat\n",
      "Attempting to call pubchem API for chem_name:  Trametinib:JQ1\n",
      "Attempting to call pubchem API for chem_name:  Trametinib:Palbociclib\n",
      "Attempting to call pubchem API for chem_name:  Trametinib:Panobinostat\n",
      "Attempting to call pubchem API for chem_name:  Trametinib:Venetoclax\n",
      "Attempting to call pubchem API for chem_name:  Venetoclax:Ibrutinib\n",
      "Attempting to call pubchem API for chem_name:  Venetoclax:JQ1\n",
      "Attempting to call pubchem API for chem_name:  Venetoclax:Palbociclib\n",
      "Attempting to call pubchem API for chem_name:  Venetoclax:Panobinostat\n",
      "Attempting to call pubchem API for chem_name:  AGI-6780:CM\n",
      "Attempting to call pubchem API for chem_name:  Azacitidine-CM\n",
      "Attempting to call pubchem API for chem_name:  Gilteritinib\n",
      "Attempting to call pubchem API for chem_name:  Lenalidomide\n",
      "Attempting to call pubchem API for chem_name:  Lovastatin\n",
      "Attempting to call pubchem API for chem_name:  SR9011\n",
      "Attempting to call pubchem API for chem_name:  Artemisinin\n",
      "Attempting to call pubchem API for chem_name:  Selinexor\n",
      "Attempting to call pubchem API for chem_name:  Entrectinib\n",
      "Attempting to call pubchem API for chem_name:  Venetoclax:Artemisinin\n",
      "Attempting to call pubchem API for chem_name:  Olaparib\n",
      "Attempting to call pubchem API for chem_name:  Indisulam\n",
      "Attempting to call pubchem API for chem_name:  F_S_V\n",
      "Attempting to call pubchem API for chem_name:  AT-101\n",
      "Attempting to call pubchem API for chem_name:  Birinapant\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attempting to call pubchem API for chem_name:  GSK-2879552\n",
      "Attempting to call pubchem API for chem_name:  Azacytidine:Venetoclax\n",
      "Attempting to call pubchem API for chem_name:  Bortezomib:AT-101\n",
      "Attempting to call pubchem API for chem_name:  Bortezomib:Birinapant\n",
      "Attempting to call pubchem API for chem_name:  Bortezomib:GSK2879552\n",
      "Attempting to call pubchem API for chem_name:  Bortezomib:OTX-015\n",
      "Attempting to call pubchem API for chem_name:  Bortezomib:Palbociclib\n",
      "Attempting to call pubchem API for chem_name:  Bortezomib:Venetoclax\n",
      "Attempting to call pubchem API for chem_name:  Doramapimod:AT-101\n",
      "Attempting to call pubchem API for chem_name:  Doramapimod:Birinapant\n",
      "Attempting to call pubchem API for chem_name:  Doramapimod:GSK2879552\n",
      "Attempting to call pubchem API for chem_name:  Doramapimod:OTX-015\n",
      "Attempting to call pubchem API for chem_name:  GW-2580:AT-101\n",
      "Attempting to call pubchem API for chem_name:  GW-2580:Birinapant\n",
      "Attempting to call pubchem API for chem_name:  GW-2580:GSK2879552\n",
      "Attempting to call pubchem API for chem_name:  GW-2580:OTX-015\n",
      "Attempting to call pubchem API for chem_name:  GW-2580:Palbociclib\n",
      "Attempting to call pubchem API for chem_name:  GW-2580:Venetoclax\n",
      "Attempting to call pubchem API for chem_name:  Idelalisib:AT-101\n",
      "Attempting to call pubchem API for chem_name:  Idelalisib:Birinapant\n",
      "Attempting to call pubchem API for chem_name:  Idelalisib:GSK2879552\n",
      "Attempting to call pubchem API for chem_name:  Idelalisib:OTX-015\n",
      "Attempting to call pubchem API for chem_name:  Olaparib:AT-101\n",
      "Attempting to call pubchem API for chem_name:  Olaparib:Birinapant\n",
      "Attempting to call pubchem API for chem_name:  Olaparib:GSK2879552\n",
      "Attempting to call pubchem API for chem_name:  Olaparib:OTX-015\n",
      "Attempting to call pubchem API for chem_name:  Olaparib:Palbociclib\n",
      "Attempting to call pubchem API for chem_name:  Olaparib:Trametinib\n",
      "Attempting to call pubchem API for chem_name:  Olaparib:Venetoclax\n",
      "Attempting to call pubchem API for chem_name:  Quizartinib:AT-101\n",
      "Attempting to call pubchem API for chem_name:  Quizartinib:Birinapant\n",
      "Attempting to call pubchem API for chem_name:  Quizartinib:GSK2879552\n",
      "Attempting to call pubchem API for chem_name:  Quizartinib:OTX-015\n",
      "Attempting to call pubchem API for chem_name:  Ruxolitinib:AT-101\n",
      "Attempting to call pubchem API for chem_name:  Ruxolitinib:Birinapant\n",
      "Attempting to call pubchem API for chem_name:  Ruxolitinib:GSK2879552\n",
      "Attempting to call pubchem API for chem_name:  Ruxolitinib:OTX-015\n",
      "Attempting to call pubchem API for chem_name:  Trametinib:AT-101\n",
      "Attempting to call pubchem API for chem_name:  Trametinib:Birinapant\n",
      "Attempting to call pubchem API for chem_name:  Trametinib:GSK2879552\n",
      "Attempting to call pubchem API for chem_name:  Trametinib:OTX-015\n",
      "Attempting to call pubchem API for chem_name:  BLZ945\n",
      "Attempting to call pubchem API for chem_name:  Baiclein\n",
      "Attempting to call pubchem API for chem_name:  Etomoxir\n",
      "Attempting to call pubchem API for chem_name:  Metformin\n",
      "Attempting to call pubchem API for chem_name:  NVP-AEW541\n",
      "Attempting to call pubchem API for chem_name:  PH-797804\n",
      "Attempting to call pubchem API for chem_name:  Perhexiline maleate\n",
      "Attempting to call pubchem API for chem_name:  Ralimetinib\n",
      "Attempting to call pubchem API for chem_name:  Ranolazine\n",
      "Attempting to call pubchem API for chem_name:  Venetoclax:PH797804\n",
      "Attempting to call pubchem API for chem_name:  XMD 8-87\n",
      "Attempting to call pubchem API for chem_name:  BMS-754807\n",
      "Attempting to call pubchem API for other_name:  GS-9973\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[WARNING] /var/folders/wt/_yr7rg_13t76sq_q5cw57hxw0000gn/T/ipykernel_84914/3316055387.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  drug_res.rename(columns={\"inchikey\": \"inCHIKey\"}, inplace=True)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "drug_path = \"drug_response.csv\"\n",
    "drug_map_path = \"drugs_map.tsv.gz\"\n",
    "drug_url = \"https://raw.githubusercontent.com/PNNL-CompBio/candleDataProcessing/main/cell_line/drugs_by_structure.tsv.gz\"\n",
    "\n",
    "download_from_github(drug_url, drug_map_path)\n",
    "drug_map = format_drug_map(drug_map_path)\n",
    "d_df = format_drug_df(drug_path)\n",
    "d_df = update_dataframe_with_pubchem(d_df)\n",
    "d_res = merge_drug_info(d_df, drug_map)\n",
    "d_res = add_improve_id(drug_map, d_res)\n",
    "\n",
    "#split df to drug\n",
    "drug_res = d_res[[\"improve_drug_id\",\"chem_name\",\"formula\",\"weight\",\"inchikey\",\"canSMILES\",\"isoSMILES\"]]\n",
    "drug_res.rename(columns={\"inchikey\": \"inCHIKey\"}, inplace=True)\n",
    "\n",
    "# drug_res.to_csv(\"drugs.tsv\",sep=\"\\t\")\n",
    "\n",
    "# split df to experiment\n",
    "exp_res = d_res[[\"sample_id\",\"improve_drug_id\",\"auc\"]]\n",
    "exp_res = map_exp_to_improve(exp_res,improve_map_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d4a02693",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>improve_drug_id</th>\n",
       "      <th>chem_name</th>\n",
       "      <th>formula</th>\n",
       "      <th>weight</th>\n",
       "      <th>inCHIKey</th>\n",
       "      <th>canSMILES</th>\n",
       "      <th>isoSMILES</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SMI_56919</td>\n",
       "      <td>AKT Inhibitor IV</td>\n",
       "      <td>C31H27IN4S</td>\n",
       "      <td>614.5</td>\n",
       "      <td>NAYRELMNTQSBIN-UHFFFAOYSA-M</td>\n",
       "      <td>CC[N+]1=C(N(C2=C1C=C(C=C2)C3=NC4=CC=CC=C4S3)C5...</td>\n",
       "      <td>CC[N+]1=C(N(C2=C1C=C(C=C2)C3=NC4=CC=CC=C4S3)C5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SMI_56921</td>\n",
       "      <td>AKT Inhibitor X</td>\n",
       "      <td>C20H26Cl2N2O</td>\n",
       "      <td>381.3</td>\n",
       "      <td>SVKSJUIYYCQZEC-UHFFFAOYSA-N</td>\n",
       "      <td>CCN(CC)CCCCN1C2=CC=CC=C2OC3=C1C=C(C=C3)Cl.Cl</td>\n",
       "      <td>CCN(CC)CCCCN1C2=CC=CC=C2OC3=C1C=C(C=C3)Cl.Cl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SMI_56939</td>\n",
       "      <td>AMPK Inhibitor</td>\n",
       "      <td>C24H25N5O</td>\n",
       "      <td>399.5</td>\n",
       "      <td>XHBVYDAKJHETMP-UHFFFAOYSA-N</td>\n",
       "      <td>C1CCN(CC1)CCOC2=CC=C(C=C2)C3=CN4C(=C(C=N4)C5=C...</td>\n",
       "      <td>C1CCN(CC1)CCOC2=CC=C(C=C2)C3=CN4C(=C(C=N4)C5=C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SMI_56922</td>\n",
       "      <td>AST-487</td>\n",
       "      <td>C26H30F3N7O2</td>\n",
       "      <td>529.6</td>\n",
       "      <td>ODPGGGTTYSGTGO-UHFFFAOYSA-N</td>\n",
       "      <td>CCN1CCN(CC1)CC2=C(C=C(C=C2)NC(=O)NC3=CC=C(C=C3...</td>\n",
       "      <td>CCN1CCN(CC1)CC2=C(C=C(C=C2)NC(=O)NC3=CC=C(C=C3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SMI_49320</td>\n",
       "      <td>AZD1152-HQPA</td>\n",
       "      <td>C26H30FN7O3</td>\n",
       "      <td>507.6</td>\n",
       "      <td>QYZOGCMHVIGURT-UHFFFAOYSA-N</td>\n",
       "      <td>CCN(CCCOC1=CC2=C(C=C1)C(=NC=N2)NC3=NNC(=C3)CC(...</td>\n",
       "      <td>CCN(CCCOC1=CC2=C(C=C1)C(=NC=N2)NC3=NNC(=C3)CC(...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27109</th>\n",
       "      <td>SMI_48842</td>\n",
       "      <td>Vismodegib</td>\n",
       "      <td>C19H14Cl2N2O3S</td>\n",
       "      <td>421.3</td>\n",
       "      <td>BPQMGSKTAYIVFO-UHFFFAOYSA-N</td>\n",
       "      <td>CS(=O)(=O)C1=CC(=C(C=C1)C(=O)NC2=CC(=C(C=C2)Cl...</td>\n",
       "      <td>CS(=O)(=O)C1=CC(=C(C=C1)C(=O)NC2=CC(=C(C=C2)Cl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27110</th>\n",
       "      <td>SMI_49301</td>\n",
       "      <td>Volasertib</td>\n",
       "      <td>C34H50N8O3</td>\n",
       "      <td>618.8</td>\n",
       "      <td>SXNJFOWDRLKDSF-XKHVUIRMSA-N</td>\n",
       "      <td>CCC1C(=O)N(C2=CN=C(N=C2N1C(C)C)NC3=C(C=C(C=C3)...</td>\n",
       "      <td>CC[C@@H]1C(=O)N(C2=CN=C(N=C2N1C(C)C)NC3=C(C=C(...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27111</th>\n",
       "      <td>SMI_48825</td>\n",
       "      <td>XAV-939</td>\n",
       "      <td>C14H11F3N2OS</td>\n",
       "      <td>312.31</td>\n",
       "      <td>KLGQSVMIPOVQAX-UHFFFAOYSA-N</td>\n",
       "      <td>C1CSCC2=C1N=C(NC2=O)C3=CC=C(C=C3)C(F)(F)F</td>\n",
       "      <td>C1CSCC2=C1N=C(NC2=O)C3=CC=C(C=C3)C(F)(F)F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27112</th>\n",
       "      <td>SMI_56928</td>\n",
       "      <td>XMD 8-87</td>\n",
       "      <td>C24H27N7O2</td>\n",
       "      <td>445.5</td>\n",
       "      <td>LGLHCXISMKHLIK-UHFFFAOYSA-N</td>\n",
       "      <td>CN1CCN(CC1)C2=CC(=C(C=C2)NC3=NC=C4C(=N3)N(C5=C...</td>\n",
       "      <td>CN1CCN(CC1)C2=CC(=C(C=C2)NC3=NC=C4C(=N3)N(C5=C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27113</th>\n",
       "      <td>SMI_54566</td>\n",
       "      <td>YM-155</td>\n",
       "      <td>C20H19BrN4O3</td>\n",
       "      <td>443.3</td>\n",
       "      <td>QBIYUDDJPRGKNJ-UHFFFAOYSA-M</td>\n",
       "      <td>CC1=[N+](C2=C(N1CCOC)C(=O)C3=CC=CC=C3C2=O)CC4=...</td>\n",
       "      <td>CC1=[N+](C2=C(N1CCOC)C(=O)C3=CC=CC=C3C2=O)CC4=...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27114 rows  7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      improve_drug_id         chem_name         formula  weight  \\\n",
       "0           SMI_56919  AKT Inhibitor IV      C31H27IN4S   614.5   \n",
       "1           SMI_56921   AKT Inhibitor X    C20H26Cl2N2O   381.3   \n",
       "2           SMI_56939    AMPK Inhibitor       C24H25N5O   399.5   \n",
       "3           SMI_56922           AST-487    C26H30F3N7O2   529.6   \n",
       "4           SMI_49320      AZD1152-HQPA     C26H30FN7O3   507.6   \n",
       "...               ...               ...             ...     ...   \n",
       "27109       SMI_48842        Vismodegib  C19H14Cl2N2O3S   421.3   \n",
       "27110       SMI_49301        Volasertib      C34H50N8O3   618.8   \n",
       "27111       SMI_48825           XAV-939    C14H11F3N2OS  312.31   \n",
       "27112       SMI_56928          XMD 8-87      C24H27N7O2   445.5   \n",
       "27113       SMI_54566            YM-155    C20H19BrN4O3   443.3   \n",
       "\n",
       "                          inCHIKey  \\\n",
       "0      NAYRELMNTQSBIN-UHFFFAOYSA-M   \n",
       "1      SVKSJUIYYCQZEC-UHFFFAOYSA-N   \n",
       "2      XHBVYDAKJHETMP-UHFFFAOYSA-N   \n",
       "3      ODPGGGTTYSGTGO-UHFFFAOYSA-N   \n",
       "4      QYZOGCMHVIGURT-UHFFFAOYSA-N   \n",
       "...                            ...   \n",
       "27109  BPQMGSKTAYIVFO-UHFFFAOYSA-N   \n",
       "27110  SXNJFOWDRLKDSF-XKHVUIRMSA-N   \n",
       "27111  KLGQSVMIPOVQAX-UHFFFAOYSA-N   \n",
       "27112  LGLHCXISMKHLIK-UHFFFAOYSA-N   \n",
       "27113  QBIYUDDJPRGKNJ-UHFFFAOYSA-M   \n",
       "\n",
       "                                               canSMILES  \\\n",
       "0      CC[N+]1=C(N(C2=C1C=C(C=C2)C3=NC4=CC=CC=C4S3)C5...   \n",
       "1           CCN(CC)CCCCN1C2=CC=CC=C2OC3=C1C=C(C=C3)Cl.Cl   \n",
       "2      C1CCN(CC1)CCOC2=CC=C(C=C2)C3=CN4C(=C(C=N4)C5=C...   \n",
       "3      CCN1CCN(CC1)CC2=C(C=C(C=C2)NC(=O)NC3=CC=C(C=C3...   \n",
       "4      CCN(CCCOC1=CC2=C(C=C1)C(=NC=N2)NC3=NNC(=C3)CC(...   \n",
       "...                                                  ...   \n",
       "27109  CS(=O)(=O)C1=CC(=C(C=C1)C(=O)NC2=CC(=C(C=C2)Cl...   \n",
       "27110  CCC1C(=O)N(C2=CN=C(N=C2N1C(C)C)NC3=C(C=C(C=C3)...   \n",
       "27111          C1CSCC2=C1N=C(NC2=O)C3=CC=C(C=C3)C(F)(F)F   \n",
       "27112  CN1CCN(CC1)C2=CC(=C(C=C2)NC3=NC=C4C(=N3)N(C5=C...   \n",
       "27113  CC1=[N+](C2=C(N1CCOC)C(=O)C3=CC=CC=C3C2=O)CC4=...   \n",
       "\n",
       "                                               isoSMILES  \n",
       "0      CC[N+]1=C(N(C2=C1C=C(C=C2)C3=NC4=CC=CC=C4S3)C5...  \n",
       "1           CCN(CC)CCCCN1C2=CC=CC=C2OC3=C1C=C(C=C3)Cl.Cl  \n",
       "2      C1CCN(CC1)CCOC2=CC=C(C=C2)C3=CN4C(=C(C=N4)C5=C...  \n",
       "3      CCN1CCN(CC1)CC2=C(C=C(C=C2)NC(=O)NC3=CC=C(C=C3...  \n",
       "4      CCN(CCCOC1=CC2=C(C=C1)C(=NC=N2)NC3=NNC(=C3)CC(...  \n",
       "...                                                  ...  \n",
       "27109  CS(=O)(=O)C1=CC(=C(C=C1)C(=O)NC2=CC(=C(C=C2)Cl...  \n",
       "27110  CC[C@@H]1C(=O)N(C2=CN=C(N=C2N1C(C)C)NC3=C(C=C(...  \n",
       "27111          C1CSCC2=C1N=C(NC2=O)C3=CC=C(C=C3)C(F)(F)F  \n",
       "27112  CN1CCN(CC1)C2=CC(=C(C=C2)NC3=NC=C4C(=N3)N(C5=C...  \n",
       "27113  CC1=[N+](C2=C(N1CCOC)C(=O)C3=CC=CC=C3C2=O)CC4=...  \n",
       "\n",
       "[27114 rows x 7 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b941f202",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>improve_sample_id</th>\n",
       "      <th>improve_drug_id</th>\n",
       "      <th>auc</th>\n",
       "      <th>source</th>\n",
       "      <th>study</th>\n",
       "      <th>ic50</th>\n",
       "      <th>ec50</th>\n",
       "      <th>ec50se</th>\n",
       "      <th>einf</th>\n",
       "      <th>hs</th>\n",
       "      <th>aac1</th>\n",
       "      <th>auc1</th>\n",
       "      <th>dss1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>111374.0</td>\n",
       "      <td>SMI_56919</td>\n",
       "      <td>109.414835</td>\n",
       "      <td>synapse</td>\n",
       "      <td>BeatAML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>111374.0</td>\n",
       "      <td>SMI_56921</td>\n",
       "      <td>183.984149</td>\n",
       "      <td>synapse</td>\n",
       "      <td>BeatAML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>111374.0</td>\n",
       "      <td>SMI_56939</td>\n",
       "      <td>183.478483</td>\n",
       "      <td>synapse</td>\n",
       "      <td>BeatAML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>111374.0</td>\n",
       "      <td>SMI_56922</td>\n",
       "      <td>242.784602</td>\n",
       "      <td>synapse</td>\n",
       "      <td>BeatAML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>111374.0</td>\n",
       "      <td>SMI_49320</td>\n",
       "      <td>188.768480</td>\n",
       "      <td>synapse</td>\n",
       "      <td>BeatAML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27109</th>\n",
       "      <td>111583.0</td>\n",
       "      <td>SMI_48842</td>\n",
       "      <td>243.001338</td>\n",
       "      <td>synapse</td>\n",
       "      <td>BeatAML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27110</th>\n",
       "      <td>111583.0</td>\n",
       "      <td>SMI_49301</td>\n",
       "      <td>156.133267</td>\n",
       "      <td>synapse</td>\n",
       "      <td>BeatAML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27111</th>\n",
       "      <td>111583.0</td>\n",
       "      <td>SMI_48825</td>\n",
       "      <td>252.977539</td>\n",
       "      <td>synapse</td>\n",
       "      <td>BeatAML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27112</th>\n",
       "      <td>111583.0</td>\n",
       "      <td>SMI_56928</td>\n",
       "      <td>165.025345</td>\n",
       "      <td>synapse</td>\n",
       "      <td>BeatAML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27113</th>\n",
       "      <td>111583.0</td>\n",
       "      <td>SMI_54566</td>\n",
       "      <td>211.998433</td>\n",
       "      <td>synapse</td>\n",
       "      <td>BeatAML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27114 rows  13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       improve_sample_id improve_drug_id         auc   source    study  ic50  \\\n",
       "0               111374.0       SMI_56919  109.414835  synapse  BeatAML   NaN   \n",
       "1               111374.0       SMI_56921  183.984149  synapse  BeatAML   NaN   \n",
       "2               111374.0       SMI_56939  183.478483  synapse  BeatAML   NaN   \n",
       "3               111374.0       SMI_56922  242.784602  synapse  BeatAML   NaN   \n",
       "4               111374.0       SMI_49320  188.768480  synapse  BeatAML   NaN   \n",
       "...                  ...             ...         ...      ...      ...   ...   \n",
       "27109           111583.0       SMI_48842  243.001338  synapse  BeatAML   NaN   \n",
       "27110           111583.0       SMI_49301  156.133267  synapse  BeatAML   NaN   \n",
       "27111           111583.0       SMI_48825  252.977539  synapse  BeatAML   NaN   \n",
       "27112           111583.0       SMI_56928  165.025345  synapse  BeatAML   NaN   \n",
       "27113           111583.0       SMI_54566  211.998433  synapse  BeatAML   NaN   \n",
       "\n",
       "       ec50  ec50se  einf  hs  aac1  auc1  dss1  \n",
       "0       NaN     NaN   NaN NaN   NaN   NaN   NaN  \n",
       "1       NaN     NaN   NaN NaN   NaN   NaN   NaN  \n",
       "2       NaN     NaN   NaN NaN   NaN   NaN   NaN  \n",
       "3       NaN     NaN   NaN NaN   NaN   NaN   NaN  \n",
       "4       NaN     NaN   NaN NaN   NaN   NaN   NaN  \n",
       "...     ...     ...   ...  ..   ...   ...   ...  \n",
       "27109   NaN     NaN   NaN NaN   NaN   NaN   NaN  \n",
       "27110   NaN     NaN   NaN NaN   NaN   NaN   NaN  \n",
       "27111   NaN     NaN   NaN NaN   NaN   NaN   NaN  \n",
       "27112   NaN     NaN   NaN NaN   NaN   NaN   NaN  \n",
       "27113   NaN     NaN   NaN NaN   NaN   NaN   NaN  \n",
       "\n",
       "[27114 rows x 13 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f364b8c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Experiment Data\n",
    "# auc\n",
    "# improve_sample_id\n",
    "# improve drug id\n",
    "# source\n",
    "# study\n",
    "# empty:\n",
    "# ic50\n",
    "# ec50\n",
    "# ec50se\n",
    "# einf\n",
    "# hs\n",
    "# aac1\n",
    "# auc1\n",
    "# dss1\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9ca346c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fa551f3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da2ceba3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "7dbc7221",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # All Non mapped drugs\n",
    "# na_chem_names = result_df[result_df['improve_drug_id'].isna()]['chem_name'].unique().tolist()\n",
    "# print(len(na_chem_names))\n",
    "# print(na_chem_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "17a7bb68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Non mapped drugs excluding interactions\n",
    "# print(len([item for item in na_chem_names if ':' not in item]))\n",
    "# print([item for item in na_chem_names if ':' not in item])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "15e6f6fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # All Non mapped drugs - better view\n",
    "# pd.set_option('display.max_rows', 320)\n",
    "# pd.set_option('display.min_rows', 320)\n",
    "# result_df[[\"chem_name\", \"isoSMILES\",\"improve_drug_id\"]].drop_duplicates()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a6ad96bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pd.set_option('display.max_rows', 20)\n",
    "# pd.set_option('display.min_rows', 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a5b7390",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ae3cf05",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4e541bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15b2cd87",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "6c4f5c18",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PNNL_clinical_summary_12_08_2021_updated_OS_4patients_02_28_2022.txt is messy\n",
    "#This is needed to remove a random quote from the 110th line. And to fix the encoding. \n",
    "\n",
    "# def remove_sixth_from_last_quote(s):\n",
    "#     reversed_s = s[::-1]  # Reverse the string\n",
    "#     count = 0\n",
    "#     index_to_remove = None\n",
    "\n",
    "#     for i, char in enumerate(reversed_s):\n",
    "#         if char == '\"':\n",
    "#             count += 1\n",
    "#         if count == 6:\n",
    "#             index_to_remove = i\n",
    "#             break\n",
    "\n",
    "#     # Remove the 6th occurrence of the quote character\n",
    "#     if index_to_remove is not None:\n",
    "#         reversed_s = reversed_s[:index_to_remove] + reversed_s[index_to_remove+1:]\n",
    "\n",
    "#     return reversed_s[::-1]  # Reverse again to obtain the original order\n",
    "\n",
    "# def modify_file(file_path):\n",
    "#     with open(file_path, 'r', encoding='ISO-8859-1') as f:\n",
    "#         lines = f.readlines()\n",
    "\n",
    "#     # Modify the 110th line (0-based index is 109)\n",
    "#     lines[109] = remove_sixth_from_last_quote(lines[109])\n",
    "\n",
    "#     with open(file_path, 'w', encoding='utf-8') as f:\n",
    "#         f.writelines(lines)\n",
    "\n",
    "# file_path = 'PNNL_clinical_summary_12_08_2021_updated_OS_4patients_02_28_2022.txt'\n",
    "# modify_file(file_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c19bca9a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "8594a866",
   "metadata": {},
   "outputs": [],
   "source": [
    "# samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df1146c3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
