<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>API Reference &#8212; CoderData</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=d75fae25" />
    <link rel="stylesheet" type="text/css" href="_static/sphinxdoc.css?v=34905f61" />
    <link rel="stylesheet" type="text/css" href="_static/custom.css?v=08cfd9eb" />
    <script src="_static/documentation_options.js?v=8d563738"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="canonical" href="https://pnnl-compbio.github.io/coderdata/APIreference.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Datasets Included" href="datasets_included.html" />
    <link rel="prev" title="Usage" href="usage.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="Related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="datasets_included.html" title="Datasets Included"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="usage.html" title="Usage"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">CoderData</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">API Reference</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="api-reference">
<h1>API Reference<a class="headerlink" href="#api-reference" title="Link to this heading">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
<section id="module-coderdata.download.downloader">
<span id="coderdata-object"></span><h2>CoderData Object<a class="headerlink" href="#module-coderdata.download.downloader" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="coderdata.download.downloader.download">
<span class="sig-prename descclassname"><span class="pre">coderdata.download.downloader.</span></span><span class="sig-name descname"><span class="pre">download</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">local_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">PathLike</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">PosixPath('/home/runner/work/coderdata/coderdata')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exist_ok</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coderdata.download.downloader.download" title="Link to this definition">¶</a></dt>
<dd><p>Download the most recent version of files from a Figshare dataset,
filtered by a specific prefix or all files.</p>
<p>This function queries the Figshare API to retrieve details of a
dataset and then downloads files from it. Files can be filtered by a
specified prefix such as hcmi, beataml, etc. If ‘all’, an empty
string, or None is passed as the prefix, all files in the dataset
are downloaded. The function identifies the most recent version of a
file by selecting the one with the highest ID among duplicates with
the same name.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dataset_prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – The prefix of the dataset to download (e.g., ‘hcmi’). If ‘all’,
an empty string, or None, all files in the dataset are
downloaded. Default is None.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The function downloads files to the local repository and does
not return any value.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<p id="module-coderdata.utils.utils">Collection of small utility and helper functions.</p>
<dl class="py function">
<dt class="sig sig-object py" id="coderdata.utils.utils.list_datasets">
<span class="sig-prename descclassname"><span class="pre">coderdata.utils.utils.</span></span><span class="sig-name descname"><span class="pre">list_datasets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="headerlink" href="#coderdata.utils.utils.list_datasets" title="Link to this definition">¶</a></dt>
<dd><p>Hepler function that returns a list of available datasets including
a short description and additional information available.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>raw</strong> (<em>bool</em><em>, </em><em>default=False</em>) – If set to True returns a yaml dictionary containing all
available datasets including additional information. If set to
false prints information to stdout and returns None.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Returns a dict containing the information if <code class="docutils literal notranslate"><span class="pre">raw==True</span></code>,
otherwise prints information to stdout and returns <cite>None</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict | None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coderdata.utils.utils.version">
<span class="sig-prename descclassname"><span class="pre">coderdata.utils.utils.</span></span><span class="sig-name descname"><span class="pre">version</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="headerlink" href="#coderdata.utils.utils.version" title="Link to this definition">¶</a></dt>
<dd><p>Helper function that returns the version strings for the package and
the dataset build.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Contains package and dataset build version.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
</dd></dl>

<p id="module-coderdata.cli">Command Line Interface to retrieve coderdata datasets.</p>
<dl class="py function">
<dt class="sig sig-object py" id="coderdata.cli.check_folder">
<span class="sig-prename descclassname"><span class="pre">coderdata.cli.</span></span><span class="sig-name descname"><span class="pre">check_folder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">PathLike</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Path</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Path</span></span></span><a class="headerlink" href="#coderdata.cli.check_folder" title="Link to this definition">¶</a></dt>
<dd><p>Helper function to check if a defined folder exists.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Cleaned path object with the absolute path to the folder passed
to the function.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Path</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>TypeError</strong> – If passed path argument is not of the requested type.</p></li>
<li><p><strong>OSError</strong> – If the passed path argument does not link to a valid existing
    folder.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coderdata.cli.info">
<span class="sig-prename descclassname"><span class="pre">coderdata.cli.</span></span><span class="sig-name descname"><span class="pre">info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coderdata.cli.info" title="Link to this definition">¶</a></dt>
<dd><p>Helper function that takes the parsed command line arguments and
prints either verison information or information on the available
datasets depending on the arguments in <code class="docutils literal notranslate"><span class="pre">args</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>args</strong> (<em>Namespace</em>) – A Namespace object that contains commandline arguments parsed by
<code class="docutils literal notranslate"><span class="pre">ArgumentParser.parse_args()</span></code>.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="dataset-object">
<h2>Dataset Object<a class="headerlink" href="#dataset-object" title="Link to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="coderdata.dataset.dataset.Dataset">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">coderdata.dataset.dataset.</span></span><span class="sig-name descname"><span class="pre">Dataset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transcriptomics</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proteomics</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy_number</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drugs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drug_descriptors</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mirna</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">experiments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">methylation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metabolomics</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combinations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#coderdata.dataset.dataset.Dataset" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coderdata.dataset.dataset.Dataset.save">
<span class="sig-prename descclassname"><span class="pre">coderdata.dataset.dataset.Dataset.</span></span><span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#coderdata.dataset.dataset.Dataset.save" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<p id="module-coderdata.dataset.dataset">_summary_</p>
<dl class="py function">
<dt class="sig sig-object py" id="coderdata.dataset.dataset.load">
<span class="sig-prename descclassname"><span class="pre">coderdata.dataset.dataset.</span></span><span class="sig-name descname"><span class="pre">load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">local_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Path</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">PosixPath('/home/runner/work/coderdata/coderdata')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_pickle</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#coderdata.dataset.dataset.Dataset" title="coderdata.dataset.dataset.Dataset"><span class="pre">Dataset</span></a></span></span><a class="headerlink" href="#coderdata.dataset.dataset.load" title="Link to this definition">¶</a></dt>
<dd><p>_summary_</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – _description_</p></li>
<li><p><strong>directory</strong> (<em>str</em><em> | </em><em>Path</em><em>, </em><em>optional</em>) – _description_, by default Path.cwd()</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>_description_</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#coderdata.dataset.dataset.Dataset" title="coderdata.dataset.dataset.Dataset">Dataset</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>OSError</strong> – _description_</p></li>
<li><p><strong>TypeError</strong> – _description_</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coderdata.dataset.dataset.train_test_validate">
<span class="sig-prename descclassname"><span class="pre">coderdata.dataset.dataset.</span></span><span class="sig-name descname"><span class="pre">train_test_validate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#coderdata.dataset.dataset.Dataset" title="coderdata.dataset.dataset.Dataset"><span class="pre">Dataset</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">split_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'mixed-set'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'drug-blind'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'cancer-blind'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'mixed-set'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ratio</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(8,</span> <span class="pre">1,</span> <span class="pre">1)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stratify_by</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">balance</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">RandomState</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Split</span></span></span><a class="headerlink" href="#coderdata.dataset.dataset.train_test_validate" title="Link to this definition">¶</a></dt>
<dd><p>Splits a <cite>CoderData</cite> object (see also
<cite>coderdata.load.loader.DatasetLoader</cite>) into three subsets for
training, testing and validating machine learning algorithms.</p>
<p>The size of the splits can be adjusted to be different from 80:10:10
(the default)for train:test:validate. The function also allows for
additional optional arguments, that define the type of split that is
performed (‘mixed-set’, ‘drug-blind’, ‘cancer-blind’), if the splits
should be stratified (and which drug response metric to use), as
well as a random seed to enable the creation of reproducable splits.
Furhermore, a list of keyword arguments can be defined that will be
passed to the stratification function if so desired.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>DatasetLoader</em>) – CoderData object containing a full dataset either downloaded
from the CoderData repository (see also
<cite>coderdata.download.downloader.download_data_by_prefix</cite>) or
built locally via the <cite>build_all</cite> process. The object must first
be loaded via <cite>coderdata.load.loader.DatasetLoader</cite>.</p></li>
<li><p><strong>split_type</strong> (<em>{'mixed-set'</em><em>, </em><em>'drug-blind'</em><em>, </em><em>'cancer-blind'}</em><em>,         </em><em>default='mixed-set'</em>) – <p>Defines the type of split that should be generated:</p>
<ul>
<li><dl class="simple">
<dt><em>mixed-set</em>: Splits randomly independent of drug / cancer</dt><dd><p>association of the samples. Individual drugs or cancer types
can appear in all three splits</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><em>drug-blind</em>: Splits according to drug association. Any sample</dt><dd><p>associated with a drug will be unique to one of the splits.
For example samples with association to drug A will only be
present in the train split, but never in test or validate.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><em>cancer-blind</em>: Splits according to cancer association.</dt><dd><p>Equivalent to drug-blind, except cancer types will be unique
to splits.</p>
</dd>
</dl>
</li>
</ul>
</p></li>
<li><p><strong>ratio</strong> (<em>tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>, </em><em>int</em><em>]</em><em>, </em><em>default=</em><em>(</em><em>8</em><em>,</em><em>1</em><em>,</em><em>1</em><em>)</em>) – Defines the size ratio of the resulting test, train and
validation sets.</p></li>
<li><p><strong>stratify_by</strong> (<em>str</em><em> | </em><em>None</em><em>, </em><em>default=None</em>) – Defines if the training, testing and validation sets should be
stratified. Any value other than None indicates stratification
and defines which drug response value should be used as basis
for the stratification. _None_ indicates that no stratfication
should be performed.</p></li>
<li><p><strong>random_state</strong> (<em>int</em><em> | </em><em>RandomState</em><em> | </em><em>None</em><em>, </em><em>defaul=None</em>) – Defines a seed value for the randomization of the splits. Will
get passed to internal functions. Providing the seed will enable
reproducability of the generated splits.</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments that will be passed to the function
that generates classes for the stratification
(see also <code class="docutils literal notranslate"><span class="pre">_create_classes</span></code>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>Splits</strong> – A <code class="docutils literal notranslate"><span class="pre">Split</span></code> object that contains three Dataset objects as
attributes (<code class="docutils literal notranslate"><span class="pre">Split.train</span></code>, <code class="docutils literal notranslate"><span class="pre">Split.test</span></code>,
<code class="docutils literal notranslate"><span class="pre">Split.validate</span></code>)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Split</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError :</strong> – </p></li>
<li><p><strong>If supplied split_type is not in the list</strong><strong> of </strong><strong>accepted values.</strong> – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<p id="module-coderdata.utils.stats">Collection of helper scripts to generate general statistics on the data
contained in a CoderData Object.</p>
<dl class="py function">
<dt class="sig sig-object py" id="coderdata.utils.stats.plot_response_metric">
<span class="sig-prename descclassname"><span class="pre">coderdata.utils.stats.</span></span><span class="sig-name descname"><span class="pre">plot_response_metric</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#coderdata.dataset.dataset.Dataset" title="coderdata.dataset.dataset.Dataset"><span class="pre">Dataset</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'auc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Axes</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#coderdata.utils.stats.plot_response_metric" title="Link to this definition">¶</a></dt>
<dd><p>Creates a histogram detailing the distribution of dose response
values for a given dose respones metric.</p>
<p>If used in conjunction with <cite>matplotlib.pyplot.subplot</cite> or
<cite>matplotlib.pyplot.subplots</cite> and the axes object is passed to the
function, the function populates the axes object with the generated
plot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>coderdata.DataLoader</em>) – A full CoderData object of a dataset</p></li>
<li><p><strong>metric</strong> (<em>str</em><em>, </em><em>default='auc'</em>) – A string that defines the response metric that should be plotted</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>default=None</em>) – An <cite>Axes</cite> object can be defined. This is uesful if a multipannel
subplot has been defined prior via <cite>matplotlib.pyplot.subplots</cite>.
Passing the location of the axes to the function will then
populate the subplot at the given location with the generated
plot.</p></li>
<li><p><strong>**kwargs</strong> (<em>dict</em><em>, </em><em>optional</em>) – Additional keyword arguments that can be passed to the function
- bins : int - sets the number of bins; passed to
<cite>seaborn.histplot</cite>
- title : str - sets the title of the axes
- kde : bool - adds a kernel density estimate plot into the
histogram</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>In a Jupyter Notebook environment the following snippet can be used
to display a histgram detailing the distribution of drug response
AUC measures in the beataml dataset.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">coderdata</span> <span class="k">as</span> <span class="nn">cd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">beataml</span> <span class="o">=</span> <span class="n">cd</span><span class="o">.</span><span class="n">DataLoader</span><span class="p">(</span><span class="s1">&#39;beataml&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cd</span><span class="o">.</span><span class="n">plot_response_metric</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">beataml</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;auc&#39;</span><span class="p">,</span> <span class="nb">bin</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
<p>For generating multipanel plots we can make use of matplotlib and
the <cite>ax</cite> parameter of this function. Furthermore, other features /
parameters of the cerated figure can be changed (e.g. the title of
the figure via <cite>suptitle()</cite>). Finally it can be saved.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">coderdata</span> <span class="k">as</span> <span class="nn">cd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">beataml</span> <span class="o">=</span> <span class="n">cd</span><span class="o">.</span><span class="n">DataLoader</span><span class="p">(</span><span class="s1">&#39;beataml&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plot_response_metric</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">data</span><span class="o">=</span><span class="n">beataml</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;auc&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">bins</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">... </span>    <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plot_response_metric</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">data</span><span class="o">=</span><span class="n">beataml</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;aac&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">bins</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">... </span>    <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fig</span><span class="o">.</span><span class="n">set_layout_engine</span><span class="p">(</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;Distribution of drug response values&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;figure.png&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="coderdata.utils.stats.summarize_response_metric">
<span class="sig-prename descclassname"><span class="pre">coderdata.utils.stats.</span></span><span class="sig-name descname"><span class="pre">summarize_response_metric</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#coderdata.dataset.dataset.Dataset" title="coderdata.dataset.dataset.Dataset"><span class="pre">Dataset</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="headerlink" href="#coderdata.utils.stats.summarize_response_metric" title="Link to this definition">¶</a></dt>
<dd><p>Helper function to extract basic statistics for the <cite>experiments</cite>
object in a CoderData object. Uses <cite>pandas.DataFrame.describe()</cite>
internally to generate count, mean, standard deviation, minimum,
25-, 50- and 75-percentile as well as maximum for
<cite>dose_response_value</cite> for each <cite>dose_response_metric</cite> present in
<cite>experiments</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data</strong> (<em>coderdata.cd.Dataset</em>) – A full CoderData object of a dataset</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A <cite>pandas.DataFrame</cite> containing basic statistics for each
dose response metric.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>The Example assumes that a dataset with the prefix ‘beataml’ has
been downloaded previously. See also <code class="docutils literal notranslate"><span class="pre">coderdata.download()</span></code></p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">coderdata</span> <span class="k">as</span> <span class="nn">cd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">beataml</span> <span class="o">=</span> <span class="n">cd</span><span class="o">.</span><span class="n">DataLoader</span><span class="p">(</span><span class="s1">&#39;beataml&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">summary_stats</span> <span class="o">=</span> <span class="n">summarize_response_metric</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">beataml</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">summary_stats</span>
<span class="go">                        count          mean           std</span>
<span class="go">dose_response_metric</span>
<span class="go">aac                   23378.0  3.028061e-01  1.821265e-01  ...</span>
<span class="go">auc                   23378.0  6.971939e-01  1.821265e-01  ...</span>
<span class="go">dss                   23378.0  3.218484e-01  5.733492e-01  ...</span>
<span class="go">...                   ...      ...           ...           ...</span>
</pre></div>
</div>
</dd></dl>

</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/coderdata_logo3.jpg" alt="Logo of CoderData"/>
            </a></p><div  class="sphinxdoc-sidebar">
    <h3>Resource Links</h3>
    <ul>
        <li><a href="https://github.com/PNNL-CompBio/coderdata/" target="_blank">Github</a></li>
        <li><a href="https://pypi.org/project/coderdata/" target="_blank">PYPI</a></li>
        <li><a href="https://figshare.com/articles/dataset/CODERData2_1_0/28823159" target="_blank">FigShare</a></li>
    </ul>
</div>

<div class="sphinxdoc-sidebar">
    <h3>Quick Access</h3>
    <ul>
        <li><a href="index.html">Home</a></li>
        <li><a href="usage.html">Usage</a></li>
        <li><a href="#">API Reference</a></li>
        <li><a href="datasets_included.html">Datasets</a></li>
        <li><a href="_notebooks/docsite_tutorial.html">Tutorial</a></li>
        <li><a href="contribution_guide.html">Contributing</a></li>
    </ul>
</div>
  <div>
    <h3><a href="index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">API Reference</a><ul>
<li><a class="reference internal" href="#module-coderdata.download.downloader">CoderData Object</a><ul>
<li><a class="reference internal" href="#coderdata.download.downloader.download"><code class="docutils literal notranslate"><span class="pre">download()</span></code></a></li>
<li><a class="reference internal" href="#coderdata.utils.utils.list_datasets"><code class="docutils literal notranslate"><span class="pre">list_datasets()</span></code></a></li>
<li><a class="reference internal" href="#coderdata.utils.utils.version"><code class="docutils literal notranslate"><span class="pre">version()</span></code></a></li>
<li><a class="reference internal" href="#coderdata.cli.check_folder"><code class="docutils literal notranslate"><span class="pre">check_folder()</span></code></a></li>
<li><a class="reference internal" href="#coderdata.cli.info"><code class="docutils literal notranslate"><span class="pre">info()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#dataset-object">Dataset Object</a><ul>
<li><a class="reference internal" href="#coderdata.dataset.dataset.Dataset"><code class="docutils literal notranslate"><span class="pre">Dataset</span></code></a></li>
<li><a class="reference internal" href="#coderdata.dataset.dataset.Dataset.save"><code class="docutils literal notranslate"><span class="pre">save()</span></code></a></li>
<li><a class="reference internal" href="#coderdata.dataset.dataset.load"><code class="docutils literal notranslate"><span class="pre">load()</span></code></a></li>
<li><a class="reference internal" href="#coderdata.dataset.dataset.train_test_validate"><code class="docutils literal notranslate"><span class="pre">train_test_validate()</span></code></a></li>
<li><a class="reference internal" href="#coderdata.utils.stats.plot_response_metric"><code class="docutils literal notranslate"><span class="pre">plot_response_metric()</span></code></a></li>
<li><a class="reference internal" href="#coderdata.utils.stats.summarize_response_metric"><code class="docutils literal notranslate"><span class="pre">summarize_response_metric()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>

  </div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="Related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="datasets_included.html" title="Datasets Included"
             >next</a> |</li>
        <li class="right" >
          <a href="usage.html" title="Usage"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">CoderData</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">API Reference</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2025, Sara Gosline, Yannick Mahlich, Jeremy Jacobson, Marissa Andrews.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.1.3.
    </div>
  </body>
</html>