{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "11f00d32",
   "metadata": {},
   "source": [
    "# Dataset Documentation for Coderdata\n",
    "### Table of Contents:\n",
    "* Publication of Origin\n",
    "* Data Modalities Available\n",
    "* Data Transformations\n",
    "* Samples x Drugs combinations\n",
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41e754b1",
   "metadata": {},
   "source": [
    "## Publication of Origin\n",
    "BeatAML -  \n",
    "bladderpdo\n",
    "broad_sanger\n",
    "cptac\n",
    "crcpdo\n",
    "hcmi\n",
    "lincs\n",
    "liverpdo\n",
    "mpnst\n",
    "novartispdx\n",
    "pancpdo\n",
    "sarcpdo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "c11a3fc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from copy import deepcopy\n",
    "\n",
    "import coderdata as cd\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import os\n",
    "import glob\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "b4fdfec2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in all Datasets\n",
    "dataset_names = ['hcmi','beataml','mpnst','pancreatic','cptac','sarcoma','colorectal','bladder','liver','novartis']\n",
    "broad_sanger_datasets = [\"ccle\",\"ctrpv2\",\"fimm\",\"gdscv1\",\"gdscv2\",\"gcsi\",\"prism\",\"nci60\"]\n",
    "all_datasets = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fdcbc65",
   "metadata": {},
   "source": [
    "### Reading in # of Samples and Drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "519a7386",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1711\n",
      "1022\n",
      "50\n",
      "49\n",
      "1139\n",
      "36\n",
      "61\n",
      "134\n",
      "62\n",
      "386\n",
      "502\n",
      "846\n",
      "52\n",
      "984\n",
      "806\n",
      "569\n",
      "478\n",
      "83\n"
     ]
    }
   ],
   "source": [
    "# samples first\n",
    "for dataset in dataset_names + broad_sanger_datasets:\n",
    "    sample_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_samples.*\".format(dataset))\n",
    "    if sample_file:\n",
    "        samples_df = pd.read_csv(sample_file[0], sep=\"\\t\" if sample_file[0].endswith(\".tsv\") else \",\")\n",
    "        print(samples_df['improve_sample_id'].nunique())\n",
    "    else:\n",
    "        print(f\"No sample file found for dataset: {dataset}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "51776d2e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No drug file found for dataset: hcmi\n",
      "/Users/imal967/pnnl/projects/coderdata/all_datasets/beataml_drugs.tsv\n",
      "164\n",
      "/Users/imal967/pnnl/projects/coderdata/all_datasets/mpnst_drugs.tsv\n",
      "30\n",
      "/Users/imal967/pnnl/projects/coderdata/all_datasets/pancreatic_drugs.tsv\n",
      "25\n",
      "No drug file found for dataset: cptac\n",
      "/Users/imal967/pnnl/projects/coderdata/all_datasets/sarcoma_drugs.tsv\n",
      "34\n",
      "/Users/imal967/pnnl/projects/coderdata/all_datasets/colorectal_drugs.tsv\n",
      "10\n",
      "/Users/imal967/pnnl/projects/coderdata/all_datasets/bladder_drugs.tsv\n",
      "50\n",
      "/Users/imal967/pnnl/projects/coderdata/all_datasets/liver_drugs.tsv\n",
      "76\n",
      "/Users/imal967/pnnl/projects/coderdata/all_datasets/novartis_drugs.tsv\n",
      "25\n",
      "/Users/imal967/pnnl/projects/coderdata/all_datasets/ccle_drugs.tsv\n",
      "24\n",
      "/Users/imal967/pnnl/projects/coderdata/all_datasets/ctrpv2_drugs.tsv\n",
      "459\n",
      "/Users/imal967/pnnl/projects/coderdata/all_datasets/fimm_drugs.tsv\n",
      "52\n",
      "/Users/imal967/pnnl/projects/coderdata/all_datasets/gdscv1_drugs.tsv\n",
      "294\n",
      "/Users/imal967/pnnl/projects/coderdata/all_datasets/gdscv2_drugs.tsv\n",
      "171\n",
      "/Users/imal967/pnnl/projects/coderdata/all_datasets/gcsi_drugs.tsv\n",
      "44\n",
      "/Users/imal967/pnnl/projects/coderdata/all_datasets/prism_drugs.tsv\n",
      "1419\n",
      "/Users/imal967/pnnl/projects/coderdata/all_datasets/nci60_drugs.tsv\n",
      "55157\n"
     ]
    }
   ],
   "source": [
    "# drugs next\n",
    "for dataset in dataset_names + broad_sanger_datasets:\n",
    "    drug_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_drugs.*\".format(dataset))\n",
    "    if drug_file:\n",
    "        print(drug_file[0])\n",
    "        drugs_df = pd.read_csv(drug_file[0], sep=\"\\t\" if drug_file[0].endswith(\".tsv\") else \",\")\n",
    "        print(drugs_df['improve_drug_id'].nunique())\n",
    "    else:\n",
    "        print(f\"No drug file found for dataset: {dataset}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "3c1250ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No experimental data file found for dataset: hcmi\n",
      "31926\n",
      "272\n",
      "190\n",
      "No experimental data file found for dataset: cptac\n",
      "275\n",
      "140\n",
      "3300\n",
      "4453\n",
      "1766\n",
      "11543\n",
      "309401\n",
      "2663\n",
      "247753\n",
      "115440\n",
      "13398\n",
      "638983\n",
      "2960756\n"
     ]
    }
   ],
   "source": [
    "# now getting unique sample-drug pairs in the experimental data\n",
    "for dataset in dataset_names + broad_sanger_datasets:\n",
    "    exp_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_experiments.*\".format(dataset))\n",
    "    if exp_file:\n",
    "        exp_df = pd.read_csv(exp_file[0], sep=\"\\t\" if exp_file[0].endswith(\".tsv\") else \",\")\n",
    "        print(exp_df[['improve_sample_id','improve_drug_id']].drop_duplicates().shape[0])\n",
    "    else:\n",
    "        print(f\"No experimental data file found for dataset: {dataset}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "85d54e9e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No experimental data file found for dataset: hcmi\n",
      "       improve_sample_id improve_drug_id\n",
      "0                   2771       SMI_11173\n",
      "1                   2771       SMI_11676\n",
      "2                   2771       SMI_14986\n",
      "3                   2771       SMI_16677\n",
      "4                   2771       SMI_17300\n",
      "...                  ...             ...\n",
      "25192               2780        SMI_5869\n",
      "25193               2780        SMI_6359\n",
      "25194               2780        SMI_7087\n",
      "25195               2780        SMI_7916\n",
      "25196               2780        SMI_9857\n",
      "\n",
      "[4137 rows x 2 columns]\n",
      "4137\n",
      "      improve_sample_id improve_drug_id\n",
      "0                  3824       SMI_13126\n",
      "3                  3824       SMI_16729\n",
      "6                  3824       SMI_17006\n",
      "9                  3824       SMI_17091\n",
      "11                 3824       SMI_17210\n",
      "...                 ...             ...\n",
      "3000               3816       SMI_13126\n",
      "3004               3816        SMI_6182\n",
      "3008               3816       SMI_56074\n",
      "3012               3821       SMI_13126\n",
      "3016               3821        SMI_6182\n",
      "\n",
      "[193 rows x 2 columns]\n",
      "193\n",
      "     improve_sample_id improve_drug_id\n",
      "0                 3076        SMI_2497\n",
      "1                 3076       SMI_28024\n",
      "2                 3076       SMI_41675\n",
      "3                 3076       SMI_55963\n",
      "4                 3076       SMI_56276\n",
      "..                 ...             ...\n",
      "185               3124        SMI_2497\n",
      "186               3124       SMI_28024\n",
      "187               3124       SMI_41675\n",
      "188               3124       SMI_55963\n",
      "189               3124       SMI_56276\n",
      "\n",
      "[190 rows x 2 columns]\n",
      "190\n",
      "No experimental data file found for dataset: cptac\n",
      "     improve_sample_id improve_drug_id\n",
      "0                 3260       SMI_56377\n",
      "1                 3260       SMI_48031\n",
      "2                 3260       SMI_31997\n",
      "3                 3260       SMI_56565\n",
      "4                 3260       SMI_56276\n",
      "..                 ...             ...\n",
      "270               3290       SMI_56099\n",
      "271               3290       SMI_15489\n",
      "272               3290       SMI_56064\n",
      "273               3290       SMI_39219\n",
      "274               3290       SMI_56381\n",
      "\n",
      "[234 rows x 2 columns]\n",
      "234\n",
      "     improve_sample_id improve_drug_id\n",
      "20                3748       SMI_17123\n",
      "21                3748        SMI_2497\n",
      "22                3748       SMI_26850\n",
      "23                3748       SMI_33483\n",
      "24                3748       SMI_36067\n",
      "25                3748       SMI_39219\n",
      "26                3748       SMI_43849\n",
      "27                3748         SMI_456\n",
      "28                3748       SMI_48585\n",
      "29                3748       SMI_56276\n",
      "30                3754       SMI_17123\n",
      "31                3754        SMI_2497\n",
      "32                3754       SMI_26850\n",
      "33                3754       SMI_33483\n",
      "34                3754       SMI_36067\n",
      "35                3754       SMI_39219\n",
      "36                3754       SMI_43849\n",
      "37                3754         SMI_456\n",
      "38                3754       SMI_48585\n",
      "39                3754       SMI_56276\n",
      "40                3756       SMI_17123\n",
      "41                3756        SMI_2497\n",
      "42                3756       SMI_26850\n",
      "43                3756       SMI_33483\n",
      "44                3756       SMI_36067\n",
      "45                3756       SMI_39219\n",
      "46                3756       SMI_43849\n",
      "47                3756         SMI_456\n",
      "48                3756       SMI_48585\n",
      "49                3756       SMI_56276\n",
      "70                3768       SMI_17123\n",
      "71                3768        SMI_2497\n",
      "72                3768       SMI_26850\n",
      "73                3768       SMI_33483\n",
      "74                3768       SMI_36067\n",
      "75                3768       SMI_39219\n",
      "76                3768       SMI_43849\n",
      "77                3768         SMI_456\n",
      "78                3768       SMI_48585\n",
      "79                3768       SMI_56276\n",
      "90                3772       SMI_17123\n",
      "91                3772        SMI_2497\n",
      "92                3772       SMI_26850\n",
      "93                3772       SMI_33483\n",
      "94                3772       SMI_36067\n",
      "95                3772       SMI_39219\n",
      "96                3772       SMI_43849\n",
      "97                3772         SMI_456\n",
      "98                3772       SMI_48585\n",
      "99                3772       SMI_56276\n",
      "100               3774       SMI_17123\n",
      "101               3774        SMI_2497\n",
      "102               3774       SMI_26850\n",
      "103               3774       SMI_33483\n",
      "104               3774       SMI_36067\n",
      "105               3774       SMI_39219\n",
      "106               3774       SMI_43849\n",
      "107               3774         SMI_456\n",
      "108               3774       SMI_48585\n",
      "109               3774       SMI_56276\n",
      "60\n",
      "      improve_sample_id improve_drug_id\n",
      "0                  3125       SMI_16091\n",
      "1                  3125         SMI_285\n",
      "2                  3125       SMI_31538\n",
      "3                  3125       SMI_31997\n",
      "4                  3125       SMI_33483\n",
      "...                 ...             ...\n",
      "3065               3234       SMI_39219\n",
      "3066               3234       SMI_39590\n",
      "3067               3234         SMI_456\n",
      "3068               3234       SMI_56243\n",
      "3069               3234       SMI_56276\n",
      "\n",
      "[840 rows x 2 columns]\n",
      "840\n",
      "      improve_sample_id improve_drug_id\n",
      "0                  3295       SMI_10817\n",
      "1                  3295       SMI_11173\n",
      "2                  3295         SMI_117\n",
      "3                  3295       SMI_13126\n",
      "4                  3295       SMI_14986\n",
      "...                 ...             ...\n",
      "4448               3356       SMI_57357\n",
      "4449               3356        SMI_7087\n",
      "4450               3356        SMI_7916\n",
      "4451               3356         SMI_848\n",
      "4452               3356        SMI_9134\n",
      "\n",
      "[4453 rows x 2 columns]\n",
      "4453\n",
      "      improve_sample_id improve_drug_id\n",
      "0                3359.0       SMI_57366\n",
      "4                3359.0       SMI_57374\n",
      "8                3359.0       SMI_57372\n",
      "12               3359.0       SMI_57358\n",
      "16               3359.0       SMI_57376\n",
      "...                 ...             ...\n",
      "6930             3407.0       SMI_57373\n",
      "6934             3407.0       SMI_57360\n",
      "6938             3407.0       SMI_57363\n",
      "6942             3407.0       SMI_41675\n",
      "6946             3407.0       SMI_27773\n",
      "\n",
      "[1734 rows x 2 columns]\n",
      "1734\n",
      "       improve_sample_id improve_drug_id\n",
      "0                      1       SMI_15489\n",
      "1                      1       SMI_17091\n",
      "2                      1       SMI_17237\n",
      "3                      1        SMI_1964\n",
      "4                      1       SMI_22349\n",
      "...                  ...             ...\n",
      "11538               1929       SMI_48746\n",
      "11539               1929       SMI_56064\n",
      "11540               1929       SMI_56243\n",
      "11541               1929       SMI_57322\n",
      "11542               1929        SMI_5869\n",
      "\n",
      "[10887 rows x 2 columns]\n",
      "10887\n",
      "        improve_sample_id improve_drug_id\n",
      "0                       1       SMI_10160\n",
      "1                       1       SMI_10817\n",
      "2                       1       SMI_11308\n",
      "3                       1       SMI_13126\n",
      "4                       1       SMI_14081\n",
      "...                   ...             ...\n",
      "309396               1929        SMI_7916\n",
      "309397               1929         SMI_910\n",
      "309398               1929        SMI_9134\n",
      "309399               1929        SMI_9544\n",
      "309400               1929        SMI_9725\n",
      "\n",
      "[300507 rows x 2 columns]\n",
      "300507\n",
      "      improve_sample_id improve_drug_id\n",
      "0                    87       SMI_11173\n",
      "1                    87       SMI_13126\n",
      "2                    87       SMI_14986\n",
      "3                    87       SMI_15489\n",
      "4                    87       SMI_15902\n",
      "...                 ...             ...\n",
      "2658               1916       SMI_56457\n",
      "2659               1916       SMI_56477\n",
      "2660               1916        SMI_5869\n",
      "2661               1916        SMI_7087\n",
      "2662               1916        SMI_7781\n",
      "\n",
      "[2457 rows x 2 columns]\n",
      "2457\n",
      "        improve_sample_id improve_drug_id\n",
      "0                      11       SMI_10100\n",
      "1                      11       SMI_10817\n",
      "2                      11       SMI_11676\n",
      "3                      11       SMI_11932\n",
      "4                      11       SMI_12289\n",
      "...                   ...             ...\n",
      "247748               2022       SMI_56477\n",
      "247749               2022       SMI_56478\n",
      "247750               2022       SMI_56489\n",
      "247751               2022       SMI_56490\n",
      "247752               2022        SMI_6534\n",
      "\n",
      "[245220 rows x 2 columns]\n",
      "245220\n",
      "        improve_sample_id improve_drug_id\n",
      "0                      11       SMI_10729\n",
      "1                      11       SMI_10817\n",
      "2                      11       SMI_12168\n",
      "3                      11       SMI_13126\n",
      "4                      11       SMI_13386\n",
      "...                   ...             ...\n",
      "115435               1924        SMI_6677\n",
      "115436               1924        SMI_6982\n",
      "115437               1924        SMI_7916\n",
      "115438               1924        SMI_8838\n",
      "115439               1924        SMI_9857\n",
      "\n",
      "[114373 rows x 2 columns]\n",
      "114373\n",
      "       improve_sample_id improve_drug_id\n",
      "0                      1       SMI_18995\n",
      "1                      1        SMI_1964\n",
      "2                      1       SMI_20749\n",
      "3                      1       SMI_22736\n",
      "4                      1       SMI_26850\n",
      "...                  ...             ...\n",
      "13393               1924       SMI_56264\n",
      "13394               1924       SMI_56276\n",
      "13395               1924       SMI_56281\n",
      "13396               1924       SMI_56314\n",
      "13397               1924         SMI_910\n",
      "\n",
      "[12506 rows x 2 columns]\n",
      "12506\n",
      "        improve_sample_id improve_drug_id\n",
      "0                       1       SMI_10015\n",
      "1                       1       SMI_10160\n",
      "2                       1       SMI_10550\n",
      "3                       1       SMI_10572\n",
      "4                       1       SMI_10708\n",
      "...                   ...             ...\n",
      "638978               1929        SMI_9619\n",
      "638979               1929        SMI_9725\n",
      "638980               1929        SMI_9857\n",
      "638981               1929        SMI_9877\n",
      "638982               1929        SMI_9922\n",
      "\n",
      "[632078 rows x 2 columns]\n",
      "632078\n",
      "         improve_sample_id improve_drug_id\n",
      "0                       37       SMI_10550\n",
      "1                       37       SMI_11919\n",
      "2                       37       SMI_16729\n",
      "3                       37       SMI_17910\n",
      "4                       37       SMI_21347\n",
      "...                    ...             ...\n",
      "2969872               1916        SMI_9995\n",
      "2969873               1916        SMI_9996\n",
      "2969874               1916        SMI_9997\n",
      "2969875               1916        SMI_9998\n",
      "2969876               1916        SMI_9999\n",
      "\n",
      "[2329149 rows x 2 columns]\n",
      "2329149\n"
     ]
    }
   ],
   "source": [
    "# now doing rnaseq x drugs\n",
    "# going to need rnaseq data and experimental data\n",
    "for dataset in dataset_names + broad_sanger_datasets:\n",
    "    exp_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_experiments.*\".format(dataset))\n",
    "    transcriptomics_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_transcriptomics.*\".format(dataset))\n",
    "    if transcriptomics_file:\n",
    "        transcriptomics_df = pd.read_csv(transcriptomics_file[0], sep=\"\\t\" if transcriptomics_file[0].endswith(\".tsv\") else \",\")\n",
    "        all_transcriptomics_samples = transcriptomics_df['improve_sample_id'].unique() # unique samples with transcriptomics data\n",
    "    if exp_file:\n",
    "        exp_df = pd.read_csv(exp_file[0], sep=\"\\t\" if exp_file[0].endswith(\".tsv\") else \",\")\n",
    "        # now filter to only those samples with transcriptomics data\n",
    "        exp_transcriptomics_df = exp_df[exp_df['improve_sample_id'].isin(all_transcriptomics_samples)]\n",
    "        number_of_unique_sample_drug_pairs = exp_transcriptomics_df[['improve_sample_id','improve_drug_id']].drop_duplicates().shape[0]\n",
    "        print(exp_transcriptomics_df[['improve_sample_id','improve_drug_id']].drop_duplicates())\n",
    "        print(number_of_unique_sample_drug_pairs)\n",
    "    else:\n",
    "        print(f\"No experimental data file found for dataset: {dataset}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "8f97d5e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No experimental data file found for dataset: hcmi\n",
      "3958\n",
      "184\n",
      "185\n",
      "No experimental data file found for dataset: cptac\n",
      "187\n",
      "60\n",
      "640\n",
      "4453\n",
      "1734\n"
     ]
    }
   ],
   "source": [
    "# rnaseq x mutations x drugs \n",
    "# loading in mutation data\n",
    "for dataset in dataset_names:\n",
    "    mutation_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_mutations.*\".format(dataset))\n",
    "    if mutation_file:\n",
    "        mutations_df = pd.read_csv(mutation_file[0], sep=\"\\t\" if mutation_file[0].endswith(\".tsv\") else \",\")\n",
    "        all_mutation_samples = mutations_df['improve_sample_id'].unique() # unique samples with mutation data\n",
    "    exp_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_experiments.*\".format(dataset))\n",
    "    transcriptomics_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_transcriptomics.*\".format(dataset))\n",
    "    if transcriptomics_file:\n",
    "        transcriptomics_df = pd.read_csv(transcriptomics_file[0], sep=\"\\t\" if transcriptomics_file[0].endswith(\".tsv\") else \",\")\n",
    "        all_transcriptomics_samples = transcriptomics_df['improve_sample_id'].unique() # unique samples with transcriptomics data\n",
    "    if exp_file:  \n",
    "        exp_df = pd.read_csv(exp_file[0], sep=\"\\t\" if exp_file[0].endswith(\".tsv\") else \",\")\n",
    "        # now filter to only those samples with transcriptomics data and mutation data\n",
    "        exp_transcriptomics_mutation_df = exp_df[exp_df['improve_sample_id'].isin(all_transcriptomics_samples) & exp_df['improve_sample_id'].isin(all_mutation_samples)]\n",
    "        number_of_unique_sample_drug_pairs = exp_transcriptomics_mutation_df[['improve_sample_id','improve_drug_id']].drop_duplicates().shape[0]\n",
    "        print(number_of_unique_sample_drug_pairs)\n",
    "    else:\n",
    "        print(f\"No experimental data file found for dataset: {dataset}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "8cb33073",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['hcmi', nan, nan, nan, nan, nan]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan], ['mpnst', 272, 193, 184, 191, 184]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan], ['mpnst', 272, 193, 184, 191, 184], ['pancreatic', 190, 190, 185, 185, 185]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan], ['mpnst', 272, 193, 184, 191, 184], ['pancreatic', 190, 190, 185, 185, 185], ['cptac', nan, nan, nan, nan, nan]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan], ['mpnst', 272, 193, 184, 191, 184], ['pancreatic', 190, 190, 185, 185, 185], ['cptac', nan, nan, nan, nan, nan], ['sarcoma', 275, 234, 187, nan, nan]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan], ['mpnst', 272, 193, 184, 191, 184], ['pancreatic', 190, 190, 185, 185, 185], ['cptac', nan, nan, nan, nan, nan], ['sarcoma', 275, 234, 187, nan, nan], ['colorectal', 140, 60, 60, 60, 140]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan], ['mpnst', 272, 193, 184, 191, 184], ['pancreatic', 190, 190, 185, 185, 185], ['cptac', nan, nan, nan, nan, nan], ['sarcoma', 275, 234, 187, nan, nan], ['colorectal', 140, 60, 60, 60, 140], ['bladder', 3300, 840, 640, 640, 3100]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan], ['mpnst', 272, 193, 184, 191, 184], ['pancreatic', 190, 190, 185, 185, 185], ['cptac', nan, nan, nan, nan, nan], ['sarcoma', 275, 234, 187, nan, nan], ['colorectal', 140, 60, 60, 60, 140], ['bladder', 3300, 840, 640, 640, 3100], ['liver', 4453, 4453, 4453, 4453, 4453]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan], ['mpnst', 272, 193, 184, 191, 184], ['pancreatic', 190, 190, 185, 185, 185], ['cptac', nan, nan, nan, nan, nan], ['sarcoma', 275, 234, 187, nan, nan], ['colorectal', 140, 60, 60, 60, 140], ['bladder', 3300, 840, 640, 640, 3100], ['liver', 4453, 4453, 4453, 4453, 4453], ['novartis', 1766, 1734, 1734, 1723, 1723]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan], ['mpnst', 272, 193, 184, 191, 184], ['pancreatic', 190, 190, 185, 185, 185], ['cptac', nan, nan, nan, nan, nan], ['sarcoma', 275, 234, 187, nan, nan], ['colorectal', 140, 60, 60, 60, 140], ['bladder', 3300, 840, 640, 640, 3100], ['liver', 4453, 4453, 4453, 4453, 4453], ['novartis', 1766, 1734, 1734, 1723, 1723], ['ccle', 11543, 10887, 10792, 10887, 11118]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan], ['mpnst', 272, 193, 184, 191, 184], ['pancreatic', 190, 190, 185, 185, 185], ['cptac', nan, nan, nan, nan, nan], ['sarcoma', 275, 234, 187, nan, nan], ['colorectal', 140, 60, 60, 60, 140], ['bladder', 3300, 840, 640, 640, 3100], ['liver', 4453, 4453, 4453, 4453, 4453], ['novartis', 1766, 1734, 1734, 1723, 1723], ['ccle', 11543, 10887, 10792, 10887, 11118], ['ctrpv2', 309401, 300507, 295742, 299698, 300616]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan], ['mpnst', 272, 193, 184, 191, 184], ['pancreatic', 190, 190, 185, 185, 185], ['cptac', nan, nan, nan, nan, nan], ['sarcoma', 275, 234, 187, nan, nan], ['colorectal', 140, 60, 60, 60, 140], ['bladder', 3300, 840, 640, 640, 3100], ['liver', 4453, 4453, 4453, 4453, 4453], ['novartis', 1766, 1734, 1734, 1723, 1723], ['ccle', 11543, 10887, 10792, 10887, 11118], ['ctrpv2', 309401, 300507, 295742, 299698, 300616], ['fimm', 2663, 2457, 2457, 2457, 2611]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan], ['mpnst', 272, 193, 184, 191, 184], ['pancreatic', 190, 190, 185, 185, 185], ['cptac', nan, nan, nan, nan, nan], ['sarcoma', 275, 234, 187, nan, nan], ['colorectal', 140, 60, 60, 60, 140], ['bladder', 3300, 840, 640, 640, 3100], ['liver', 4453, 4453, 4453, 4453, 4453], ['novartis', 1766, 1734, 1734, 1723, 1723], ['ccle', 11543, 10887, 10792, 10887, 11118], ['ctrpv2', 309401, 300507, 295742, 299698, 300616], ['fimm', 2663, 2457, 2457, 2457, 2611], ['gdscv1', 247753, 245220, 241999, 241240, 242570]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan], ['mpnst', 272, 193, 184, 191, 184], ['pancreatic', 190, 190, 185, 185, 185], ['cptac', nan, nan, nan, nan, nan], ['sarcoma', 275, 234, 187, nan, nan], ['colorectal', 140, 60, 60, 60, 140], ['bladder', 3300, 840, 640, 640, 3100], ['liver', 4453, 4453, 4453, 4453, 4453], ['novartis', 1766, 1734, 1734, 1723, 1723], ['ccle', 11543, 10887, 10792, 10887, 11118], ['ctrpv2', 309401, 300507, 295742, 299698, 300616], ['fimm', 2663, 2457, 2457, 2457, 2611], ['gdscv1', 247753, 245220, 241999, 241240, 242570], ['gdscv2', 115440, 114373, 112829, 112523, 113133]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan], ['mpnst', 272, 193, 184, 191, 184], ['pancreatic', 190, 190, 185, 185, 185], ['cptac', nan, nan, nan, nan, nan], ['sarcoma', 275, 234, 187, nan, nan], ['colorectal', 140, 60, 60, 60, 140], ['bladder', 3300, 840, 640, 640, 3100], ['liver', 4453, 4453, 4453, 4453, 4453], ['novartis', 1766, 1734, 1734, 1723, 1723], ['ccle', 11543, 10887, 10792, 10887, 11118], ['ctrpv2', 309401, 300507, 295742, 299698, 300616], ['fimm', 2663, 2457, 2457, 2457, 2611], ['gdscv1', 247753, 245220, 241999, 241240, 242570], ['gdscv2', 115440, 114373, 112829, 112523, 113133], ['gcsi', 13398, 12506, 12338, 12506, 13112]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan], ['mpnst', 272, 193, 184, 191, 184], ['pancreatic', 190, 190, 185, 185, 185], ['cptac', nan, nan, nan, nan, nan], ['sarcoma', 275, 234, 187, nan, nan], ['colorectal', 140, 60, 60, 60, 140], ['bladder', 3300, 840, 640, 640, 3100], ['liver', 4453, 4453, 4453, 4453, 4453], ['novartis', 1766, 1734, 1734, 1723, 1723], ['ccle', 11543, 10887, 10792, 10887, 11118], ['ctrpv2', 309401, 300507, 295742, 299698, 300616], ['fimm', 2663, 2457, 2457, 2457, 2611], ['gdscv1', 247753, 245220, 241999, 241240, 242570], ['gdscv2', 115440, 114373, 112829, 112523, 113133], ['gcsi', 13398, 12506, 12338, 12506, 13112], ['prism', 638983, 632078, 630672, 632078, 636226]]\n",
      "[['hcmi', nan, nan, nan, nan, nan], ['beataml', 31926, 4137, 3958, nan, nan], ['mpnst', 272, 193, 184, 191, 184], ['pancreatic', 190, 190, 185, 185, 185], ['cptac', nan, nan, nan, nan, nan], ['sarcoma', 275, 234, 187, nan, nan], ['colorectal', 140, 60, 60, 60, 140], ['bladder', 3300, 840, 640, 640, 3100], ['liver', 4453, 4453, 4453, 4453, 4453], ['novartis', 1766, 1734, 1734, 1723, 1723], ['ccle', 11543, 10887, 10792, 10887, 11118], ['ctrpv2', 309401, 300507, 295742, 299698, 300616], ['fimm', 2663, 2457, 2457, 2457, 2611], ['gdscv1', 247753, 245220, 241999, 241240, 242570], ['gdscv2', 115440, 114373, 112829, 112523, 113133], ['gcsi', 13398, 12506, 12338, 12506, 13112], ['prism', 638983, 632078, 630672, 632078, 636226], ['nci60', 2960756, 2329149, 2329132, 2329149, 2784474]]\n"
     ]
    }
   ],
   "source": [
    "# adding to the dataframe\n",
    "to_add = []\n",
    "for dataset in dataset_names + broad_sanger_datasets:\n",
    "    sample_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_samples.*\".format(dataset))\n",
    "    drug_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_drugs.*\".format(dataset))\n",
    "    exp_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_experiments.*\".format(dataset))\n",
    "    transcriptomics_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_transcriptomics.*\".format(dataset))\n",
    "    mutation_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_mutations.*\".format(dataset))\n",
    "    copy_number_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_copy_number.*\".format(dataset)) \n",
    "    # getting number of samples\n",
    "    # if sample_file:\n",
    "    #     samples_df = pd.read_csv(sample_file[0], sep=\"\\t\" if sample_file[0].endswith(\".tsv\") else \",\")\n",
    "    #     num_samples = samples_df['improve_sample_id'].nunique() # unique samples\n",
    "    # else:\n",
    "    #     print(f\"Missing files for dataset: {dataset}\")\n",
    "    #     num_samples = np.nan\n",
    "    # # getting number of drugs\n",
    "    # if drug_file:\n",
    "    #     drugs_df = pd.read_csv(drug_file[0], sep=\"\\t\" if drug_file[0].endswith(\".tsv\") else \",\")\n",
    "    #     num_drugs = drugs_df['improve_drug_id'].nunique() # unique drugs\n",
    "    # else:\n",
    "    #     print(f\"Missing files for dataset: {dataset}\")\n",
    "    #     num_drugs = np.nan\n",
    "    # getting number of unique sample-drug pairs\n",
    "    if exp_file:\n",
    "        exp_df = pd.read_csv(exp_file[0], sep=\"\\t\" if exp_file[0].endswith(\".tsv\") else \",\")\n",
    "        num_sample_drug_pairs = exp_df[['improve_sample_id','improve_drug_id']].drop_duplicates().shape[0]\n",
    "    else:\n",
    "        num_sample_drug_pairs = np.nan\n",
    "    # getting number of unique sample-drug pairs with transcriptomics data\n",
    "    if transcriptomics_file and exp_file:\n",
    "        transcriptomics_df = pd.read_csv(transcriptomics_file[0], sep=\"\\t\" if transcriptomics_file[0].endswith(\".tsv\") else \",\")\n",
    "        all_transcriptomics_samples = transcriptomics_df['improve_sample_id'].unique() # unique samples with transcriptomics data\n",
    "        # now filter to only those samples with transcriptomics data\n",
    "        exp_transcriptomics_df = exp_df[exp_df['improve_sample_id'].isin(all_transcriptomics_samples)]\n",
    "        num_sample_drug_transcript_pairs = exp_transcriptomics_df[['improve_sample_id','improve_drug_id']].drop_duplicates().shape[0]\n",
    "    else:\n",
    "        num_sample_drug_transcript_pairs = np.nan\n",
    "    # unique sample-drug pairs with mutation data and transcriptomics data\n",
    "    if transcriptomics_file and mutation_file and exp_file:\n",
    "        mutations_df = pd.read_csv(mutation_file[0], sep=\"\\t\" if mutation_file[0].endswith(\".tsv\") else \",\")\n",
    "        all_mutation_samples = mutations_df['improve_sample_id'].unique() # unique samples with mutation data\n",
    "        all_transcriptomics_samples = transcriptomics_df['improve_sample_id'].unique() # unique samples with transcriptomics data\n",
    "        # now filter to only those samples with transcriptomics data and mutation data\n",
    "        exp_transcriptomics_mutation_df = exp_df[exp_df['improve_sample_id'].isin(all_transcriptomics_samples) & exp_df['improve_sample_id'].isin(all_mutation_samples)]\n",
    "        num_sample_drug_transcript_mutation_pairs = exp_transcriptomics_mutation_df[['improve_sample_id','improve_drug_id']].drop_duplicates().shape[0]\n",
    "    else:\n",
    "        num_sample_drug_transcript_mutation_pairs = np.nan\n",
    "    # unique sample-drug pairs with copynum data and transcriptomics data\n",
    "    if transcriptomics_file and copy_number_file and exp_file:\n",
    "        copy_number_df = pd.read_csv(copy_number_file[0], sep=\"\\t\" if copy_number_file[0].endswith(\".tsv\") else \",\")\n",
    "        all_copy_number_samples = copy_number_df['improve_sample_id'].unique() # unique samples with copy number data\n",
    "        all_transcriptomics_samples = transcriptomics_df['improve_sample_id'].unique() # unique samples with transcriptomics data\n",
    "        # now filter to only those samples with transcriptomics data and copy number data\n",
    "        exp_transcriptomics_copy_number_df = exp_df[exp_df['improve_sample_id'].isin(all_transcriptomics_samples) & exp_df['improve_sample_id'].isin(all_copy_number_samples)]\n",
    "        num_sample_drug_transcript_copynum_pairs = exp_transcriptomics_copy_number_df[['improve_sample_id','improve_drug_id']].drop_duplicates().shape[0]\n",
    "    else:\n",
    "        num_sample_drug_transcript_copynum_pairs = np.nan\n",
    "    # unique sample-drug pairs with mutation data and copynum data\n",
    "    if mutation_file and copy_number_file and exp_file:\n",
    "        all_mutation_samples = mutations_df['improve_sample_id'].unique()\n",
    "        all_copy_number_samples = copy_number_df['improve_sample_id'].unique()\n",
    "        exp_mutation_copy_number_df = exp_df[exp_df['improve_sample_id'].isin(all_mutation_samples) & exp_df['improve_sample_id'].isin(all_copy_number_samples)]\n",
    "        num_sample_drug_mutation_copynum_pairs = exp_mutation_copy_number_df[['improve_sample_id','improve_drug_id']].drop_duplicates().shape[0]\n",
    "    else:\n",
    "        num_sample_drug_mutation_copynum_pairs = np.nan\n",
    "    # adding to dataframe\n",
    "    to_add.append([dataset, num_sample_drug_pairs, num_sample_drug_transcript_pairs, num_sample_drug_transcript_mutation_pairs, num_sample_drug_transcript_copynum_pairs, num_sample_drug_mutation_copynum_pairs])\n",
    "    #print(to_add)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "b2c4a077",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['hcmi', nan, nan, nan, nan, nan],\n",
       " ['beataml', 31926, 4137, 3958, nan, nan],\n",
       " ['mpnst', 272, 193, 184, 191, 184],\n",
       " ['pancreatic', 190, 190, 185, 185, 185],\n",
       " ['cptac', nan, nan, nan, nan, nan],\n",
       " ['sarcoma', 275, 234, 187, nan, nan],\n",
       " ['colorectal', 140, 60, 60, 60, 140],\n",
       " ['bladder', 3300, 840, 640, 640, 3100],\n",
       " ['liver', 4453, 4453, 4453, 4453, 4453],\n",
       " ['novartis', 1766, 1734, 1734, 1723, 1723],\n",
       " ['ccle', 11543, 10887, 10792, 10887, 11118],\n",
       " ['ctrpv2', 309401, 300507, 295742, 299698, 300616],\n",
       " ['fimm', 2663, 2457, 2457, 2457, 2611],\n",
       " ['gdscv1', 247753, 245220, 241999, 241240, 242570],\n",
       " ['gdscv2', 115440, 114373, 112829, 112523, 113133],\n",
       " ['gcsi', 13398, 12506, 12338, 12506, 13112],\n",
       " ['prism', 638983, 632078, 630672, 632078, 636226],\n",
       " ['nci60', 2960756, 2329149, 2329132, 2329149, 2784474]]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "to_add"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "271fad6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "datset_stats = pd.DataFrame(to_add, columns = ['dataset','sample_drug_pairs','sample_drug_transcript_pairs','sample_drug_transcriptomics_mutation_pairs','sample_drug_transcriptomics_copynumber_pairs','sample_drug_mutation_copynumber_pairs'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "c7c21581",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "dataset",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "sample_drug_pairs",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "sample_drug_transcript_pairs",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "sample_drug_transcriptomics_mutation_pairs",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "sample_drug_transcriptomics_copynumber_pairs",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "sample_drug_mutation_copynumber_pairs",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "4c6d619b-c868-4804-b176-201a9458f7c9",
       "rows": [
        [
         "1",
         "beataml",
         "31926.0",
         "4137.0",
         "3958.0",
         null,
         null
        ],
        [
         "7",
         "bladder",
         "3300.0",
         "840.0",
         "640.0",
         "640.0",
         "3100.0"
        ],
        [
         "10",
         "ccle",
         "11543.0",
         "10887.0",
         "10792.0",
         "10887.0",
         "11118.0"
        ],
        [
         "6",
         "colorectal",
         "140.0",
         "60.0",
         "60.0",
         "60.0",
         "140.0"
        ],
        [
         "4",
         "cptac",
         null,
         null,
         null,
         null,
         null
        ],
        [
         "11",
         "ctrpv2",
         "309401.0",
         "300507.0",
         "295742.0",
         "299698.0",
         "300616.0"
        ],
        [
         "12",
         "fimm",
         "2663.0",
         "2457.0",
         "2457.0",
         "2457.0",
         "2611.0"
        ],
        [
         "15",
         "gcsi",
         "13398.0",
         "12506.0",
         "12338.0",
         "12506.0",
         "13112.0"
        ],
        [
         "13",
         "gdscv1",
         "247753.0",
         "245220.0",
         "241999.0",
         "241240.0",
         "242570.0"
        ],
        [
         "14",
         "gdscv2",
         "115440.0",
         "114373.0",
         "112829.0",
         "112523.0",
         "113133.0"
        ],
        [
         "0",
         "hcmi",
         null,
         null,
         null,
         null,
         null
        ],
        [
         "8",
         "liver",
         "4453.0",
         "4453.0",
         "4453.0",
         "4453.0",
         "4453.0"
        ],
        [
         "2",
         "mpnst",
         "272.0",
         "193.0",
         "184.0",
         "191.0",
         "184.0"
        ],
        [
         "17",
         "nci60",
         "2960756.0",
         "2329149.0",
         "2329132.0",
         "2329149.0",
         "2784474.0"
        ],
        [
         "9",
         "novartis",
         "1766.0",
         "1734.0",
         "1734.0",
         "1723.0",
         "1723.0"
        ],
        [
         "3",
         "pancreatic",
         "190.0",
         "190.0",
         "185.0",
         "185.0",
         "185.0"
        ],
        [
         "16",
         "prism",
         "638983.0",
         "632078.0",
         "630672.0",
         "632078.0",
         "636226.0"
        ],
        [
         "5",
         "sarcoma",
         "275.0",
         "234.0",
         "187.0",
         null,
         null
        ]
       ],
       "shape": {
        "columns": 6,
        "rows": 18
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>sample_drug_pairs</th>\n",
       "      <th>sample_drug_transcript_pairs</th>\n",
       "      <th>sample_drug_transcriptomics_mutation_pairs</th>\n",
       "      <th>sample_drug_transcriptomics_copynumber_pairs</th>\n",
       "      <th>sample_drug_mutation_copynumber_pairs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>beataml</td>\n",
       "      <td>31926.0</td>\n",
       "      <td>4137.0</td>\n",
       "      <td>3958.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>bladder</td>\n",
       "      <td>3300.0</td>\n",
       "      <td>840.0</td>\n",
       "      <td>640.0</td>\n",
       "      <td>640.0</td>\n",
       "      <td>3100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ccle</td>\n",
       "      <td>11543.0</td>\n",
       "      <td>10887.0</td>\n",
       "      <td>10792.0</td>\n",
       "      <td>10887.0</td>\n",
       "      <td>11118.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>colorectal</td>\n",
       "      <td>140.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>140.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cptac</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ctrpv2</td>\n",
       "      <td>309401.0</td>\n",
       "      <td>300507.0</td>\n",
       "      <td>295742.0</td>\n",
       "      <td>299698.0</td>\n",
       "      <td>300616.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>fimm</td>\n",
       "      <td>2663.0</td>\n",
       "      <td>2457.0</td>\n",
       "      <td>2457.0</td>\n",
       "      <td>2457.0</td>\n",
       "      <td>2611.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>gcsi</td>\n",
       "      <td>13398.0</td>\n",
       "      <td>12506.0</td>\n",
       "      <td>12338.0</td>\n",
       "      <td>12506.0</td>\n",
       "      <td>13112.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>gdscv1</td>\n",
       "      <td>247753.0</td>\n",
       "      <td>245220.0</td>\n",
       "      <td>241999.0</td>\n",
       "      <td>241240.0</td>\n",
       "      <td>242570.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>gdscv2</td>\n",
       "      <td>115440.0</td>\n",
       "      <td>114373.0</td>\n",
       "      <td>112829.0</td>\n",
       "      <td>112523.0</td>\n",
       "      <td>113133.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hcmi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>liver</td>\n",
       "      <td>4453.0</td>\n",
       "      <td>4453.0</td>\n",
       "      <td>4453.0</td>\n",
       "      <td>4453.0</td>\n",
       "      <td>4453.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mpnst</td>\n",
       "      <td>272.0</td>\n",
       "      <td>193.0</td>\n",
       "      <td>184.0</td>\n",
       "      <td>191.0</td>\n",
       "      <td>184.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>nci60</td>\n",
       "      <td>2960756.0</td>\n",
       "      <td>2329149.0</td>\n",
       "      <td>2329132.0</td>\n",
       "      <td>2329149.0</td>\n",
       "      <td>2784474.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>novartis</td>\n",
       "      <td>1766.0</td>\n",
       "      <td>1734.0</td>\n",
       "      <td>1734.0</td>\n",
       "      <td>1723.0</td>\n",
       "      <td>1723.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pancreatic</td>\n",
       "      <td>190.0</td>\n",
       "      <td>190.0</td>\n",
       "      <td>185.0</td>\n",
       "      <td>185.0</td>\n",
       "      <td>185.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>prism</td>\n",
       "      <td>638983.0</td>\n",
       "      <td>632078.0</td>\n",
       "      <td>630672.0</td>\n",
       "      <td>632078.0</td>\n",
       "      <td>636226.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sarcoma</td>\n",
       "      <td>275.0</td>\n",
       "      <td>234.0</td>\n",
       "      <td>187.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       dataset  sample_drug_pairs  sample_drug_transcript_pairs  \\\n",
       "1      beataml            31926.0                        4137.0   \n",
       "7      bladder             3300.0                         840.0   \n",
       "10        ccle            11543.0                       10887.0   \n",
       "6   colorectal              140.0                          60.0   \n",
       "4        cptac                NaN                           NaN   \n",
       "11      ctrpv2           309401.0                      300507.0   \n",
       "12        fimm             2663.0                        2457.0   \n",
       "15        gcsi            13398.0                       12506.0   \n",
       "13      gdscv1           247753.0                      245220.0   \n",
       "14      gdscv2           115440.0                      114373.0   \n",
       "0         hcmi                NaN                           NaN   \n",
       "8        liver             4453.0                        4453.0   \n",
       "2        mpnst              272.0                         193.0   \n",
       "17       nci60          2960756.0                     2329149.0   \n",
       "9     novartis             1766.0                        1734.0   \n",
       "3   pancreatic              190.0                         190.0   \n",
       "16       prism           638983.0                      632078.0   \n",
       "5      sarcoma              275.0                         234.0   \n",
       "\n",
       "    sample_drug_transcriptomics_mutation_pairs  \\\n",
       "1                                       3958.0   \n",
       "7                                        640.0   \n",
       "10                                     10792.0   \n",
       "6                                         60.0   \n",
       "4                                          NaN   \n",
       "11                                    295742.0   \n",
       "12                                      2457.0   \n",
       "15                                     12338.0   \n",
       "13                                    241999.0   \n",
       "14                                    112829.0   \n",
       "0                                          NaN   \n",
       "8                                       4453.0   \n",
       "2                                        184.0   \n",
       "17                                   2329132.0   \n",
       "9                                       1734.0   \n",
       "3                                        185.0   \n",
       "16                                    630672.0   \n",
       "5                                        187.0   \n",
       "\n",
       "    sample_drug_transcriptomics_copynumber_pairs  \\\n",
       "1                                            NaN   \n",
       "7                                          640.0   \n",
       "10                                       10887.0   \n",
       "6                                           60.0   \n",
       "4                                            NaN   \n",
       "11                                      299698.0   \n",
       "12                                        2457.0   \n",
       "15                                       12506.0   \n",
       "13                                      241240.0   \n",
       "14                                      112523.0   \n",
       "0                                            NaN   \n",
       "8                                         4453.0   \n",
       "2                                          191.0   \n",
       "17                                     2329149.0   \n",
       "9                                         1723.0   \n",
       "3                                          185.0   \n",
       "16                                      632078.0   \n",
       "5                                            NaN   \n",
       "\n",
       "    sample_drug_mutation_copynumber_pairs  \n",
       "1                                     NaN  \n",
       "7                                  3100.0  \n",
       "10                                11118.0  \n",
       "6                                   140.0  \n",
       "4                                     NaN  \n",
       "11                               300616.0  \n",
       "12                                 2611.0  \n",
       "15                                13112.0  \n",
       "13                               242570.0  \n",
       "14                               113133.0  \n",
       "0                                     NaN  \n",
       "8                                  4453.0  \n",
       "2                                   184.0  \n",
       "17                              2784474.0  \n",
       "9                                  1723.0  \n",
       "3                                   185.0  \n",
       "16                               636226.0  \n",
       "5                                     NaN  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datset_stats.sort_values(by='dataset', inplace=True)\n",
    "datset_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "0cd0450e",
   "metadata": {},
   "outputs": [],
   "source": [
    "datset_stats.to_csv(\"../docs/source/_static/dataset_summary_statistics.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c555a37",
   "metadata": {},
   "source": [
    "## Drug Curve Metrics Collected\n",
    "This looks at the experiments results and shows how many drugs we have for each curve metric (found in the dose_response_metric column)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "80906bf8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No experimental data file found for dataset: hcmi\n",
      "No experimental data file found for dataset: cptac\n"
     ]
    }
   ],
   "source": [
    "curve_metric_stats = pd.DataFrame(columns=['dataset','curve_metric','num_drugs'])\n",
    "for dataset in dataset_names + broad_sanger_datasets:\n",
    "    exp_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_experiments.*\".format(dataset))\n",
    "    if exp_file:\n",
    "        exp_df = pd.read_csv(exp_file[0], sep=\"\\t\" if exp_file[0].endswith(\".tsv\") else \",\")\n",
    "        exp_df = exp_df[['improve_drug_id','dose_response_metric']].drop_duplicates()\n",
    "        curve_metric_counts = exp_df.groupby('dose_response_metric')['improve_drug_id'].nunique().reset_index()\n",
    "        curve_metric_counts['dataset'] = dataset\n",
    "        curve_metric_stats = pd.concat([curve_metric_stats, curve_metric_counts.rename(columns={'dose_response_metric':'curve_metric','improve_drug_id':'num_drugs'})], ignore_index=True)\n",
    "    else:\n",
    "        print(f\"No experimental data file found for dataset: {dataset}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "a03f363c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "dataset",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "curve_metric",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "num_drugs",
         "rawType": "object",
         "type": "unknown"
        }
       ],
       "ref": "bc6713e0-def1-481a-acd7-48188b0f2b50",
       "rows": [
        [
         "0",
         "beataml",
         "aac",
         "138"
        ],
        [
         "1",
         "beataml",
         "auc",
         "138"
        ],
        [
         "2",
         "beataml",
         "dss",
         "138"
        ],
        [
         "3",
         "beataml",
         "fit_auc",
         "138"
        ],
        [
         "4",
         "beataml",
         "fit_ec50",
         "138"
        ],
        [
         "5",
         "beataml",
         "fit_ec50se",
         "138"
        ],
        [
         "6",
         "beataml",
         "fit_einf",
         "138"
        ],
        [
         "7",
         "beataml",
         "fit_hs",
         "138"
        ],
        [
         "8",
         "beataml",
         "fit_ic50",
         "138"
        ],
        [
         "9",
         "beataml",
         "fit_r2",
         "138"
        ],
        [
         "10",
         "mpnst",
         "TGI",
         "8"
        ],
        [
         "11",
         "mpnst",
         "aac",
         "27"
        ],
        [
         "12",
         "mpnst",
         "abc",
         "8"
        ],
        [
         "13",
         "mpnst",
         "auc",
         "27"
        ],
        [
         "14",
         "mpnst",
         "dss",
         "27"
        ],
        [
         "15",
         "mpnst",
         "fit_auc",
         "27"
        ],
        [
         "16",
         "mpnst",
         "fit_ec50",
         "27"
        ],
        [
         "17",
         "mpnst",
         "fit_ec50se",
         "27"
        ],
        [
         "18",
         "mpnst",
         "fit_einf",
         "27"
        ],
        [
         "19",
         "mpnst",
         "fit_hs",
         "27"
        ],
        [
         "20",
         "mpnst",
         "fit_ic50",
         "27"
        ],
        [
         "21",
         "mpnst",
         "fit_r2",
         "27"
        ],
        [
         "22",
         "mpnst",
         "lmm",
         "8"
        ],
        [
         "23",
         "mpnst",
         "mRESCIST",
         "8"
        ],
        [
         "24",
         "pancreatic",
         "aac",
         "5"
        ],
        [
         "25",
         "pancreatic",
         "auc",
         "5"
        ],
        [
         "26",
         "pancreatic",
         "dss",
         "5"
        ],
        [
         "27",
         "pancreatic",
         "fit_auc",
         "5"
        ],
        [
         "28",
         "pancreatic",
         "fit_ec50",
         "5"
        ],
        [
         "29",
         "pancreatic",
         "fit_ec50se",
         "5"
        ],
        [
         "30",
         "pancreatic",
         "fit_einf",
         "5"
        ],
        [
         "31",
         "pancreatic",
         "fit_hs",
         "5"
        ],
        [
         "32",
         "pancreatic",
         "fit_ic50",
         "5"
        ],
        [
         "33",
         "pancreatic",
         "fit_r2",
         "5"
        ],
        [
         "34",
         "sarcoma",
         "published_auc",
         "32"
        ],
        [
         "35",
         "colorectal",
         "aac",
         "10"
        ],
        [
         "36",
         "colorectal",
         "auc",
         "10"
        ],
        [
         "37",
         "colorectal",
         "dss",
         "10"
        ],
        [
         "38",
         "colorectal",
         "fit_auc",
         "10"
        ],
        [
         "39",
         "colorectal",
         "fit_ec50",
         "10"
        ],
        [
         "40",
         "colorectal",
         "fit_ec50se",
         "10"
        ],
        [
         "41",
         "colorectal",
         "fit_einf",
         "10"
        ],
        [
         "42",
         "colorectal",
         "fit_hs",
         "10"
        ],
        [
         "43",
         "colorectal",
         "fit_ic50",
         "10"
        ],
        [
         "44",
         "colorectal",
         "fit_r2",
         "10"
        ],
        [
         "45",
         "bladder",
         "aac",
         "50"
        ],
        [
         "46",
         "bladder",
         "auc",
         "50"
        ],
        [
         "47",
         "bladder",
         "dss",
         "50"
        ],
        [
         "48",
         "bladder",
         "fit_auc",
         "50"
        ],
        [
         "49",
         "bladder",
         "fit_ec50",
         "50"
        ]
       ],
       "shape": {
        "columns": 3,
        "rows": 149
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>curve_metric</th>\n",
       "      <th>num_drugs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>beataml</td>\n",
       "      <td>aac</td>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>beataml</td>\n",
       "      <td>auc</td>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>beataml</td>\n",
       "      <td>dss</td>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>beataml</td>\n",
       "      <td>fit_auc</td>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>beataml</td>\n",
       "      <td>fit_ec50</td>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>nci60</td>\n",
       "      <td>fit_ec50se</td>\n",
       "      <td>55157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>nci60</td>\n",
       "      <td>fit_einf</td>\n",
       "      <td>55157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>nci60</td>\n",
       "      <td>fit_hs</td>\n",
       "      <td>55157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>nci60</td>\n",
       "      <td>fit_ic50</td>\n",
       "      <td>55157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>nci60</td>\n",
       "      <td>fit_r2</td>\n",
       "      <td>55157</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>149 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     dataset curve_metric num_drugs\n",
       "0    beataml          aac       138\n",
       "1    beataml          auc       138\n",
       "2    beataml          dss       138\n",
       "3    beataml      fit_auc       138\n",
       "4    beataml     fit_ec50       138\n",
       "..       ...          ...       ...\n",
       "144    nci60   fit_ec50se     55157\n",
       "145    nci60     fit_einf     55157\n",
       "146    nci60       fit_hs     55157\n",
       "147    nci60     fit_ic50     55157\n",
       "148    nci60       fit_r2     55157\n",
       "\n",
       "[149 rows x 3 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curve_metric_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "d131414c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "dataset",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "num_drugs",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "aac",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "abc",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "auc",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "dss",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "fit_auc",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "fit_ec50",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "fit_ec50se",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "fit_einf",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "fit_hs",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "fit_ic50",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "fit_r2",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "lmm",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mRESCIST",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "published_auc",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "TGI",
         "rawType": "object",
         "type": "string"
        }
       ],
       "ref": "01cfbf24-3326-45be-a05e-fcfa8c03938e",
       "rows": [
        [
         "0",
         "beataml",
         "164",
         "X",
         "",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "",
         "",
         "",
         ""
        ],
        [
         "1",
         "bladder",
         "50",
         "X",
         "",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "",
         "",
         "",
         ""
        ],
        [
         "2",
         "ccle",
         "24",
         "X",
         "",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "",
         "",
         "",
         ""
        ],
        [
         "3",
         "colorectal",
         "10",
         "X",
         "",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "",
         "",
         "",
         ""
        ],
        [
         "4",
         "ctrpv2",
         "459",
         "X",
         "",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "",
         "",
         "",
         ""
        ],
        [
         "5",
         "fimm",
         "52",
         "X",
         "",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "",
         "",
         "",
         ""
        ],
        [
         "6",
         "gcsi",
         "44",
         "X",
         "",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "",
         "",
         "",
         ""
        ],
        [
         "7",
         "gdscv1",
         "294",
         "X",
         "",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "",
         "",
         "",
         ""
        ],
        [
         "8",
         "gdscv2",
         "171",
         "X",
         "",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "",
         "",
         "",
         ""
        ],
        [
         "9",
         "liver",
         "76",
         "X",
         "",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "",
         "",
         "",
         ""
        ],
        [
         "10",
         "mpnst",
         "30",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "",
         "X"
        ],
        [
         "11",
         "nci60",
         "55157",
         "X",
         "",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "",
         "",
         "",
         ""
        ],
        [
         "12",
         "novartis",
         "25",
         "",
         "X",
         "",
         "",
         "",
         "",
         "",
         "",
         "",
         "",
         "",
         "X",
         "X",
         "",
         "X"
        ],
        [
         "13",
         "pancreatic",
         "25",
         "X",
         "",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "",
         "",
         "",
         ""
        ],
        [
         "14",
         "prism",
         "1419",
         "X",
         "",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "X",
         "",
         "",
         "",
         ""
        ],
        [
         "15",
         "sarcoma",
         "34",
         "",
         "",
         "",
         "",
         "",
         "",
         "",
         "",
         "",
         "",
         "",
         "",
         "",
         "X",
         ""
        ]
       ],
       "shape": {
        "columns": 17,
        "rows": 16
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>num_drugs</th>\n",
       "      <th>aac</th>\n",
       "      <th>abc</th>\n",
       "      <th>auc</th>\n",
       "      <th>dss</th>\n",
       "      <th>fit_auc</th>\n",
       "      <th>fit_ec50</th>\n",
       "      <th>fit_ec50se</th>\n",
       "      <th>fit_einf</th>\n",
       "      <th>fit_hs</th>\n",
       "      <th>fit_ic50</th>\n",
       "      <th>fit_r2</th>\n",
       "      <th>lmm</th>\n",
       "      <th>mRESCIST</th>\n",
       "      <th>published_auc</th>\n",
       "      <th>TGI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>beataml</td>\n",
       "      <td>164</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bladder</td>\n",
       "      <td>50</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ccle</td>\n",
       "      <td>24</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>colorectal</td>\n",
       "      <td>10</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ctrpv2</td>\n",
       "      <td>459</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>fimm</td>\n",
       "      <td>52</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>gcsi</td>\n",
       "      <td>44</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>gdscv1</td>\n",
       "      <td>294</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>gdscv2</td>\n",
       "      <td>171</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>liver</td>\n",
       "      <td>76</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>mpnst</td>\n",
       "      <td>30</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>nci60</td>\n",
       "      <td>55157</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>novartis</td>\n",
       "      <td>25</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>pancreatic</td>\n",
       "      <td>25</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>prism</td>\n",
       "      <td>1419</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>sarcoma</td>\n",
       "      <td>34</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       dataset num_drugs aac abc auc dss fit_auc fit_ec50 fit_ec50se fit_einf  \\\n",
       "0      beataml       164   X       X   X       X        X          X        X   \n",
       "1      bladder        50   X       X   X       X        X          X        X   \n",
       "2         ccle        24   X       X   X       X        X          X        X   \n",
       "3   colorectal        10   X       X   X       X        X          X        X   \n",
       "4       ctrpv2       459   X       X   X       X        X          X        X   \n",
       "5         fimm        52   X       X   X       X        X          X        X   \n",
       "6         gcsi        44   X       X   X       X        X          X        X   \n",
       "7       gdscv1       294   X       X   X       X        X          X        X   \n",
       "8       gdscv2       171   X       X   X       X        X          X        X   \n",
       "9        liver        76   X       X   X       X        X          X        X   \n",
       "10       mpnst        30   X   X   X   X       X        X          X        X   \n",
       "11       nci60     55157   X       X   X       X        X          X        X   \n",
       "12    novartis        25       X                                                \n",
       "13  pancreatic        25   X       X   X       X        X          X        X   \n",
       "14       prism      1419   X       X   X       X        X          X        X   \n",
       "15     sarcoma        34                                                        \n",
       "\n",
       "   fit_hs fit_ic50 fit_r2 lmm mRESCIST published_auc TGI  \n",
       "0       X        X      X                                 \n",
       "1       X        X      X                                 \n",
       "2       X        X      X                                 \n",
       "3       X        X      X                                 \n",
       "4       X        X      X                                 \n",
       "5       X        X      X                                 \n",
       "6       X        X      X                                 \n",
       "7       X        X      X                                 \n",
       "8       X        X      X                                 \n",
       "9       X        X      X                                 \n",
       "10      X        X      X   X        X                 X  \n",
       "11      X        X      X                                 \n",
       "12                          X        X                 X  \n",
       "13      X        X      X                                 \n",
       "14      X        X      X                                 \n",
       "15                                                 X      "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "presence_absence = (\n",
    "    curve_metric_stats.assign(present=\"X\")\n",
    "      .pivot_table(index=\"dataset\", columns=\"curve_metric\", values=\"present\", aggfunc=\"first\", fill_value=\"\")\n",
    "      .reset_index()\n",
    ")\n",
    "presence_absence = presence_absence.merge(datset_stats[['dataset','num_drugs']], on='dataset')\n",
    "sorted_columns = ['dataset','num_drugs'] + sorted([col for col in presence_absence.columns if col not in ['dataset','num_drugs','TGI']]) + ['TGI']\n",
    "presence_absence = presence_absence[sorted_columns]\n",
    "presence_absence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "29ed18cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "presence_absence.to_csv(\"../docs/source/_static/dataset_curve_metrics_wide.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e0f049c",
   "metadata": {},
   "source": [
    "# Dataset Modalities Overview DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "8881426f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing files for dataset: hcmi\n",
      "hcmi 1711 nan  X X X\n",
      "beataml 1022 164 X X X \n",
      "mpnst 50 30 X X X X\n",
      "pancreatic 49 25 X X X X\n",
      "Missing files for dataset: cptac\n",
      "cptac 1139 nan  X X X\n",
      "sarcoma 36 34 X X X \n",
      "colorectal 61 10 X X X X\n",
      "bladder 134 50 X X X X\n",
      "liver 62 76 X X X X\n",
      "novartis 386 25 X X X X\n",
      "ccle 502 24 X X X X\n",
      "ctrpv2 846 459 X X X X\n",
      "fimm 52 52 X X X X\n",
      "gdscv1 984 294 X X X X\n",
      "gdscv2 806 171 X X X X\n",
      "gcsi 569 44 X X X X\n",
      "prism 478 1419 X X X X\n",
      "nci60 83 55157 X X X X\n"
     ]
    }
   ],
   "source": [
    "datset_modalities_df = pd.DataFrame(columns=[\"Dataset\", \"References\", \"Sample\", \"Transcriptomics\", \"Proteomics\", \"Mutations\", \"Copy Number\", \"Drug\", \"Drug Descriptor\", \"Experiments\"])\n",
    "list_of_rows = list()\n",
    "\n",
    "for dataset in dataset_names + broad_sanger_datasets:\n",
    "    sample_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_samples.*\".format(dataset))\n",
    "    drug_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_drugs.*\".format(dataset))\n",
    "    exp_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_experiments.*\".format(dataset))\n",
    "    transcriptomics_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_transcriptomics.*\".format(dataset))\n",
    "    mutation_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_mutations.*\".format(dataset))\n",
    "    copy_number_file = glob.glob(\"/Users/imal967/pnnl/projects/coderdata/all_datasets/{}_copy_number.*\".format(dataset))\n",
    "    # getting number of samples\n",
    "    if sample_file:\n",
    "        samples_df = pd.read_csv(sample_file[0], sep=\"\\t\" if sample_file[0].endswith(\".tsv\") else \",\")\n",
    "        num_samples = samples_df['improve_sample_id'].nunique() # unique samples\n",
    "    else:\n",
    "        print(f\"Missing files for dataset: {dataset}\")\n",
    "        num_samples = np.nan\n",
    "    # getting number of drugs\n",
    "    if drug_file:\n",
    "        drugs_df = pd.read_csv(drug_file[0], sep=\"\\t\" if drug_file[0].endswith(\".tsv\") else \",\")\n",
    "        num_drugs = drugs_df['improve_drug_id'].nunique() # unique drugs\n",
    "    else:\n",
    "        print(f\"Missing files for dataset: {dataset}\")\n",
    "        num_drugs = np.nan\n",
    "    if exp_file:\n",
    "        exp_present = \"X\"\n",
    "    else:\n",
    "        exp_present = \"\"\n",
    "    if transcriptomics_file:\n",
    "        transcriptomics_present = \"X\"\n",
    "    else:\n",
    "        transcriptomics_present = \"\"\n",
    "    if mutation_file:\n",
    "        mutation_present = \"X\"\n",
    "    else:\n",
    "        mutation_present = \"\"\n",
    "    if copy_number_file:\n",
    "        copy_number_present = \"X\"\n",
    "    else:\n",
    "        copy_number_present = \"\"\n",
    "    print(dataset, num_samples, num_drugs, exp_present, transcriptomics_present, mutation_present, copy_number_present)\n",
    "    # list_of_rows = list_of_rows.append({'dataset':[dataset],'sample_drug_pairs':[num_sample_drug_pairs],'sample_drug_transcriptomics_pairs':num_sample_drug_transcript_pairs,'sample_drug_transcriptomics_mutation_pairs':num_sample_drug_transcript_mutation_pairs,'sample_drug_transcriptomics_copynumberpairs':num_sample_drug_transcript_copynum_pairs,'sample_drug_mutation_copynumber_pairs':num_sample_drug_mutation_copynum_pairs})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3bd65bb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "f603d8bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Dataset",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "References",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Sample",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Transcriptomics",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Proteomics",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Mutations",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Copy Number",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Drug",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Drug Descriptor",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Experiments",
         "rawType": "object",
         "type": "string"
        }
       ],
       "ref": "15d68210-da92-419e-ace6-09413314c4f2",
       "rows": [],
       "shape": {
        "columns": 10,
        "rows": 0
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th>References</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Transcriptomics</th>\n",
       "      <th>Proteomics</th>\n",
       "      <th>Mutations</th>\n",
       "      <th>Copy Number</th>\n",
       "      <th>Drug</th>\n",
       "      <th>Drug Descriptor</th>\n",
       "      <th>Experiments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Dataset, References, Sample, Transcriptomics, Proteomics, Mutations, Copy Number, Drug, Drug Descriptor, Experiments]\n",
       "Index: []"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datset_stats"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cdc_organoids",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
