<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Datasets Included &#8212; CoderData</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=d75fae25" />
    <link rel="stylesheet" type="text/css" href="_static/sphinxdoc.css?v=34905f61" />
    <link rel="stylesheet" type="text/css" href="_static/custom.css?v=08cfd9eb" />
    <script src="_static/documentation_options.js?v=8d563738"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="canonical" href="https://pnnl-compbio.github.io/coderdata/datasets_included.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="How to Contribute" href="contribution.html" />
    <link rel="prev" title="API Reference" href="APIreference.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="Related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="contribution.html" title="How to Contribute"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="APIreference.html" title="API Reference"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">CoderData</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Datasets Included</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="datasets-included">
<h1>Datasets Included<a class="headerlink" href="#datasets-included" title="Link to this heading">¶</a></h1>
<p>This page provides an overview of the datasets included in CoderData version 2.2.0. This package collects 18 diverse sets of paired molecular datasets with corresponding drug sensitivity data. All data here is reprocessed and standardized so it can be easily used as a benchmark dataset for drug response prediction machine learning models.</p>
<p>The dataset files are in csv format and are available at the link below:</p>
<p>Figshare record: <a class="reference external" href="https://api.figshare.com/v2/articles/28823159">https://api.figshare.com/v2/articles/28823159</a></p>
<p>Version: 2.2.0</p>
<section id="dataset-overview">
<h2>Dataset Overview<a class="headerlink" href="#dataset-overview" title="Link to this heading">¶</a></h2>
<table class="docutils align-default" id="id54">
<caption><span class="caption-text">Datasets and Modalities</span><a class="headerlink" href="#id54" title="Link to this table">¶</a></caption>
<colgroup>
<col style="width: 12.2%" />
<col style="width: 10.4%" />
<col style="width: 5.2%" />
<col style="width: 7.0%" />
<col style="width: 13.0%" />
<col style="width: 10.4%" />
<col style="width: 10.4%" />
<col style="width: 10.4%" />
<col style="width: 10.4%" />
<col style="width: 10.4%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Dataset</p></th>
<th class="head"><p>References</p></th>
<th class="head"><p>Sample</p></th>
<th class="head"><p>Drug</p></th>
<th class="head"><p>Drug Descriptor</p></th>
<th class="head"><p>Experiments</p></th>
<th class="head"><p>Transcriptomics</p></th>
<th class="head"><p>Proteomics</p></th>
<th class="head"><p>Mutations</p></th>
<th class="head"><p>Copy Number</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>BeatAML</p></td>
<td><p><a class="footnote-reference brackets" href="#id29" id="id1" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>, <a class="footnote-reference brackets" href="#id30" id="id2" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a></p></td>
<td><p>1022</p></td>
<td><p>164</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Bladder</p></td>
<td><p><a class="footnote-reference brackets" href="#id31" id="id3" role="doc-noteref"><span class="fn-bracket">[</span>3<span class="fn-bracket">]</span></a></p></td>
<td><p>134</p></td>
<td><p>50</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
</tr>
<tr class="row-even"><td><p>CCLE</p></td>
<td><p><a class="footnote-reference brackets" href="#id32" id="id4" role="doc-noteref"><span class="fn-bracket">[</span>4<span class="fn-bracket">]</span></a></p></td>
<td><p>502</p></td>
<td><p>24</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
</tr>
<tr class="row-odd"><td><p>Colorectal</p></td>
<td><p><a class="footnote-reference brackets" href="#id46" id="id5" role="doc-noteref"><span class="fn-bracket">[</span>18<span class="fn-bracket">]</span></a></p></td>
<td><p>61</p></td>
<td><p>10</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
</tr>
<tr class="row-even"><td><p>CPTAC</p></td>
<td><p><a class="footnote-reference brackets" href="#id33" id="id6" role="doc-noteref"><span class="fn-bracket">[</span>5<span class="fn-bracket">]</span></a></p></td>
<td><p>1139</p></td>
<td></td>
<td></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
</tr>
<tr class="row-odd"><td><p>CTRPv2</p></td>
<td><p><a class="footnote-reference brackets" href="#id34" id="id7" role="doc-noteref"><span class="fn-bracket">[</span>6<span class="fn-bracket">]</span></a>, <a class="footnote-reference brackets" href="#id35" id="id8" role="doc-noteref"><span class="fn-bracket">[</span>7<span class="fn-bracket">]</span></a>, <a class="footnote-reference brackets" href="#id36" id="id9" role="doc-noteref"><span class="fn-bracket">[</span>8<span class="fn-bracket">]</span></a></p></td>
<td><p>846</p></td>
<td><p>459</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
</tr>
<tr class="row-even"><td><p>FIMM</p></td>
<td><p><a class="footnote-reference brackets" href="#id37" id="id10" role="doc-noteref"><span class="fn-bracket">[</span>9<span class="fn-bracket">]</span></a>, <a class="footnote-reference brackets" href="#id38" id="id11" role="doc-noteref"><span class="fn-bracket">[</span>10<span class="fn-bracket">]</span></a></p></td>
<td><p>52</p></td>
<td><p>52</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>GDSC v1</p></td>
<td><p><a class="footnote-reference brackets" href="#id51" id="id12" role="doc-noteref"><span class="fn-bracket">[</span>23<span class="fn-bracket">]</span></a>, <a class="footnote-reference brackets" href="#id52" id="id13" role="doc-noteref"><span class="fn-bracket">[</span>24<span class="fn-bracket">]</span></a>, <a class="footnote-reference brackets" href="#id53" id="id14" role="doc-noteref"><span class="fn-bracket">[</span>25<span class="fn-bracket">]</span></a></p></td>
<td><p>984</p></td>
<td><p>294</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
</tr>
<tr class="row-even"><td><p>GDSC v2</p></td>
<td><p><a class="footnote-reference brackets" href="#id51" id="id15" role="doc-noteref"><span class="fn-bracket">[</span>23<span class="fn-bracket">]</span></a>, <a class="footnote-reference brackets" href="#id52" id="id16" role="doc-noteref"><span class="fn-bracket">[</span>24<span class="fn-bracket">]</span></a>, <a class="footnote-reference brackets" href="#id53" id="id17" role="doc-noteref"><span class="fn-bracket">[</span>25<span class="fn-bracket">]</span></a></p></td>
<td><p>806</p></td>
<td><p>171</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
</tr>
<tr class="row-odd"><td><p>gCSI</p></td>
<td><p><a class="footnote-reference brackets" href="#id49" id="id18" role="doc-noteref"><span class="fn-bracket">[</span>21<span class="fn-bracket">]</span></a>, <a class="footnote-reference brackets" href="#id50" id="id19" role="doc-noteref"><span class="fn-bracket">[</span>22<span class="fn-bracket">]</span></a></p></td>
<td><p>569</p></td>
<td><p>44</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
</tr>
<tr class="row-even"><td><p>HCMI</p></td>
<td><p><a class="footnote-reference brackets" href="#id39" id="id20" role="doc-noteref"><span class="fn-bracket">[</span>11<span class="fn-bracket">]</span></a></p></td>
<td><p>886</p></td>
<td></td>
<td></td>
<td></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
</tr>
<tr class="row-odd"><td><p>Liver</p></td>
<td><p><a class="footnote-reference brackets" href="#id47" id="id21" role="doc-noteref"><span class="fn-bracket">[</span>19<span class="fn-bracket">]</span></a></p></td>
<td><p>62</p></td>
<td><p>76</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
</tr>
<tr class="row-even"><td><p>MPNST</p></td>
<td><p><a class="footnote-reference brackets" href="#id40" id="id22" role="doc-noteref"><span class="fn-bracket">[</span>12<span class="fn-bracket">]</span></a></p></td>
<td><p>50</p></td>
<td><p>30</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
</tr>
<tr class="row-odd"><td><p>NCI60</p></td>
<td><p><a class="footnote-reference brackets" href="#id41" id="id23" role="doc-noteref"><span class="fn-bracket">[</span>13<span class="fn-bracket">]</span></a></p></td>
<td><p>83</p></td>
<td><p>55157</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Novartis</p></td>
<td><p><a class="footnote-reference brackets" href="#id48" id="id24" role="doc-noteref"><span class="fn-bracket">[</span>20<span class="fn-bracket">]</span></a></p></td>
<td><p>386</p></td>
<td><p>25</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
</tr>
<tr class="row-odd"><td><p>Pancreatic</p></td>
<td><p><a class="footnote-reference brackets" href="#id42" id="id25" role="doc-noteref"><span class="fn-bracket">[</span>14<span class="fn-bracket">]</span></a></p></td>
<td><p>70</p></td>
<td><p>25</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
</tr>
<tr class="row-even"><td><p>PRISM</p></td>
<td><p><a class="footnote-reference brackets" href="#id43" id="id26" role="doc-noteref"><span class="fn-bracket">[</span>15<span class="fn-bracket">]</span></a>, <a class="footnote-reference brackets" href="#id44" id="id27" role="doc-noteref"><span class="fn-bracket">[</span>16<span class="fn-bracket">]</span></a></p></td>
<td><p>478</p></td>
<td><p>1419</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Sarcoma</p></td>
<td><p><a class="footnote-reference brackets" href="#id45" id="id28" role="doc-noteref"><span class="fn-bracket">[</span>17<span class="fn-bracket">]</span></a></p></td>
<td><p>36</p></td>
<td><p>34</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td></td>
</tr>
</tbody>
</table>
<p>The table above lists the datasets included in CoderData version 2.2.0, along with references to their original publications, counts of samples and drugs, and the types of data available for each dataset.</p>
<p>CoderData includes the following data:</p>
<ul class="simple">
<li><p>Sample - cell lines, patient-derived samples, or patient-derived organoids</p></li>
<li><p>Drug - compounds tested for sensitivity</p></li>
<li><p>Drug Descriptor - molecular descriptors for each drug (computed using RDKit)</p></li>
<li><p>Experiments - dose-response experiments (various metrics such as AUC, IC50, etc.)</p></li>
<li><p>Transcriptomics - gene expression (in transcripts per million, TPM)</p></li>
<li><p>Proteomics - protein expression (in log2 ratio to reference)</p></li>
<li><p>Mutations - gene mutations (variant calls)</p></li>
<li><p>Copy Number - gene copy number variations (number of copies of each gene, 2 being diploid)</p></li>
</ul>
<p>An “X” indicates the presence of a particular data type for the corresponding dataset.  Each sample in the datasets corresponds to either a cancer cell line, a patient-derived xenograft, or a patient-derived organoid, depending on the specific dataset.</p>
</section>
<section id="dataset-summary-statistics">
<h2>Dataset Summary Statistics<a class="headerlink" href="#dataset-summary-statistics" title="Link to this heading">¶</a></h2>
<p>The following table summarizes combination counts for each dataset. This includes the number of experimental sample-drug pairs, with different molecular data types. Each column represents the number of unique combinations of samples and drugs with the specified molecular data types available.
For example, the “Sample-Drug-Transcriptomics-Mutations” column indicates the number of unique sample-drug pairs that have both transcriptomics and mutation data available. These counts let you estimate how much paired data is available for tasks like building predictive models with transcriptomics and drug response.</p>
<blockquote>
<div><table class="docutils align-default" id="id55">
<caption><span class="caption-text">Dataset Summary Statistics</span><a class="headerlink" href="#id55" title="Link to this table">¶</a></caption>
<tbody>
<tr class="row-odd"><td><p>dataset</p></td>
<td><p>sample_drug_pairs</p></td>
<td><p>sample_drug_transcript_pairs</p></td>
<td><p>sample_drug_transcriptomics_mutation_pairs</p></td>
<td><p>sample_drug_transcriptomics_copynumber_pairs</p></td>
<td><p>sample_drug_mutation_copynumber_pairs</p></td>
</tr>
<tr class="row-even"><td><p>beataml</p></td>
<td><p>31926.0</p></td>
<td><p>4137.0</p></td>
<td><p>3958.0</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>bladder</p></td>
<td><p>3300.0</p></td>
<td><p>840.0</p></td>
<td><p>640.0</p></td>
<td><p>640.0</p></td>
<td><p>3100.0</p></td>
</tr>
<tr class="row-even"><td><p>ccle</p></td>
<td><p>11543.0</p></td>
<td><p>10887.0</p></td>
<td><p>10792.0</p></td>
<td><p>10887.0</p></td>
<td><p>11118.0</p></td>
</tr>
<tr class="row-odd"><td><p>colorectal</p></td>
<td><p>140.0</p></td>
<td><p>60.0</p></td>
<td><p>60.0</p></td>
<td><p>60.0</p></td>
<td><p>140.0</p></td>
</tr>
<tr class="row-even"><td><p>cptac</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>ctrpv2</p></td>
<td><p>309401.0</p></td>
<td><p>300507.0</p></td>
<td><p>295742.0</p></td>
<td><p>299698.0</p></td>
<td><p>300616.0</p></td>
</tr>
<tr class="row-even"><td><p>fimm</p></td>
<td><p>2663.0</p></td>
<td><p>2457.0</p></td>
<td><p>2457.0</p></td>
<td><p>2457.0</p></td>
<td><p>2611.0</p></td>
</tr>
<tr class="row-odd"><td><p>gcsi</p></td>
<td><p>13398.0</p></td>
<td><p>12506.0</p></td>
<td><p>12338.0</p></td>
<td><p>12506.0</p></td>
<td><p>13112.0</p></td>
</tr>
<tr class="row-even"><td><p>gdscv1</p></td>
<td><p>247753.0</p></td>
<td><p>245220.0</p></td>
<td><p>241999.0</p></td>
<td><p>241240.0</p></td>
<td><p>242570.0</p></td>
</tr>
<tr class="row-odd"><td><p>gdscv2</p></td>
<td><p>115440.0</p></td>
<td><p>114373.0</p></td>
<td><p>112829.0</p></td>
<td><p>112523.0</p></td>
<td><p>113133.0</p></td>
</tr>
<tr class="row-even"><td><p>hcmi</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>liver</p></td>
<td><p>4453.0</p></td>
<td><p>4453.0</p></td>
<td><p>4453.0</p></td>
<td><p>4453.0</p></td>
<td><p>4453.0</p></td>
</tr>
<tr class="row-even"><td><p>mpnst</p></td>
<td><p>272.0</p></td>
<td><p>193.0</p></td>
<td><p>184.0</p></td>
<td><p>191.0</p></td>
<td><p>184.0</p></td>
</tr>
<tr class="row-odd"><td><p>nci60</p></td>
<td><p>2960756.0</p></td>
<td><p>2329149.0</p></td>
<td><p>2329132.0</p></td>
<td><p>2329149.0</p></td>
<td><p>2784474.0</p></td>
</tr>
<tr class="row-even"><td><p>novartis</p></td>
<td><p>1766.0</p></td>
<td><p>1734.0</p></td>
<td><p>1734.0</p></td>
<td><p>1723.0</p></td>
<td><p>1723.0</p></td>
</tr>
<tr class="row-odd"><td><p>pancreatic</p></td>
<td><p>190.0</p></td>
<td><p>190.0</p></td>
<td><p>185.0</p></td>
<td><p>185.0</p></td>
<td><p>185.0</p></td>
</tr>
<tr class="row-even"><td><p>prism</p></td>
<td><p>638983.0</p></td>
<td><p>632078.0</p></td>
<td><p>630672.0</p></td>
<td><p>632078.0</p></td>
<td><p>636226.0</p></td>
</tr>
<tr class="row-odd"><td><p>sarcoma</p></td>
<td><p>275.0</p></td>
<td><p>234.0</p></td>
<td><p>187.0</p></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</div></blockquote>
</section>
<section id="drug-curve-metrics-collected">
<h2>Drug Curve Metrics Collected<a class="headerlink" href="#drug-curve-metrics-collected" title="Link to this heading">¶</a></h2>
<p>The following table summarizes the number of drugs associated with each dose-response metric across the datasets.</p>
<blockquote>
<div><table class="docutils align-default" id="id56">
<caption><span class="caption-text">Drug Curve Metrics Summary</span><a class="headerlink" href="#id56" title="Link to this table">¶</a></caption>
<tbody>
<tr class="row-odd"><td><p>dataset</p></td>
<td><p>num_drugs</p></td>
<td><p>aac</p></td>
<td><p>abc</p></td>
<td><p>auc</p></td>
<td><p>dss</p></td>
<td><p>fit_auc</p></td>
<td><p>fit_ec50</p></td>
<td><p>fit_ec50se</p></td>
<td><p>fit_einf</p></td>
<td><p>fit_hs</p></td>
<td><p>fit_ic50</p></td>
<td><p>fit_r2</p></td>
<td><p>lmm</p></td>
<td><p>mRESCIST</p></td>
<td><p>published_auc</p></td>
<td><p>TGI</p></td>
</tr>
<tr class="row-even"><td><p>beataml</p></td>
<td><p>164</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>bladder</p></td>
<td><p>50</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>ccle</p></td>
<td><p>24</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>colorectal</p></td>
<td><p>10</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>ctrpv2</p></td>
<td><p>459</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>fimm</p></td>
<td><p>52</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>gcsi</p></td>
<td><p>44</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>gdscv1</p></td>
<td><p>294</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>gdscv2</p></td>
<td><p>171</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>liver</p></td>
<td><p>76</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>mpnst</p></td>
<td><p>30</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
</tr>
<tr class="row-odd"><td><p>nci60</p></td>
<td><p>55157</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>novartis</p></td>
<td><p>25</p></td>
<td></td>
<td><p>X</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
</tr>
<tr class="row-odd"><td><p>pancreatic</p></td>
<td><p>25</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>prism</p></td>
<td><p>1419</p></td>
<td><p>X</p></td>
<td></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td><p>X</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>sarcoma</p></td>
<td><p>34</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td><p>X</p></td>
<td></td>
</tr>
</tbody>
</table>
</div></blockquote>
<p>Types of dose-response metrics collected include:</p>
<ul class="simple">
<li><p>AAC - Area above the response curve; the complement value of AUC.</p></li>
<li><p>ABC - Area between curves, the difference between the AUC of the control and the treated cells.</p></li>
<li><p>AUC - Area under the fitted hill slope curve across all doses present. Lower AUC signifies lower levels of growth.</p></li>
<li><p>DSS - A multiparametric dose response value that takes into account control and treated cells.</p></li>
<li><p>fit_auc - Area under the fitted hill slope curve across the common interval of −log10[M], where the molar concentration ranges from 10⁻⁴ to 10⁻¹⁰.</p></li>
<li><p>fit_ec50 - The fitted curve prediction of the −log10M concentration at which 50% of the maximal effect is observed.</p></li>
<li><p>fit_ec50se - Standard error of the Fit_EC50 estimate.</p></li>
<li><p>fit_einf - The fraction of cells that are unaffected even at an infinite dose concentration. Calculated as the lower asymptote of the hill slope function.</p></li>
<li><p>fit_hs - The estimated hill slope binding cooperativity, calculated as the slope of the sigmoidal hill curve.</p></li>
<li><p>fit_ic50 - The fitted curve prediction of the −log10M concentration required to reduce tumor growth by 50%.</p></li>
<li><p>fit_r2 - Coefficient of determination between observed growth and the fitted hill slope curve, indicating goodness of fit.</p></li>
<li><p>lmm - The resulting “time and treatment interaction” in a linear mixed model with fixed effects as time and treatment and patient as a random effect. Indicates how much the treatment changes the slope of log(volume) over time compared to the control.</p></li>
<li><p>mRESCIST - Disease status classified into PD (progressive disease), SD (stable disease), PR (partial response), and CR (complete response), based on percent volume change and cumulative average response.</p></li>
<li><p>published_auc - Published Area Under the Curve</p></li>
<li><p>TG - Tumor growth inhibition between the control and treatment time-volume curves.</p></li>
</ul>
</section>
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Link to this heading">¶</a></h2>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id29" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">1</a><span class="fn-bracket">]</span></span>
<p>Bottomly D, Long N, Schultz AR, et al. <em>Integrative analysis of drug response and clinical outcome in acute myeloid leukemia.</em> Cancer Cell. 2022;40(8):850-864.e9. doi:<a class="reference external" href="https://doi.org/10.1016/j.ccell.2022.07.002">10.1016/j.ccell.2022.07.002</a></p>
</aside>
<aside class="footnote brackets" id="id30" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id2">2</a><span class="fn-bracket">]</span></span>
<p>Pino JC, Posso C, Joshi SK, et al. <em>Mapping the proteogenomic landscape enables prediction of drug response in acute myeloid leukemia.</em> Cell Rep Med. 2024;5(1):101359. doi:<a class="reference external" href="https://doi.org/10.1016/j.xcrm.2023.101359">10.1016/j.xcrm.2023.101359</a></p>
</aside>
<aside class="footnote brackets" id="id31" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">3</a><span class="fn-bracket">]</span></span>
<p>Lee SH, Hu W, Matulay JT, et al. <em>Tumor Evolution and Drug Response in Patient-Derived Organoid Models of Bladder Cancer.</em> Cell. 2018;173(2):515-528.e17. doi:<a class="reference external" href="https://doi.org/10.1016/j.cell.2018.03.017">10.1016/j.cell.2018.03.017</a></p>
</aside>
<aside class="footnote brackets" id="id32" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id4">4</a><span class="fn-bracket">]</span></span>
<p>Barretina J, Caponigro G, Stransky N, et al. <em>The Cancer Cell Line Encyclopedia enables predictive modelling of anticancer drug sensitivity.</em> Nature. 2012;483(7391):603-607. doi:<a class="reference external" href="https://doi.org/10.1038/nature11003">10.1038/nature11003</a></p>
</aside>
<aside class="footnote brackets" id="id33" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id6">5</a><span class="fn-bracket">]</span></span>
<p>Lindgren CM, Adams DW, Kimball B, et al. <em>Simplified and Unified Access to Cancer Proteogenomic Data.</em> J Proteome Res. 2021;20(4):1902-1910. doi:<a class="reference external" href="https://doi.org/10.1021/acs.jproteome.0c00919">10.1021/acs.jproteome.0c00919</a></p>
</aside>
<aside class="footnote brackets" id="id34" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id7">6</a><span class="fn-bracket">]</span></span>
<p>Rees MG, Seashore-Ludlow B, Cheah JH, et al. <em>Correlating chemical sensitivity and basal gene expression reveals mechanism of action.</em> Nat Chem Biol. 2016;12(2):109-116. doi:<a class="reference external" href="https://doi.org/10.1038/nchembio.1986">10.1038/nchembio.1986</a></p>
</aside>
<aside class="footnote brackets" id="id35" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id8">7</a><span class="fn-bracket">]</span></span>
<p>Seashore-Ludlow B, Rees MG, Cheah JH, et al. <em>Harnessing Connectivity in a Large-Scale Small-Molecule Sensitivity Dataset.</em> Cancer Discov. 2015;5(11):1210-1223. doi:<a class="reference external" href="https://doi.org/10.1158/2159-8290.CD-15-0235">10.1158/2159-8290.CD-15-0235</a></p>
</aside>
<aside class="footnote brackets" id="id36" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id9">8</a><span class="fn-bracket">]</span></span>
<p>Basu A, Bodycombe NE, Cheah JH, et al. <em>An interactive resource to identify cancer genetic and lineage dependencies targeted by small molecules.</em> Cell. 2013;154(5):1151-1161. doi:<a class="reference external" href="https://doi.org/10.1016/j.cell.2013.08.003">10.1016/j.cell.2013.08.003</a></p>
</aside>
<aside class="footnote brackets" id="id37" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id10">9</a><span class="fn-bracket">]</span></span>
<p>Mpindi JP, Yadav B, Östling P, et al. <em>Consistency in drug response profiling.</em> Nature. 2016;540(7631):E5-E6. doi:<a class="reference external" href="https://doi.org/10.1038/nature20171">10.1038/nature20171</a></p>
</aside>
<aside class="footnote brackets" id="id38" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id11">10</a><span class="fn-bracket">]</span></span>
<p>Pemovska T, Kontro M, Yadav B, et al. <em>Individualized systems medicine strategy to tailor treatments for patients with chemorefractory acute myeloid leukemia.</em> Cancer Discov. 2013;3(12):1416-1429. doi:<a class="reference external" href="https://doi.org/10.1158/2159-8290.CD-13-0350">10.1159/2159-8290.CD-13-0350</a></p>
</aside>
<aside class="footnote brackets" id="id39" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id20">11</a><span class="fn-bracket">]</span></span>
<p>Human Cancer Models Initiative (HCMI). dbGaP accession phs001486. <a class="reference external" href="https://cancer.gov/ccg/research/functional-genomics/hcmi">https://cancer.gov/ccg/research/functional-genomics/hcmi</a></p>
</aside>
<aside class="footnote brackets" id="id40" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id22">12</a><span class="fn-bracket">]</span></span>
<p>Dehner C, Moon CI, Zhang X, et al. <em>Chromosome 8 gain is associated with high-grade transformation in MPNST.</em> JCI Insight. 2021;6(6):e146351. doi:<a class="reference external" href="https://doi.org/10.1172/jci.insight.146351">10.1172/jci.insight.146351</a></p>
</aside>
<aside class="footnote brackets" id="id41" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id23">13</a><span class="fn-bracket">]</span></span>
<p>Shoemaker RH. <em>The NCI60 human tumour cell line anticancer drug screen.</em> Nat Rev Cancer. 2006;6(10):813-823. doi:<a class="reference external" href="https://doi.org/10.1038/nrc1951">10.1038/nrc1951</a></p>
</aside>
<aside class="footnote brackets" id="id42" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id25">14</a><span class="fn-bracket">]</span></span>
<p>Tiriac H, Belleau P, Engle DD, et al. <em>Organoid Profiling Identifies Common Responders to Chemotherapy in Pancreatic Cancer.</em> Cancer Discov. 2018;8(9):1112-1129. doi:<a class="reference external" href="https://doi.org/10.1158/2159-8290.CD-18-0349">10.1158/2159-8290.CD-18-0349</a></p>
</aside>
<aside class="footnote brackets" id="id43" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id26">15</a><span class="fn-bracket">]</span></span>
<p>Corsello SM, Nagari RT, Spangler RD, et al. <em>Discovering the anti-cancer potential of non-oncology drugs by systematic viability profiling.</em> Nat Cancer. 2020;1(2):235-248. doi:<a class="reference external" href="https://doi.org/10.1038/s43018-019-0018-6">10.1038/s43018-019-0018-6</a></p>
</aside>
<aside class="footnote brackets" id="id44" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id27">16</a><span class="fn-bracket">]</span></span>
<p>Yu C, Mannan AM, Yvone GM, et al. <em>High-throughput identification of genotype-specific cancer vulnerabilities in mixtures of barcoded tumor cell lines.</em> Nat Biotechnol. 2016;34(4):419-423. doi:<a class="reference external" href="https://doi.org/10.1038/nbt.3460">10.1038/nbt.3460</a></p>
</aside>
<aside class="footnote brackets" id="id45" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id28">17</a><span class="fn-bracket">]</span></span>
<p>Al Shihabi A, Tebon PJ, Nguyen HTL, et al. <em>The landscape of drug sensitivity and resistance in sarcoma.</em> Cell Stem Cell. 2024;31(10):1524-1542.e4. doi:<a class="reference external" href="https://doi.org/10.1016/j.stem.2024.08.010">10.1016/j.stem.2024.08.010</a></p>
</aside>
<aside class="footnote brackets" id="id46" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id5">18</a><span class="fn-bracket">]</span></span>
<p>van de Wetering M, Francies HE, Francis JM, et al. <em>Prospective derivation of a living organoid biobank of colorectal cancer patients.</em> Cell. 2015;161(4):933-945. doi:<a class="reference external" href="https://doi.org/10.1016/j.cell.2015.03.053">10.1016/j.cell.2015.03.053</a></p>
</aside>
<aside class="footnote brackets" id="id47" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id21">19</a><span class="fn-bracket">]</span></span>
<p>Ji S, Feng L, Fu Z, et al. <em>Pharmaco-proteogenomic characterization of liver cancer organoids for precision oncology.</em> Sci Transl Med. 2023;15(706):eadg3358. doi:<a class="reference external" href="https://doi.org/10.1126/scitranslmed.adg3358">10.1126/scitranslmed.adg3358</a></p>
</aside>
<aside class="footnote brackets" id="id48" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id24">20</a><span class="fn-bracket">]</span></span>
<p>Gao H, Korn JM, Ferretti S, et al. <em>High-throughput screening using patient-derived tumor xenografts to predict clinical trial drug response.</em> Nat Med. 2015;21(11):1318–1325. doi:<a class="reference external" href="https://doi.org/10.1038/nm.3954">10.1038/nm.3954</a></p>
</aside>
<aside class="footnote brackets" id="id49" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id18">21</a><span class="fn-bracket">]</span></span>
<p>Haverty PM, Lin E, Tan J, et al. <em>Reproducible pharmacogenomic profiling of cancer cell line panels.</em> Nature. 2016;533(7603):333–337. doi:<a class="reference external" href="https://doi.org/10.1038/nature17987">10.1038/nature17987</a></p>
</aside>
<aside class="footnote brackets" id="id50" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id19">22</a><span class="fn-bracket">]</span></span>
<p>Klijn C, Durinck S, Stawiski EW, et al. <em>A comprehensive transcriptional portrait of human cancer cell lines.</em> Nat Biotechnol. 2015;33(3):306–312. doi:<a class="reference external" href="https://doi.org/10.1038/nbt.3080">10.1038/nbt.3080</a></p>
</aside>
<aside class="footnote brackets" id="id51" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span>23<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id12">1</a>,<a role="doc-backlink" href="#id15">2</a>)</span>
<p>Garnett MJ, Edelman EJ, Heidorn SJ, et al. <em>Systematic identification of genomic markers of drug sensitivity in cancer cells.</em> Nature. 2012;483(7391):570–575. doi:<a class="reference external" href="https://doi.org/10.1038/nature11005">10.1038/nature11005</a></p>
</aside>
<aside class="footnote brackets" id="id52" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span>24<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id13">1</a>,<a role="doc-backlink" href="#id16">2</a>)</span>
<p>Iorio F, Knijnenburg TA, Vis DJ, et al. <em>A Landscape of Pharmacogenomic Interactions in Cancer.</em> Cell. 2016;166(3):740–754. doi:<a class="reference external" href="https://doi.org/10.1016/j.cell.2016.06.017">10.1016/j.cell.2016.06.017</a></p>
</aside>
<aside class="footnote brackets" id="id53" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span>25<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id14">1</a>,<a role="doc-backlink" href="#id17">2</a>)</span>
<p>Yang W, Soares J, Greninger P, et al. <em>Genomics of Drug Sensitivity in Cancer (GDSC): a resource for therapeutic biomarker discovery in cancer cells.</em> Nucleic Acids Res. 2013;41(Database issue):D955–D961. doi:<a class="reference external" href="https://doi.org/10.1093/nar/gks1111">10.1093/nar/gks1111</a></p>
</aside>
</aside>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/coderdata_logo3.jpg" alt="Logo of CoderData"/>
            </a></p><div  class="sphinxdoc-sidebar">
    <h3>Resource Links</h3>
    <ul>
        <li><a href="https://github.com/PNNL-CompBio/coderdata/" target="_blank">Github</a></li>
        <li><a href="https://pypi.org/project/coderdata/" target="_blank">PYPI</a></li>
        <li><a href="https://figshare.com/articles/dataset/CODERData2_1_0/28823159" target="_blank">FigShare</a></li>
    </ul>
</div>

<div class="sphinxdoc-sidebar">
    <h3>Quick Access</h3>
    <ul>
        <li><a href="index.html">Home</a></li>
        <li><a href="usage.html">Usage</a></li>
        <li><a href="APIreference.html">API Reference</a></li>
        <li><a href="#">Datasets</a></li>
        <li><a href="_notebooks/docsite_tutorial.html">Tutorial</a></li>
        <li><a href="contribution_guide.html">Contributing</a></li>
    </ul>
</div>
  <div>
    <h3><a href="index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">Datasets Included</a><ul>
<li><a class="reference internal" href="#dataset-overview">Dataset Overview</a></li>
<li><a class="reference internal" href="#dataset-summary-statistics">Dataset Summary Statistics</a></li>
<li><a class="reference internal" href="#drug-curve-metrics-collected">Drug Curve Metrics Collected</a></li>
<li><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
</ul>

  </div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="Related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="contribution.html" title="How to Contribute"
             >next</a> |</li>
        <li class="right" >
          <a href="APIreference.html" title="API Reference"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">CoderData</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Datasets Included</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2025, Sara Gosline, Yannick Mahlich, Jeremy Jacobson, Marissa Andrews.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.4.7.
    </div>
  </body>
</html>